{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7952a075-7d7e-4fbc-aafa-539b21d7a47a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available tables:\n",
      "- action_type\n",
      "- assay_type\n",
      "- chembl_id_lookup\n",
      "- confidence_score_lookup\n",
      "- curation_lookup\n",
      "- chembl_release\n",
      "- source\n",
      "- relationship_type\n",
      "- target_type\n",
      "- variant_sequences\n",
      "- bioassay_ontology\n",
      "- data_validity_lookup\n",
      "- activity_smid\n",
      "- activity_stds_lookup\n",
      "- assay_classification\n",
      "- atc_classification\n",
      "- bio_component_sequences\n",
      "- component_sequences\n",
      "- protein_classification\n",
      "- domains\n",
      "- go_classification\n",
      "- structural_alert_sets\n",
      "- products\n",
      "- frac_classification\n",
      "- hrac_classification\n",
      "- irac_classification\n",
      "- research_stem\n",
      "- organism_class\n",
      "- patent_use_codes\n",
      "- usan_stems\n",
      "- version\n",
      "- cell_dictionary\n",
      "- docs\n",
      "- target_dictionary\n",
      "- tissue_dictionary\n",
      "- molecule_dictionary\n",
      "- activity_supp\n",
      "- component_class\n",
      "- component_domains\n",
      "- component_go\n",
      "- component_synonyms\n",
      "- structural_alerts\n",
      "- defined_daily_dose\n",
      "- product_patents\n",
      "- protein_class_synonyms\n",
      "- research_companies\n",
      "- assays\n",
      "- compound_records\n",
      "- binding_sites\n",
      "- biotherapeutics\n",
      "- compound_properties\n",
      "- compound_structural_alerts\n",
      "- compound_structures\n",
      "- molecule_atc_classification\n",
      "- molecule_frac_classification\n",
      "- molecule_hierarchy\n",
      "- molecule_hrac_classification\n",
      "- molecule_irac_classification\n",
      "- molecule_synonyms\n",
      "- target_components\n",
      "- target_relations\n",
      "- activities\n",
      "- assay_class_map\n",
      "- assay_parameters\n",
      "- biotherapeutic_components\n",
      "- drug_indication\n",
      "- drug_mechanism\n",
      "- drug_warning\n",
      "- formulations\n",
      "- metabolism\n",
      "- site_components\n",
      "- activity_properties\n",
      "- activity_supp_map\n",
      "- indication_refs\n",
      "- ligand_eff\n",
      "- mechanism_refs\n",
      "- metabolism_refs\n",
      "- predicted_binding_domains\n",
      "- warning_refs\n",
      "- sqlite_stat1\n"
     ]
    }
   ],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "\n",
    "# Path to your database\n",
    "db_path = 'chembl_35/chembl_35_sqlite/chembl_35.db'\n",
    "\n",
    "# Connect\n",
    "conn = sqlite3.connect(db_path)\n",
    "cursor = conn.cursor()\n",
    "\n",
    "# List available tables\n",
    "tables = cursor.execute(\"SELECT name FROM sqlite_master WHERE type='table';\").fetchall()\n",
    "print(\"Available tables:\")\n",
    "for table in tables:\n",
    "    print(\"-\", table[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9eea2f7f-b180-4f82-980d-701f5c237942",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chembl_id</th>\n",
       "      <th>canonical_smiles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL6329</td>\n",
       "      <td>Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccccc1Cl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL6328</td>\n",
       "      <td>Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccc(C#N)cc1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL265667</td>\n",
       "      <td>Cc1cc(-n2ncc(=O)[nH]c2=O)cc(C)c1C(O)c1ccc(Cl)cc1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL6362</td>\n",
       "      <td>Cc1ccc(C(=O)c2ccc(-n3ncc(=O)[nH]c3=O)cc2)cc1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL267864</td>\n",
       "      <td>Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccc(Cl)cc1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chembl_id                                  canonical_smiles\n",
       "0    CHEMBL6329      Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccccc1Cl\n",
       "1    CHEMBL6328   Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccc(C#N)cc1\n",
       "2  CHEMBL265667  Cc1cc(-n2ncc(=O)[nH]c2=O)cc(C)c1C(O)c1ccc(Cl)cc1\n",
       "3    CHEMBL6362      Cc1ccc(C(=O)c2ccc(-n3ncc(=O)[nH]c3=O)cc2)cc1\n",
       "4  CHEMBL267864    Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccc(Cl)cc1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "SELECT md.chembl_id, cs.canonical_smiles\n",
    "FROM molecule_dictionary md\n",
    "JOIN compound_structures cs ON md.molregno = cs.molregno\n",
    "LIMIT 10;\n",
    "\"\"\"\n",
    "df_smiles = pd.read_sql_query(query, conn)\n",
    "df_smiles.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0c135197-a663-4b83-99a9-777136f0eac2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "📊 Columns in table `activities`:\n",
      "\n",
      "                     name           type\n",
      "0             activity_id         BIGINT\n",
      "1                assay_id         BIGINT\n",
      "2                  doc_id         BIGINT\n",
      "3               record_id         BIGINT\n",
      "4                molregno         BIGINT\n",
      "5       standard_relation    VARCHAR(50)\n",
      "6          standard_value        NUMERIC\n",
      "7          standard_units   VARCHAR(100)\n",
      "8           standard_flag       SMALLINT\n",
      "9           standard_type   VARCHAR(250)\n",
      "10       activity_comment  VARCHAR(4000)\n",
      "11  data_validity_comment    VARCHAR(30)\n",
      "12    potential_duplicate       SMALLINT\n",
      "13          pchembl_value  NUMERIC(4, 2)\n",
      "14           bao_endpoint    VARCHAR(11)\n",
      "15               uo_units    VARCHAR(10)\n",
      "16             qudt_units    VARCHAR(70)\n",
      "17                   toid        INTEGER\n",
      "18            upper_value        NUMERIC\n",
      "19   standard_upper_value        NUMERIC\n",
      "20                 src_id        INTEGER\n",
      "21                   type   VARCHAR(250)\n",
      "22               relation    VARCHAR(50)\n",
      "23                  value        NUMERIC\n",
      "24                  units   VARCHAR(100)\n",
      "25             text_value  VARCHAR(1000)\n",
      "26    standard_text_value  VARCHAR(1000)\n",
      "27            action_type    VARCHAR(50)\n",
      "\n",
      "📊 Columns in table `molecule_dictionary`:\n",
      "\n",
      "                    name           type\n",
      "0               molregno         BIGINT\n",
      "1              pref_name   VARCHAR(255)\n",
      "2              chembl_id    VARCHAR(20)\n",
      "3              max_phase  NUMERIC(2, 1)\n",
      "4       therapeutic_flag       SMALLINT\n",
      "5       dosed_ingredient       SMALLINT\n",
      "6         structure_type    VARCHAR(10)\n",
      "7           chebi_par_id         BIGINT\n",
      "8          molecule_type    VARCHAR(30)\n",
      "9         first_approval        INTEGER\n",
      "10                  oral       SMALLINT\n",
      "11            parenteral       SMALLINT\n",
      "12               topical       SMALLINT\n",
      "13     black_box_warning       SMALLINT\n",
      "14       natural_product       SMALLINT\n",
      "15        first_in_class       SMALLINT\n",
      "16             chirality       SMALLINT\n",
      "17               prodrug       SMALLINT\n",
      "18        inorganic_flag       SMALLINT\n",
      "19             usan_year        INTEGER\n",
      "20     availability_type       SMALLINT\n",
      "21             usan_stem    VARCHAR(50)\n",
      "22          polymer_flag       SMALLINT\n",
      "23          usan_substem    VARCHAR(50)\n",
      "24  usan_stem_definition  VARCHAR(1000)\n",
      "25      indication_class  VARCHAR(1000)\n",
      "26        withdrawn_flag       SMALLINT\n",
      "27        chemical_probe       SMALLINT\n",
      "28                orphan       SMALLINT\n",
      "\n",
      "📊 Columns in table `compound_structures`:\n",
      "\n",
      "                 name           type\n",
      "0            molregno         BIGINT\n",
      "1             molfile           TEXT\n",
      "2      standard_inchi  VARCHAR(4000)\n",
      "3  standard_inchi_key    VARCHAR(27)\n",
      "4    canonical_smiles  VARCHAR(4000)\n",
      "\n",
      "📊 Columns in table `target_dictionary`:\n",
      "\n",
      "                 name          type\n",
      "0                 tid        BIGINT\n",
      "1         target_type   VARCHAR(30)\n",
      "2           pref_name  VARCHAR(200)\n",
      "3              tax_id        BIGINT\n",
      "4            organism  VARCHAR(150)\n",
      "5           chembl_id   VARCHAR(20)\n",
      "6  species_group_flag      SMALLINT\n",
      "\n",
      "📊 Columns in table `drug_indication`:\n",
      "\n",
      "                name           type\n",
      "0         drugind_id         BIGINT\n",
      "1          record_id         BIGINT\n",
      "2           molregno         BIGINT\n",
      "3  max_phase_for_ind  NUMERIC(2, 1)\n",
      "4            mesh_id    VARCHAR(20)\n",
      "5       mesh_heading   VARCHAR(200)\n",
      "6             efo_id    VARCHAR(20)\n",
      "7           efo_term   VARCHAR(200)\n",
      "\n",
      "📊 Columns in table `assays`:\n",
      "\n",
      "                          name           type\n",
      "0                     assay_id         BIGINT\n",
      "1                       doc_id         BIGINT\n",
      "2                  description  VARCHAR(4000)\n",
      "3                   assay_type     VARCHAR(1)\n",
      "4              assay_test_type    VARCHAR(20)\n",
      "5               assay_category    VARCHAR(50)\n",
      "6               assay_organism   VARCHAR(250)\n",
      "7                 assay_tax_id         BIGINT\n",
      "8                 assay_strain   VARCHAR(200)\n",
      "9                 assay_tissue   VARCHAR(100)\n",
      "10             assay_cell_type   VARCHAR(100)\n",
      "11  assay_subcellular_fraction   VARCHAR(100)\n",
      "12                         tid         BIGINT\n",
      "13           relationship_type     VARCHAR(1)\n",
      "14            confidence_score       SMALLINT\n",
      "15                  curated_by    VARCHAR(32)\n",
      "16                      src_id        INTEGER\n",
      "17                src_assay_id    VARCHAR(50)\n",
      "18                   chembl_id    VARCHAR(20)\n",
      "19                     cell_id         BIGINT\n",
      "20                  bao_format    VARCHAR(11)\n",
      "21                   tissue_id         BIGINT\n",
      "22                  variant_id         BIGINT\n",
      "23                        aidx   VARCHAR(200)\n",
      "24                 assay_group   VARCHAR(200)\n"
     ]
    }
   ],
   "source": [
    "def show_columns(table_name):\n",
    "    query = f\"PRAGMA table_info({table_name})\"\n",
    "    cols = pd.read_sql_query(query, conn)\n",
    "    print(f\"\\n📊 Columns in table `{table_name}`:\\n\")\n",
    "    print(cols[['name', 'type']])\n",
    "\n",
    "# List columns in key tables\n",
    "for table in [\n",
    "    'activities',\n",
    "    'molecule_dictionary',\n",
    "    'compound_structures',\n",
    "    'target_dictionary',\n",
    "    'drug_indication',\n",
    "    'assays'\n",
    "]:\n",
    "    show_columns(table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e5e18984-3465-4e8a-b332-3a6bb47d6f75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assay_id</th>\n",
       "      <th>molregno</th>\n",
       "      <th>drug_id</th>\n",
       "      <th>drug_name</th>\n",
       "      <th>standard_type</th>\n",
       "      <th>standard_value</th>\n",
       "      <th>standard_units</th>\n",
       "      <th>standard_relation</th>\n",
       "      <th>tid</th>\n",
       "      <th>target_name</th>\n",
       "      <th>target_organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>54505</td>\n",
       "      <td>180094</td>\n",
       "      <td>CHEMBL113081</td>\n",
       "      <td>None</td>\n",
       "      <td>IC50</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>nM</td>\n",
       "      <td>&gt;</td>\n",
       "      <td>63</td>\n",
       "      <td>DNA topoisomerase II alpha</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>83907</td>\n",
       "      <td>182268</td>\n",
       "      <td>CHEMBL324340</td>\n",
       "      <td>None</td>\n",
       "      <td>IC50</td>\n",
       "      <td>2500.0</td>\n",
       "      <td>nM</td>\n",
       "      <td>=</td>\n",
       "      <td>11653</td>\n",
       "      <td>Heparanase</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>88152</td>\n",
       "      <td>182268</td>\n",
       "      <td>CHEMBL324340</td>\n",
       "      <td>None</td>\n",
       "      <td>IC50</td>\n",
       "      <td>50000.0</td>\n",
       "      <td>nM</td>\n",
       "      <td>&gt;</td>\n",
       "      <td>22221</td>\n",
       "      <td>NON-PROTEIN TARGET</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>83907</td>\n",
       "      <td>182855</td>\n",
       "      <td>CHEMBL109600</td>\n",
       "      <td>None</td>\n",
       "      <td>IC50</td>\n",
       "      <td>9000.0</td>\n",
       "      <td>nM</td>\n",
       "      <td>=</td>\n",
       "      <td>11653</td>\n",
       "      <td>Heparanase</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>154606</td>\n",
       "      <td>252199</td>\n",
       "      <td>CHEMBL357278</td>\n",
       "      <td>None</td>\n",
       "      <td>IC50</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>nM</td>\n",
       "      <td>=</td>\n",
       "      <td>10483</td>\n",
       "      <td>Palmitoyl-CoA oxidase</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   assay_id  molregno       drug_id drug_name standard_type  standard_value  \\\n",
       "0     54505    180094  CHEMBL113081      None          IC50        100000.0   \n",
       "1     83907    182268  CHEMBL324340      None          IC50          2500.0   \n",
       "2     88152    182268  CHEMBL324340      None          IC50         50000.0   \n",
       "3     83907    182855  CHEMBL109600      None          IC50          9000.0   \n",
       "4    154606    252199  CHEMBL357278      None          IC50          4000.0   \n",
       "\n",
       "  standard_units standard_relation    tid                 target_name  \\\n",
       "0             nM                 >     63  DNA topoisomerase II alpha   \n",
       "1             nM                 =  11653                  Heparanase   \n",
       "2             nM                 >  22221          NON-PROTEIN TARGET   \n",
       "3             nM                 =  11653                  Heparanase   \n",
       "4             nM                 =  10483       Palmitoyl-CoA oxidase   \n",
       "\n",
       "     target_organism  \n",
       "0       Homo sapiens  \n",
       "1       Homo sapiens  \n",
       "2               None  \n",
       "3       Homo sapiens  \n",
       "4  Rattus norvegicus  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "SELECT \n",
    "    a.assay_id,\n",
    "    a.molregno,\n",
    "    md.chembl_id AS drug_id,\n",
    "    md.pref_name AS drug_name,\n",
    "    a.standard_type,\n",
    "    a.standard_value,\n",
    "    a.standard_units,\n",
    "    a.standard_relation,\n",
    "    ass.tid,\n",
    "    td.pref_name AS target_name,\n",
    "    td.organism AS target_organism\n",
    "FROM activities a\n",
    "JOIN molecule_dictionary md ON a.molregno = md.molregno\n",
    "JOIN assays ass ON a.assay_id = ass.assay_id\n",
    "JOIN target_dictionary td ON ass.tid = td.tid\n",
    "WHERE a.standard_type IN ('IC50', 'Ki', 'Kd')\n",
    "  AND a.standard_value IS NOT NULL\n",
    "  AND a.standard_units = 'nM'\n",
    "LIMIT 100;\n",
    "\"\"\"\n",
    "df_activity = pd.read_sql_query(query, conn)\n",
    "df_activity.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e92b6151-26c1-4dc9-a595-2183e51a2cb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug_id</th>\n",
       "      <th>canonical_smiles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL6329</td>\n",
       "      <td>Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccccc1Cl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL6328</td>\n",
       "      <td>Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccc(C#N)cc1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL265667</td>\n",
       "      <td>Cc1cc(-n2ncc(=O)[nH]c2=O)cc(C)c1C(O)c1ccc(Cl)cc1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL6362</td>\n",
       "      <td>Cc1ccc(C(=O)c2ccc(-n3ncc(=O)[nH]c3=O)cc2)cc1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL267864</td>\n",
       "      <td>Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccc(Cl)cc1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        drug_id                                  canonical_smiles\n",
       "0    CHEMBL6329      Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccccc1Cl\n",
       "1    CHEMBL6328   Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccc(C#N)cc1\n",
       "2  CHEMBL265667  Cc1cc(-n2ncc(=O)[nH]c2=O)cc(C)c1C(O)c1ccc(Cl)cc1\n",
       "3    CHEMBL6362      Cc1ccc(C(=O)c2ccc(-n3ncc(=O)[nH]c3=O)cc2)cc1\n",
       "4  CHEMBL267864    Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccc(Cl)cc1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "SELECT \n",
    "    md.chembl_id AS drug_id,\n",
    "    cs.canonical_smiles\n",
    "FROM molecule_dictionary md\n",
    "JOIN compound_structures cs ON md.molregno = cs.molregno\n",
    "WHERE cs.canonical_smiles IS NOT NULL\n",
    "LIMIT 100;\n",
    "\"\"\"\n",
    "df_smiles = pd.read_sql_query(query, conn)\n",
    "df_smiles.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9f580ce6-f667-4287-9301-c5331222d40c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug_id</th>\n",
       "      <th>mesh_id</th>\n",
       "      <th>mesh_heading</th>\n",
       "      <th>efo_id</th>\n",
       "      <th>efo_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL1201823</td>\n",
       "      <td>D045743</td>\n",
       "      <td>Scleroderma, Diffuse</td>\n",
       "      <td>EFO:0000404</td>\n",
       "      <td>diffuse scleroderma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL1201823</td>\n",
       "      <td>D001172</td>\n",
       "      <td>Arthritis, Rheumatoid</td>\n",
       "      <td>EFO:0000685</td>\n",
       "      <td>rheumatoid arthritis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL1201584</td>\n",
       "      <td>D009203</td>\n",
       "      <td>Myocardial Infarction</td>\n",
       "      <td>EFO:0000612</td>\n",
       "      <td>myocardial infarction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL271227</td>\n",
       "      <td>D011471</td>\n",
       "      <td>Prostatic Neoplasms</td>\n",
       "      <td>EFO:0001663</td>\n",
       "      <td>prostate carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL1566</td>\n",
       "      <td>D003920</td>\n",
       "      <td>Diabetes Mellitus</td>\n",
       "      <td>EFO:0000400</td>\n",
       "      <td>diabetes mellitus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         drug_id  mesh_id           mesh_heading       efo_id  \\\n",
       "0  CHEMBL1201823  D045743   Scleroderma, Diffuse  EFO:0000404   \n",
       "1  CHEMBL1201823  D001172  Arthritis, Rheumatoid  EFO:0000685   \n",
       "2  CHEMBL1201584  D009203  Myocardial Infarction  EFO:0000612   \n",
       "3   CHEMBL271227  D011471    Prostatic Neoplasms  EFO:0001663   \n",
       "4     CHEMBL1566  D003920      Diabetes Mellitus  EFO:0000400   \n",
       "\n",
       "                efo_term  \n",
       "0    diffuse scleroderma  \n",
       "1   rheumatoid arthritis  \n",
       "2  myocardial infarction  \n",
       "3     prostate carcinoma  \n",
       "4      diabetes mellitus  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "SELECT \n",
    "    md.chembl_id AS drug_id,\n",
    "    di.mesh_id,\n",
    "    di.mesh_heading,\n",
    "    di.efo_id,\n",
    "    di.efo_term\n",
    "FROM drug_indication di\n",
    "JOIN molecule_dictionary md ON di.molregno = md.molregno\n",
    "LIMIT 100;\n",
    "\"\"\"\n",
    "df_indications = pd.read_sql_query(query, conn)\n",
    "df_indications.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0affa2a9-6822-4205-ac04-3b1130472d47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "📊 Columns in `compound_properties`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>molregno</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mw_freebase</td>\n",
       "      <td>NUMERIC(9, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>alogp</td>\n",
       "      <td>NUMERIC(9, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hba</td>\n",
       "      <td>INTEGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hbd</td>\n",
       "      <td>INTEGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>psa</td>\n",
       "      <td>NUMERIC(9, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rtb</td>\n",
       "      <td>INTEGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ro3_pass</td>\n",
       "      <td>VARCHAR(3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>num_ro5_violations</td>\n",
       "      <td>SMALLINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cx_most_apka</td>\n",
       "      <td>NUMERIC(9, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cx_most_bpka</td>\n",
       "      <td>NUMERIC(9, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cx_logp</td>\n",
       "      <td>NUMERIC(9, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cx_logd</td>\n",
       "      <td>NUMERIC(9, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>molecular_species</td>\n",
       "      <td>VARCHAR(50)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>full_mwt</td>\n",
       "      <td>NUMERIC(9, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>aromatic_rings</td>\n",
       "      <td>INTEGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>heavy_atoms</td>\n",
       "      <td>INTEGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>qed_weighted</td>\n",
       "      <td>NUMERIC(3, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>mw_monoisotopic</td>\n",
       "      <td>NUMERIC(11, 4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>full_molformula</td>\n",
       "      <td>VARCHAR(100)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>hba_lipinski</td>\n",
       "      <td>INTEGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>hbd_lipinski</td>\n",
       "      <td>INTEGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>num_lipinski_ro5_violations</td>\n",
       "      <td>SMALLINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>np_likeness_score</td>\n",
       "      <td>NUMERIC(3, 2)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           name            type\n",
       "0                      molregno          BIGINT\n",
       "1                   mw_freebase   NUMERIC(9, 2)\n",
       "2                         alogp   NUMERIC(9, 2)\n",
       "3                           hba         INTEGER\n",
       "4                           hbd         INTEGER\n",
       "5                           psa   NUMERIC(9, 2)\n",
       "6                           rtb         INTEGER\n",
       "7                      ro3_pass      VARCHAR(3)\n",
       "8            num_ro5_violations        SMALLINT\n",
       "9                  cx_most_apka   NUMERIC(9, 2)\n",
       "10                 cx_most_bpka   NUMERIC(9, 2)\n",
       "11                      cx_logp   NUMERIC(9, 2)\n",
       "12                      cx_logd   NUMERIC(9, 2)\n",
       "13            molecular_species     VARCHAR(50)\n",
       "14                     full_mwt   NUMERIC(9, 2)\n",
       "15               aromatic_rings         INTEGER\n",
       "16                  heavy_atoms         INTEGER\n",
       "17                 qed_weighted   NUMERIC(3, 2)\n",
       "18              mw_monoisotopic  NUMERIC(11, 4)\n",
       "19              full_molformula    VARCHAR(100)\n",
       "20                 hba_lipinski         INTEGER\n",
       "21                 hbd_lipinski         INTEGER\n",
       "22  num_lipinski_ro5_violations        SMALLINT\n",
       "23            np_likeness_score   NUMERIC(3, 2)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "📊 Columns in `structural_alerts`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>alert_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>alert_set_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>alert_name</td>\n",
       "      <td>VARCHAR(100)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>smarts</td>\n",
       "      <td>VARCHAR(4000)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           name           type\n",
       "0      alert_id         BIGINT\n",
       "1  alert_set_id         BIGINT\n",
       "2    alert_name   VARCHAR(100)\n",
       "3        smarts  VARCHAR(4000)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "📊 Columns in `compound_structural_alerts`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cpd_str_alert_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>molregno</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>alert_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               name    type\n",
       "0  cpd_str_alert_id  BIGINT\n",
       "1          molregno  BIGINT\n",
       "2          alert_id  BIGINT"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "📊 Columns in `structural_alert_sets`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>alert_set_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>set_name</td>\n",
       "      <td>VARCHAR(100)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>priority</td>\n",
       "      <td>SMALLINT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           name          type\n",
       "0  alert_set_id        BIGINT\n",
       "1      set_name  VARCHAR(100)\n",
       "2      priority      SMALLINT"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "📊 Columns in `drug_warning`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>warning_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>record_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>molregno</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>warning_type</td>\n",
       "      <td>VARCHAR(20)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>warning_class</td>\n",
       "      <td>VARCHAR(100)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>warning_description</td>\n",
       "      <td>VARCHAR(4000)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>warning_country</td>\n",
       "      <td>VARCHAR(1000)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>warning_year</td>\n",
       "      <td>INTEGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>efo_term</td>\n",
       "      <td>VARCHAR(200)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>efo_id</td>\n",
       "      <td>VARCHAR(20)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>efo_id_for_warning_class</td>\n",
       "      <td>VARCHAR(20)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name           type\n",
       "0                 warning_id         BIGINT\n",
       "1                  record_id         BIGINT\n",
       "2                   molregno         BIGINT\n",
       "3               warning_type    VARCHAR(20)\n",
       "4              warning_class   VARCHAR(100)\n",
       "5        warning_description  VARCHAR(4000)\n",
       "6            warning_country  VARCHAR(1000)\n",
       "7               warning_year        INTEGER\n",
       "8                   efo_term   VARCHAR(200)\n",
       "9                     efo_id    VARCHAR(20)\n",
       "10  efo_id_for_warning_class    VARCHAR(20)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "📊 Columns in `metabolism`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>met_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>drug_record_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>substrate_record_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>metabolite_record_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pathway_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pathway_key</td>\n",
       "      <td>VARCHAR(50)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>enzyme_name</td>\n",
       "      <td>VARCHAR(200)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>enzyme_tid</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>met_conversion</td>\n",
       "      <td>VARCHAR(200)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>organism</td>\n",
       "      <td>VARCHAR(100)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tax_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>met_comment</td>\n",
       "      <td>VARCHAR(1000)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    name           type\n",
       "0                 met_id         BIGINT\n",
       "1         drug_record_id         BIGINT\n",
       "2    substrate_record_id         BIGINT\n",
       "3   metabolite_record_id         BIGINT\n",
       "4             pathway_id         BIGINT\n",
       "5            pathway_key    VARCHAR(50)\n",
       "6            enzyme_name   VARCHAR(200)\n",
       "7             enzyme_tid         BIGINT\n",
       "8         met_conversion   VARCHAR(200)\n",
       "9               organism   VARCHAR(100)\n",
       "10                tax_id         BIGINT\n",
       "11           met_comment  VARCHAR(1000)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "📊 Columns in `drug_mechanism`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mec_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>record_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>molregno</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mechanism_of_action</td>\n",
       "      <td>VARCHAR(250)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tid</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>site_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>action_type</td>\n",
       "      <td>VARCHAR(50)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>direct_interaction</td>\n",
       "      <td>SMALLINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>molecular_mechanism</td>\n",
       "      <td>SMALLINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>disease_efficacy</td>\n",
       "      <td>SMALLINT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>mechanism_comment</td>\n",
       "      <td>VARCHAR(2000)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>selectivity_comment</td>\n",
       "      <td>VARCHAR(1000)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>binding_site_comment</td>\n",
       "      <td>VARCHAR(1000)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>variant_id</td>\n",
       "      <td>BIGINT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    name           type\n",
       "0                 mec_id         BIGINT\n",
       "1              record_id         BIGINT\n",
       "2               molregno         BIGINT\n",
       "3    mechanism_of_action   VARCHAR(250)\n",
       "4                    tid         BIGINT\n",
       "5                site_id         BIGINT\n",
       "6            action_type    VARCHAR(50)\n",
       "7     direct_interaction       SMALLINT\n",
       "8    molecular_mechanism       SMALLINT\n",
       "9       disease_efficacy       SMALLINT\n",
       "10     mechanism_comment  VARCHAR(2000)\n",
       "11   selectivity_comment  VARCHAR(1000)\n",
       "12  binding_site_comment  VARCHAR(1000)\n",
       "13            variant_id         BIGINT"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def show_columns(table_name):\n",
    "    print(f\"\\n📊 Columns in `{table_name}`\")\n",
    "    query = f\"PRAGMA table_info({table_name})\"\n",
    "    df = pd.read_sql_query(query, conn)\n",
    "    display(df[['name', 'type']])\n",
    "\n",
    "for table in [\n",
    "    'compound_properties',\n",
    "    'structural_alerts',\n",
    "    'compound_structural_alerts',\n",
    "    'structural_alert_sets',\n",
    "    'drug_warning',\n",
    "    'metabolism',\n",
    "    'drug_mechanism'\n",
    "]:\n",
    "    show_columns(table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ce20c796-6209-431a-8b3f-19257a90822f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "📦 Table: action_type\n",
      "          name          type\n",
      "0  action_type   VARCHAR(50)\n",
      "1  description  VARCHAR(200)\n",
      "2  parent_type   VARCHAR(50)\n",
      "\n",
      "📦 Table: assay_type\n",
      "         name          type\n",
      "0  assay_type    VARCHAR(1)\n",
      "1  assay_desc  VARCHAR(250)\n",
      "\n",
      "📦 Table: chembl_id_lookup\n",
      "          name         type\n",
      "0    chembl_id  VARCHAR(20)\n",
      "1  entity_type  VARCHAR(50)\n",
      "2    entity_id       BIGINT\n",
      "3       status  VARCHAR(10)\n",
      "4  last_active      INTEGER\n",
      "\n",
      "📦 Table: confidence_score_lookup\n",
      "               name          type\n",
      "0  confidence_score      SMALLINT\n",
      "1       description  VARCHAR(100)\n",
      "2    target_mapping   VARCHAR(30)\n",
      "\n",
      "📦 Table: curation_lookup\n",
      "          name          type\n",
      "0   curated_by   VARCHAR(32)\n",
      "1  description  VARCHAR(100)\n",
      "\n",
      "📦 Table: chembl_release\n",
      "                name         type\n",
      "0  chembl_release_id      INTEGER\n",
      "1     chembl_release  VARCHAR(20)\n",
      "2      creation_date     DATETIME\n",
      "\n",
      "📦 Table: source\n",
      "              name           type\n",
      "0           src_id        INTEGER\n",
      "1  src_description   VARCHAR(500)\n",
      "2   src_short_name    VARCHAR(20)\n",
      "3      src_comment  VARCHAR(1200)\n",
      "4          src_url   VARCHAR(200)\n",
      "\n",
      "📦 Table: relationship_type\n",
      "                name          type\n",
      "0  relationship_type    VARCHAR(1)\n",
      "1  relationship_desc  VARCHAR(250)\n",
      "\n",
      "📦 Table: target_type\n",
      "          name          type\n",
      "0  target_type   VARCHAR(30)\n",
      "1  target_desc  VARCHAR(250)\n",
      "2  parent_type   VARCHAR(25)\n",
      "\n",
      "📦 Table: variant_sequences\n",
      "         name           type\n",
      "0  variant_id         BIGINT\n",
      "1    mutation  VARCHAR(2000)\n",
      "2   accession    VARCHAR(25)\n",
      "3     version         BIGINT\n",
      "4     isoform         BIGINT\n",
      "5    sequence           TEXT\n",
      "6    organism   VARCHAR(200)\n",
      "7      tax_id         BIGINT\n",
      "\n",
      "📦 Table: bioassay_ontology\n",
      "     name          type\n",
      "0  bao_id   VARCHAR(11)\n",
      "1   label  VARCHAR(100)\n",
      "\n",
      "📦 Table: data_validity_lookup\n",
      "                    name          type\n",
      "0  data_validity_comment   VARCHAR(30)\n",
      "1            description  VARCHAR(200)\n",
      "\n",
      "📦 Table: activity_smid\n",
      "   name    type\n",
      "0  smid  BIGINT\n",
      "\n",
      "📦 Table: activity_stds_lookup\n",
      "               name             type\n",
      "0        std_act_id           BIGINT\n",
      "1     standard_type     VARCHAR(250)\n",
      "2        definition     VARCHAR(500)\n",
      "3    standard_units     VARCHAR(100)\n",
      "4  normal_range_min  NUMERIC(24, 12)\n",
      "5  normal_range_max  NUMERIC(24, 12)\n",
      "\n",
      "📦 Table: assay_classification\n",
      "             name           type\n",
      "0  assay_class_id         BIGINT\n",
      "1              l1   VARCHAR(100)\n",
      "2              l2   VARCHAR(100)\n",
      "3              l3  VARCHAR(1000)\n",
      "4      class_type    VARCHAR(50)\n",
      "5          source    VARCHAR(50)\n",
      "\n",
      "📦 Table: atc_classification\n",
      "                 name           type\n",
      "0            who_name  VARCHAR(2000)\n",
      "1              level1    VARCHAR(10)\n",
      "2              level2    VARCHAR(10)\n",
      "3              level3    VARCHAR(10)\n",
      "4              level4    VARCHAR(10)\n",
      "5              level5    VARCHAR(10)\n",
      "6  level1_description  VARCHAR(2000)\n",
      "7  level2_description  VARCHAR(2000)\n",
      "8  level3_description  VARCHAR(2000)\n",
      "9  level4_description  VARCHAR(2000)\n",
      "\n",
      "📦 Table: bio_component_sequences\n",
      "              name          type\n",
      "0     component_id        BIGINT\n",
      "1   component_type   VARCHAR(50)\n",
      "2      description  VARCHAR(200)\n",
      "3         sequence          TEXT\n",
      "4  sequence_md5sum   VARCHAR(32)\n",
      "5           tax_id        BIGINT\n",
      "6         organism  VARCHAR(150)\n",
      "\n",
      "📦 Table: component_sequences\n",
      "              name          type\n",
      "0     component_id        BIGINT\n",
      "1   component_type   VARCHAR(50)\n",
      "2        accession   VARCHAR(25)\n",
      "3         sequence          TEXT\n",
      "4  sequence_md5sum   VARCHAR(32)\n",
      "5      description  VARCHAR(200)\n",
      "6           tax_id        BIGINT\n",
      "7         organism  VARCHAR(150)\n",
      "8        db_source   VARCHAR(25)\n",
      "9       db_version   VARCHAR(10)\n",
      "\n",
      "📦 Table: protein_classification\n",
      "                 name           type\n",
      "0    protein_class_id         BIGINT\n",
      "1           parent_id         BIGINT\n",
      "2           pref_name   VARCHAR(500)\n",
      "3          short_name    VARCHAR(50)\n",
      "4  protein_class_desc   VARCHAR(410)\n",
      "5          definition  VARCHAR(4000)\n",
      "6         class_level         BIGINT\n",
      "\n",
      "📦 Table: domains\n",
      "                 name          type\n",
      "0           domain_id        BIGINT\n",
      "1         domain_type   VARCHAR(20)\n",
      "2    source_domain_id   VARCHAR(20)\n",
      "3         domain_name  VARCHAR(100)\n",
      "4  domain_description  VARCHAR(500)\n",
      "\n",
      "📦 Table: go_classification\n",
      "           name           type\n",
      "0         go_id    VARCHAR(10)\n",
      "1  parent_go_id    VARCHAR(10)\n",
      "2     pref_name   VARCHAR(200)\n",
      "3   class_level       SMALLINT\n",
      "4        aspect     VARCHAR(1)\n",
      "5          path  VARCHAR(1000)\n",
      "\n",
      "📦 Table: structural_alert_sets\n",
      "           name          type\n",
      "0  alert_set_id        BIGINT\n",
      "1      set_name  VARCHAR(100)\n",
      "2      priority      SMALLINT\n",
      "\n",
      "📦 Table: products\n",
      "                   name          type\n",
      "0           dosage_form  VARCHAR(200)\n",
      "1                 route  VARCHAR(200)\n",
      "2            trade_name  VARCHAR(200)\n",
      "3         approval_date      DATETIME\n",
      "4               ad_type    VARCHAR(5)\n",
      "5                  oral      SMALLINT\n",
      "6               topical      SMALLINT\n",
      "7            parenteral      SMALLINT\n",
      "8     black_box_warning      SMALLINT\n",
      "9   applicant_full_name  VARCHAR(200)\n",
      "10    innovator_company      SMALLINT\n",
      "11           product_id   VARCHAR(30)\n",
      "12             nda_type   VARCHAR(10)\n",
      "\n",
      "📦 Table: frac_classification\n",
      "                  name           type\n",
      "0        frac_class_id         BIGINT\n",
      "1    active_ingredient   VARCHAR(500)\n",
      "2               level1     VARCHAR(2)\n",
      "3   level1_description  VARCHAR(2000)\n",
      "4               level2     VARCHAR(2)\n",
      "5   level2_description  VARCHAR(2000)\n",
      "6               level3     VARCHAR(6)\n",
      "7   level3_description  VARCHAR(2000)\n",
      "8               level4     VARCHAR(7)\n",
      "9   level4_description  VARCHAR(2000)\n",
      "10              level5     VARCHAR(8)\n",
      "11           frac_code     VARCHAR(4)\n",
      "\n",
      "📦 Table: hrac_classification\n",
      "                 name           type\n",
      "0       hrac_class_id         BIGINT\n",
      "1   active_ingredient   VARCHAR(500)\n",
      "2              level1     VARCHAR(2)\n",
      "3  level1_description  VARCHAR(2000)\n",
      "4              level2     VARCHAR(3)\n",
      "5  level2_description  VARCHAR(2000)\n",
      "6              level3     VARCHAR(5)\n",
      "7           hrac_code     VARCHAR(2)\n",
      "\n",
      "📦 Table: irac_classification\n",
      "                 name           type\n",
      "0       irac_class_id         BIGINT\n",
      "1   active_ingredient   VARCHAR(500)\n",
      "2              level1     VARCHAR(1)\n",
      "3  level1_description  VARCHAR(2000)\n",
      "4              level2     VARCHAR(3)\n",
      "5  level2_description  VARCHAR(2000)\n",
      "6              level3     VARCHAR(6)\n",
      "7  level3_description  VARCHAR(2000)\n",
      "8              level4     VARCHAR(8)\n",
      "9           irac_code     VARCHAR(3)\n",
      "\n",
      "📦 Table: research_stem\n",
      "            name         type\n",
      "0    res_stem_id       BIGINT\n",
      "1  research_stem  VARCHAR(20)\n",
      "\n",
      "📦 Table: organism_class\n",
      "     name          type\n",
      "0   oc_id        BIGINT\n",
      "1  tax_id        BIGINT\n",
      "2      l1  VARCHAR(200)\n",
      "3      l2  VARCHAR(200)\n",
      "4      l3  VARCHAR(200)\n",
      "\n",
      "📦 Table: patent_use_codes\n",
      "              name          type\n",
      "0  patent_use_code    VARCHAR(8)\n",
      "1       definition  VARCHAR(500)\n",
      "\n",
      "📦 Table: usan_stems\n",
      "           name           type\n",
      "0  usan_stem_id         BIGINT\n",
      "1          stem   VARCHAR(100)\n",
      "2      subgroup   VARCHAR(100)\n",
      "3    annotation  VARCHAR(2000)\n",
      "4    stem_class   VARCHAR(100)\n",
      "5   major_class   VARCHAR(100)\n",
      "\n",
      "📦 Table: version\n",
      "            name           type\n",
      "0           name    VARCHAR(50)\n",
      "1  creation_date       DATETIME\n",
      "2       comments  VARCHAR(2000)\n",
      "\n",
      "📦 Table: cell_dictionary\n",
      "                    name          type\n",
      "0                cell_id        BIGINT\n",
      "1              cell_name   VARCHAR(50)\n",
      "2       cell_description  VARCHAR(200)\n",
      "3     cell_source_tissue   VARCHAR(50)\n",
      "4   cell_source_organism  VARCHAR(150)\n",
      "5     cell_source_tax_id        BIGINT\n",
      "6                 clo_id   VARCHAR(11)\n",
      "7                 efo_id   VARCHAR(12)\n",
      "8         cellosaurus_id   VARCHAR(15)\n",
      "9            cl_lincs_id    VARCHAR(8)\n",
      "10             chembl_id   VARCHAR(20)\n",
      "11      cell_ontology_id   VARCHAR(10)\n",
      "\n",
      "📦 Table: docs\n",
      "                 name           type\n",
      "0              doc_id         BIGINT\n",
      "1             journal    VARCHAR(50)\n",
      "2                year        INTEGER\n",
      "3              volume    VARCHAR(50)\n",
      "4               issue    VARCHAR(50)\n",
      "5          first_page    VARCHAR(50)\n",
      "6           last_page    VARCHAR(50)\n",
      "7           pubmed_id         BIGINT\n",
      "8                 doi   VARCHAR(100)\n",
      "9           chembl_id    VARCHAR(20)\n",
      "10              title   VARCHAR(500)\n",
      "11           doc_type    VARCHAR(50)\n",
      "12            authors  VARCHAR(4000)\n",
      "13           abstract           TEXT\n",
      "14          patent_id    VARCHAR(20)\n",
      "15               ridx   VARCHAR(200)\n",
      "16             src_id        INTEGER\n",
      "17  chembl_release_id        INTEGER\n",
      "18            contact   VARCHAR(200)\n",
      "\n",
      "📦 Table: target_dictionary\n",
      "                 name          type\n",
      "0                 tid        BIGINT\n",
      "1         target_type   VARCHAR(30)\n",
      "2           pref_name  VARCHAR(200)\n",
      "3              tax_id        BIGINT\n",
      "4            organism  VARCHAR(150)\n",
      "5           chembl_id   VARCHAR(20)\n",
      "6  species_group_flag      SMALLINT\n",
      "\n",
      "📦 Table: tissue_dictionary\n",
      "        name          type\n",
      "0  tissue_id        BIGINT\n",
      "1  uberon_id   VARCHAR(15)\n",
      "2  pref_name  VARCHAR(200)\n",
      "3     efo_id   VARCHAR(20)\n",
      "4  chembl_id   VARCHAR(20)\n",
      "5     bto_id   VARCHAR(20)\n",
      "6  caloha_id    VARCHAR(7)\n",
      "\n",
      "📦 Table: molecule_dictionary\n",
      "                    name           type\n",
      "0               molregno         BIGINT\n",
      "1              pref_name   VARCHAR(255)\n",
      "2              chembl_id    VARCHAR(20)\n",
      "3              max_phase  NUMERIC(2, 1)\n",
      "4       therapeutic_flag       SMALLINT\n",
      "5       dosed_ingredient       SMALLINT\n",
      "6         structure_type    VARCHAR(10)\n",
      "7           chebi_par_id         BIGINT\n",
      "8          molecule_type    VARCHAR(30)\n",
      "9         first_approval        INTEGER\n",
      "10                  oral       SMALLINT\n",
      "11            parenteral       SMALLINT\n",
      "12               topical       SMALLINT\n",
      "13     black_box_warning       SMALLINT\n",
      "14       natural_product       SMALLINT\n",
      "15        first_in_class       SMALLINT\n",
      "16             chirality       SMALLINT\n",
      "17               prodrug       SMALLINT\n",
      "18        inorganic_flag       SMALLINT\n",
      "19             usan_year        INTEGER\n",
      "20     availability_type       SMALLINT\n",
      "21             usan_stem    VARCHAR(50)\n",
      "22          polymer_flag       SMALLINT\n",
      "23          usan_substem    VARCHAR(50)\n",
      "24  usan_stem_definition  VARCHAR(1000)\n",
      "25      indication_class  VARCHAR(1000)\n",
      "26        withdrawn_flag       SMALLINT\n",
      "27        chemical_probe       SMALLINT\n",
      "28                orphan       SMALLINT\n",
      "\n",
      "📦 Table: activity_supp\n",
      "                   name           type\n",
      "0                 as_id         BIGINT\n",
      "1                  rgid         BIGINT\n",
      "2                  smid         BIGINT\n",
      "3                  type   VARCHAR(250)\n",
      "4              relation    VARCHAR(50)\n",
      "5                 value        NUMERIC\n",
      "6                 units   VARCHAR(100)\n",
      "7            text_value  VARCHAR(1000)\n",
      "8         standard_type   VARCHAR(250)\n",
      "9     standard_relation    VARCHAR(50)\n",
      "10       standard_value        NUMERIC\n",
      "11       standard_units   VARCHAR(100)\n",
      "12  standard_text_value  VARCHAR(1000)\n",
      "13             comments  VARCHAR(4000)\n",
      "\n",
      "📦 Table: component_class\n",
      "               name    type\n",
      "0      component_id  BIGINT\n",
      "1  protein_class_id  BIGINT\n",
      "2     comp_class_id  BIGINT\n",
      "\n",
      "📦 Table: component_domains\n",
      "             name    type\n",
      "0        compd_id  BIGINT\n",
      "1       domain_id  BIGINT\n",
      "2    component_id  BIGINT\n",
      "3  start_position  BIGINT\n",
      "4    end_position  BIGINT\n",
      "\n",
      "📦 Table: component_go\n",
      "           name         type\n",
      "0    comp_go_id       BIGINT\n",
      "1  component_id       BIGINT\n",
      "2         go_id  VARCHAR(10)\n",
      "\n",
      "📦 Table: component_synonyms\n",
      "                name          type\n",
      "0         compsyn_id        BIGINT\n",
      "1       component_id        BIGINT\n",
      "2  component_synonym  VARCHAR(500)\n",
      "3           syn_type   VARCHAR(20)\n",
      "\n",
      "📦 Table: structural_alerts\n",
      "           name           type\n",
      "0      alert_id         BIGINT\n",
      "1  alert_set_id         BIGINT\n",
      "2    alert_name   VARCHAR(100)\n",
      "3        smarts  VARCHAR(4000)\n",
      "\n",
      "📦 Table: defined_daily_dose\n",
      "          name           type\n",
      "0     atc_code    VARCHAR(10)\n",
      "1    ddd_units   VARCHAR(200)\n",
      "2     ddd_admr  VARCHAR(1000)\n",
      "3  ddd_comment  VARCHAR(2000)\n",
      "4       ddd_id         BIGINT\n",
      "5    ddd_value        NUMERIC\n",
      "\n",
      "📦 Table: product_patents\n",
      "                  name         type\n",
      "0          prod_pat_id       BIGINT\n",
      "1           product_id  VARCHAR(30)\n",
      "2            patent_no  VARCHAR(20)\n",
      "3   patent_expire_date     DATETIME\n",
      "4  drug_substance_flag     SMALLINT\n",
      "5    drug_product_flag     SMALLINT\n",
      "6      patent_use_code  VARCHAR(10)\n",
      "7          delist_flag     SMALLINT\n",
      "8      submission_date     DATETIME\n",
      "\n",
      "📦 Table: protein_class_synonyms\n",
      "                    name           type\n",
      "0        protclasssyn_id         BIGINT\n",
      "1       protein_class_id         BIGINT\n",
      "2  protein_class_synonym  VARCHAR(1000)\n",
      "3               syn_type    VARCHAR(20)\n",
      "\n",
      "📦 Table: research_companies\n",
      "               name          type\n",
      "0        co_stem_id        BIGINT\n",
      "1       res_stem_id        BIGINT\n",
      "2           company  VARCHAR(100)\n",
      "3           country   VARCHAR(50)\n",
      "4  previous_company  VARCHAR(100)\n",
      "\n",
      "📦 Table: assays\n",
      "                          name           type\n",
      "0                     assay_id         BIGINT\n",
      "1                       doc_id         BIGINT\n",
      "2                  description  VARCHAR(4000)\n",
      "3                   assay_type     VARCHAR(1)\n",
      "4              assay_test_type    VARCHAR(20)\n",
      "5               assay_category    VARCHAR(50)\n",
      "6               assay_organism   VARCHAR(250)\n",
      "7                 assay_tax_id         BIGINT\n",
      "8                 assay_strain   VARCHAR(200)\n",
      "9                 assay_tissue   VARCHAR(100)\n",
      "10             assay_cell_type   VARCHAR(100)\n",
      "11  assay_subcellular_fraction   VARCHAR(100)\n",
      "12                         tid         BIGINT\n",
      "13           relationship_type     VARCHAR(1)\n",
      "14            confidence_score       SMALLINT\n",
      "15                  curated_by    VARCHAR(32)\n",
      "16                      src_id        INTEGER\n",
      "17                src_assay_id    VARCHAR(50)\n",
      "18                   chembl_id    VARCHAR(20)\n",
      "19                     cell_id         BIGINT\n",
      "20                  bao_format    VARCHAR(11)\n",
      "21                   tissue_id         BIGINT\n",
      "22                  variant_id         BIGINT\n",
      "23                        aidx   VARCHAR(200)\n",
      "24                 assay_group   VARCHAR(200)\n",
      "\n",
      "📦 Table: compound_records\n",
      "              name           type\n",
      "0        record_id         BIGINT\n",
      "1         molregno         BIGINT\n",
      "2           doc_id         BIGINT\n",
      "3     compound_key   VARCHAR(250)\n",
      "4    compound_name  VARCHAR(4000)\n",
      "5           src_id        INTEGER\n",
      "6  src_compound_id   VARCHAR(150)\n",
      "7             cidx   VARCHAR(200)\n",
      "\n",
      "📦 Table: binding_sites\n",
      "        name          type\n",
      "0    site_id        BIGINT\n",
      "1  site_name  VARCHAR(200)\n",
      "2        tid        BIGINT\n",
      "\n",
      "📦 Table: biotherapeutics\n",
      "            name           type\n",
      "0       molregno         BIGINT\n",
      "1    description  VARCHAR(2000)\n",
      "2  helm_notation  VARCHAR(4000)\n",
      "\n",
      "📦 Table: compound_properties\n",
      "                           name            type\n",
      "0                      molregno          BIGINT\n",
      "1                   mw_freebase   NUMERIC(9, 2)\n",
      "2                         alogp   NUMERIC(9, 2)\n",
      "3                           hba         INTEGER\n",
      "4                           hbd         INTEGER\n",
      "5                           psa   NUMERIC(9, 2)\n",
      "6                           rtb         INTEGER\n",
      "7                      ro3_pass      VARCHAR(3)\n",
      "8            num_ro5_violations        SMALLINT\n",
      "9                  cx_most_apka   NUMERIC(9, 2)\n",
      "10                 cx_most_bpka   NUMERIC(9, 2)\n",
      "11                      cx_logp   NUMERIC(9, 2)\n",
      "12                      cx_logd   NUMERIC(9, 2)\n",
      "13            molecular_species     VARCHAR(50)\n",
      "14                     full_mwt   NUMERIC(9, 2)\n",
      "15               aromatic_rings         INTEGER\n",
      "16                  heavy_atoms         INTEGER\n",
      "17                 qed_weighted   NUMERIC(3, 2)\n",
      "18              mw_monoisotopic  NUMERIC(11, 4)\n",
      "19              full_molformula    VARCHAR(100)\n",
      "20                 hba_lipinski         INTEGER\n",
      "21                 hbd_lipinski         INTEGER\n",
      "22  num_lipinski_ro5_violations        SMALLINT\n",
      "23            np_likeness_score   NUMERIC(3, 2)\n",
      "\n",
      "📦 Table: compound_structural_alerts\n",
      "               name    type\n",
      "0  cpd_str_alert_id  BIGINT\n",
      "1          molregno  BIGINT\n",
      "2          alert_id  BIGINT\n",
      "\n",
      "📦 Table: compound_structures\n",
      "                 name           type\n",
      "0            molregno         BIGINT\n",
      "1             molfile           TEXT\n",
      "2      standard_inchi  VARCHAR(4000)\n",
      "3  standard_inchi_key    VARCHAR(27)\n",
      "4    canonical_smiles  VARCHAR(4000)\n",
      "\n",
      "📦 Table: molecule_atc_classification\n",
      "         name         type\n",
      "0  mol_atc_id       BIGINT\n",
      "1      level5  VARCHAR(10)\n",
      "2    molregno       BIGINT\n",
      "\n",
      "📦 Table: molecule_frac_classification\n",
      "            name    type\n",
      "0    mol_frac_id  BIGINT\n",
      "1  frac_class_id  BIGINT\n",
      "2       molregno  BIGINT\n",
      "\n",
      "📦 Table: molecule_hierarchy\n",
      "              name    type\n",
      "0         molregno  BIGINT\n",
      "1  parent_molregno  BIGINT\n",
      "2  active_molregno  BIGINT\n",
      "\n",
      "📦 Table: molecule_hrac_classification\n",
      "            name    type\n",
      "0    mol_hrac_id  BIGINT\n",
      "1  hrac_class_id  BIGINT\n",
      "2       molregno  BIGINT\n",
      "\n",
      "📦 Table: molecule_irac_classification\n",
      "            name    type\n",
      "0    mol_irac_id  BIGINT\n",
      "1  irac_class_id  BIGINT\n",
      "2       molregno  BIGINT\n",
      "\n",
      "📦 Table: molecule_synonyms\n",
      "          name          type\n",
      "0     molregno        BIGINT\n",
      "1     syn_type   VARCHAR(50)\n",
      "2    molsyn_id        BIGINT\n",
      "3  res_stem_id        BIGINT\n",
      "4     synonyms  VARCHAR(250)\n",
      "\n",
      "📦 Table: target_components\n",
      "           name      type\n",
      "0           tid    BIGINT\n",
      "1  component_id    BIGINT\n",
      "2   targcomp_id    BIGINT\n",
      "3     homologue  SMALLINT\n",
      "\n",
      "📦 Table: target_relations\n",
      "           name         type\n",
      "0           tid       BIGINT\n",
      "1  relationship  VARCHAR(20)\n",
      "2   related_tid       BIGINT\n",
      "3    targrel_id       BIGINT\n",
      "\n",
      "📦 Table: activities\n",
      "                     name           type\n",
      "0             activity_id         BIGINT\n",
      "1                assay_id         BIGINT\n",
      "2                  doc_id         BIGINT\n",
      "3               record_id         BIGINT\n",
      "4                molregno         BIGINT\n",
      "5       standard_relation    VARCHAR(50)\n",
      "6          standard_value        NUMERIC\n",
      "7          standard_units   VARCHAR(100)\n",
      "8           standard_flag       SMALLINT\n",
      "9           standard_type   VARCHAR(250)\n",
      "10       activity_comment  VARCHAR(4000)\n",
      "11  data_validity_comment    VARCHAR(30)\n",
      "12    potential_duplicate       SMALLINT\n",
      "13          pchembl_value  NUMERIC(4, 2)\n",
      "14           bao_endpoint    VARCHAR(11)\n",
      "15               uo_units    VARCHAR(10)\n",
      "16             qudt_units    VARCHAR(70)\n",
      "17                   toid        INTEGER\n",
      "18            upper_value        NUMERIC\n",
      "19   standard_upper_value        NUMERIC\n",
      "20                 src_id        INTEGER\n",
      "21                   type   VARCHAR(250)\n",
      "22               relation    VARCHAR(50)\n",
      "23                  value        NUMERIC\n",
      "24                  units   VARCHAR(100)\n",
      "25             text_value  VARCHAR(1000)\n",
      "26    standard_text_value  VARCHAR(1000)\n",
      "27            action_type    VARCHAR(50)\n",
      "\n",
      "📦 Table: assay_class_map\n",
      "             name    type\n",
      "0  ass_cls_map_id  BIGINT\n",
      "1        assay_id  BIGINT\n",
      "2  assay_class_id  BIGINT\n",
      "\n",
      "📦 Table: assay_parameters\n",
      "                   name           type\n",
      "0        assay_param_id         BIGINT\n",
      "1              assay_id         BIGINT\n",
      "2                  type   VARCHAR(250)\n",
      "3              relation    VARCHAR(50)\n",
      "4                 value        NUMERIC\n",
      "5                 units   VARCHAR(100)\n",
      "6            text_value  VARCHAR(4000)\n",
      "7         standard_type   VARCHAR(250)\n",
      "8     standard_relation    VARCHAR(50)\n",
      "9        standard_value        NUMERIC\n",
      "10       standard_units   VARCHAR(100)\n",
      "11  standard_text_value  VARCHAR(4000)\n",
      "12             comments  VARCHAR(4000)\n",
      "\n",
      "📦 Table: biotherapeutic_components\n",
      "           name    type\n",
      "0    biocomp_id  BIGINT\n",
      "1      molregno  BIGINT\n",
      "2  component_id  BIGINT\n",
      "\n",
      "📦 Table: drug_indication\n",
      "                name           type\n",
      "0         drugind_id         BIGINT\n",
      "1          record_id         BIGINT\n",
      "2           molregno         BIGINT\n",
      "3  max_phase_for_ind  NUMERIC(2, 1)\n",
      "4            mesh_id    VARCHAR(20)\n",
      "5       mesh_heading   VARCHAR(200)\n",
      "6             efo_id    VARCHAR(20)\n",
      "7           efo_term   VARCHAR(200)\n",
      "\n",
      "📦 Table: drug_mechanism\n",
      "                    name           type\n",
      "0                 mec_id         BIGINT\n",
      "1              record_id         BIGINT\n",
      "2               molregno         BIGINT\n",
      "3    mechanism_of_action   VARCHAR(250)\n",
      "4                    tid         BIGINT\n",
      "5                site_id         BIGINT\n",
      "6            action_type    VARCHAR(50)\n",
      "7     direct_interaction       SMALLINT\n",
      "8    molecular_mechanism       SMALLINT\n",
      "9       disease_efficacy       SMALLINT\n",
      "10     mechanism_comment  VARCHAR(2000)\n",
      "11   selectivity_comment  VARCHAR(1000)\n",
      "12  binding_site_comment  VARCHAR(1000)\n",
      "13            variant_id         BIGINT\n",
      "\n",
      "📦 Table: drug_warning\n",
      "                        name           type\n",
      "0                 warning_id         BIGINT\n",
      "1                  record_id         BIGINT\n",
      "2                   molregno         BIGINT\n",
      "3               warning_type    VARCHAR(20)\n",
      "4              warning_class   VARCHAR(100)\n",
      "5        warning_description  VARCHAR(4000)\n",
      "6            warning_country  VARCHAR(1000)\n",
      "7               warning_year        INTEGER\n",
      "8                   efo_term   VARCHAR(200)\n",
      "9                     efo_id    VARCHAR(20)\n",
      "10  efo_id_for_warning_class    VARCHAR(20)\n",
      "\n",
      "📦 Table: formulations\n",
      "             name          type\n",
      "0      product_id   VARCHAR(30)\n",
      "1      ingredient  VARCHAR(200)\n",
      "2        strength  VARCHAR(300)\n",
      "3       record_id        BIGINT\n",
      "4        molregno        BIGINT\n",
      "5  formulation_id        BIGINT\n",
      "\n",
      "📦 Table: metabolism\n",
      "                    name           type\n",
      "0                 met_id         BIGINT\n",
      "1         drug_record_id         BIGINT\n",
      "2    substrate_record_id         BIGINT\n",
      "3   metabolite_record_id         BIGINT\n",
      "4             pathway_id         BIGINT\n",
      "5            pathway_key    VARCHAR(50)\n",
      "6            enzyme_name   VARCHAR(200)\n",
      "7             enzyme_tid         BIGINT\n",
      "8         met_conversion   VARCHAR(200)\n",
      "9               organism   VARCHAR(100)\n",
      "10                tax_id         BIGINT\n",
      "11           met_comment  VARCHAR(1000)\n",
      "\n",
      "📦 Table: site_components\n",
      "            name           type\n",
      "0    sitecomp_id         BIGINT\n",
      "1        site_id         BIGINT\n",
      "2   component_id         BIGINT\n",
      "3      domain_id         BIGINT\n",
      "4  site_residues  VARCHAR(2000)\n",
      "\n",
      "📦 Table: activity_properties\n",
      "                   name           type\n",
      "0                 ap_id         BIGINT\n",
      "1           activity_id         BIGINT\n",
      "2                  type   VARCHAR(250)\n",
      "3              relation    VARCHAR(50)\n",
      "4                 value        NUMERIC\n",
      "5                 units   VARCHAR(100)\n",
      "6            text_value  VARCHAR(2000)\n",
      "7         standard_type   VARCHAR(250)\n",
      "8     standard_relation    VARCHAR(50)\n",
      "9        standard_value        NUMERIC\n",
      "10       standard_units   VARCHAR(100)\n",
      "11  standard_text_value  VARCHAR(2000)\n",
      "12             comments  VARCHAR(2000)\n",
      "13          result_flag       SMALLINT\n",
      "\n",
      "📦 Table: activity_supp_map\n",
      "          name    type\n",
      "0     actsm_id  BIGINT\n",
      "1  activity_id  BIGINT\n",
      "2         smid  BIGINT\n",
      "\n",
      "📦 Table: indication_refs\n",
      "         name           type\n",
      "0   indref_id         BIGINT\n",
      "1  drugind_id         BIGINT\n",
      "2    ref_type    VARCHAR(50)\n",
      "3      ref_id  VARCHAR(4000)\n",
      "4     ref_url  VARCHAR(4000)\n",
      "\n",
      "📦 Table: ligand_eff\n",
      "          name           type\n",
      "0  activity_id         BIGINT\n",
      "1          bei  NUMERIC(9, 2)\n",
      "2          sei  NUMERIC(9, 2)\n",
      "3           le  NUMERIC(9, 2)\n",
      "4          lle  NUMERIC(9, 2)\n",
      "\n",
      "📦 Table: mechanism_refs\n",
      "        name          type\n",
      "0  mecref_id        BIGINT\n",
      "1     mec_id        BIGINT\n",
      "2   ref_type   VARCHAR(50)\n",
      "3     ref_id  VARCHAR(200)\n",
      "4    ref_url  VARCHAR(400)\n",
      "\n",
      "📦 Table: metabolism_refs\n",
      "        name          type\n",
      "0  metref_id        BIGINT\n",
      "1     met_id        BIGINT\n",
      "2   ref_type   VARCHAR(50)\n",
      "3     ref_id  VARCHAR(200)\n",
      "4    ref_url  VARCHAR(400)\n",
      "\n",
      "📦 Table: predicted_binding_domains\n",
      "                name         type\n",
      "0        predbind_id       BIGINT\n",
      "1        activity_id       BIGINT\n",
      "2            site_id       BIGINT\n",
      "3  prediction_method  VARCHAR(50)\n",
      "4         confidence  VARCHAR(10)\n",
      "\n",
      "📦 Table: warning_refs\n",
      "         name           type\n",
      "0  warnref_id         BIGINT\n",
      "1  warning_id         BIGINT\n",
      "2    ref_type    VARCHAR(50)\n",
      "3      ref_id  VARCHAR(4000)\n",
      "4     ref_url  VARCHAR(4000)\n",
      "\n",
      "📦 Table: sqlite_stat1\n",
      "   name type\n",
      "0   tbl     \n",
      "1   idx     \n",
      "2  stat     \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def get_table_columns(connection):\n",
    "    tables_query = \"SELECT name FROM sqlite_master WHERE type='table';\"\n",
    "    tables = pd.read_sql_query(tables_query, connection)['name'].tolist()\n",
    "    \n",
    "    column_dict = {}\n",
    "    for table in tables:\n",
    "        try:\n",
    "            cols = pd.read_sql_query(f\"PRAGMA table_info({table});\", connection)\n",
    "            column_dict[table] = cols[['name', 'type']]\n",
    "        except:\n",
    "            column_dict[table] = 'Error reading columns'\n",
    "    return column_dict\n",
    "\n",
    "all_columns = get_table_columns(conn)\n",
    "\n",
    "# Display results for verification\n",
    "for table, cols in all_columns.items():\n",
    "    print(f\"\\n📦 Table: {table}\")\n",
    "    print(cols)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e035c50-7f8d-4f78-bc56-e9278b321b68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting tdqm\n",
      "  Downloading tdqm-0.0.1.tar.gz (1.4 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Requirement already satisfied: tqdm in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from tdqm) (4.66.5)\n",
      "Requirement already satisfied: colorama in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from tqdm->tdqm) (0.4.6)\n",
      "Building wheels for collected packages: tdqm\n",
      "  Building wheel for tdqm (setup.py): started\n",
      "  Building wheel for tdqm (setup.py): finished with status 'done'\n",
      "  Created wheel for tdqm: filename=tdqm-0.0.1-py3-none-any.whl size=1395 sha256=5191731ad923de20eb37a12f5f25c2b923479d11aab233cd2e1168dc996bd3bc\n",
      "  Stored in directory: c:\\users\\konde\\appdata\\local\\pip\\cache\\wheels\\af\\02\\71\\aae0f7ee738abf19498353918ddae0f90a0d6ceb337b0bbc91\n",
      "Successfully built tdqm\n",
      "Installing collected packages: tdqm\n",
      "Successfully installed tdqm-0.0.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  DEPRECATION: Building 'tdqm' using the legacy setup.py bdist_wheel mechanism, which will be removed in a future version. pip 25.3 will enforce this behaviour change. A possible replacement is to use the standardized build interface by setting the `--use-pep517` option, (possibly combined with `--no-build-isolation`), or adding a `pyproject.toml` file to the source tree of 'tdqm'. Discussion can be found at https://github.com/pypa/pip/issues/6334\n"
     ]
    }
   ],
   "source": [
    "!pip install tdqm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "253e08b1-c7cc-4999-a953-afabfa5cf0ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🔌 Connecting to database...\n",
      "🔍 Step 1: Running TEST query with LIMIT 50000...\n",
      "✅ Query finished in 2.01 seconds\n",
      "🔢 Rows fetched: 50,000\n",
      "🧹 Step 2: Cleaning...\n",
      "✅ Cleaned in 0.09 seconds\n",
      "💾 Step 3: Saving to disk...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving CSV: 100%|█████████████████████████████████████████████████████████████| 1871/1871 [00:00<00:00, 42983.98rows/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved CSV + Parquet in 0.91 seconds\n",
      "🏁 ALL DONE in 3.01 seconds\n"
     ]
    }
   ],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import time\n",
    "from tqdm import tqdm\n",
    "\n",
    "start_total = time.time()\n",
    "\n",
    "print(\"🔌 Connecting to database...\")\n",
    "conn = sqlite3.connect(\"chembl_35/chembl_35_sqlite/chembl_35.db\")\n",
    "\n",
    "print(\"🔍 Step 1: Running TEST query with LIMIT 50000...\")\n",
    "start = time.time()\n",
    "query = \"\"\"\n",
    "SELECT \n",
    "    md.chembl_id AS drug_id,\n",
    "    cs.canonical_smiles AS SMILES,\n",
    "    cp.cx_logd AS logD,\n",
    "    cp.qed_weighted AS drug_likeness,\n",
    "    cp.psa AS psa,\n",
    "    cp.cx_logp AS logP,\n",
    "    act.standard_value AS IC50,\n",
    "    td.pref_name AS target,\n",
    "    di.efo_term AS indication,\n",
    "    sa.alert_name AS toxicity_alert,\n",
    "    dw.warning_description AS warning,\n",
    "    met.enzyme_name AS metabolism_enzyme,\n",
    "    met.met_conversion AS metabolism_conversion\n",
    "\n",
    "FROM molecule_dictionary md\n",
    "JOIN compound_structures cs ON md.molregno = cs.molregno\n",
    "JOIN compound_properties cp ON md.molregno = cp.molregno\n",
    "JOIN activities act ON md.molregno = act.molregno AND act.standard_type = 'IC50' AND act.standard_value IS NOT NULL\n",
    "JOIN assays a ON act.assay_id = a.assay_id AND a.confidence_score >= 8\n",
    "JOIN target_dictionary td ON a.tid = td.tid\n",
    "LEFT JOIN drug_indication di ON md.molregno = di.molregno\n",
    "LEFT JOIN compound_structural_alerts csa ON md.molregno = csa.molregno\n",
    "LEFT JOIN structural_alerts sa ON csa.alert_id = sa.alert_id\n",
    "LEFT JOIN drug_warning dw ON md.molregno = dw.molregno\n",
    "LEFT JOIN compound_records cr ON md.molregno = cr.molregno\n",
    "LEFT JOIN metabolism met ON cr.record_id = met.substrate_record_id\n",
    "\n",
    "LIMIT 50000\n",
    "\"\"\"\n",
    "\n",
    "df = pd.read_sql_query(query, conn)\n",
    "conn.close()\n",
    "print(f\"✅ Query finished in {round(time.time() - start, 2)} seconds\")\n",
    "print(f\"🔢 Rows fetched: {len(df):,}\")\n",
    "\n",
    "# Step 2: Cleaning\n",
    "print(\"🧹 Step 2: Cleaning...\")\n",
    "start = time.time()\n",
    "df.dropna(subset=[\"SMILES\", \"IC50\", \"target\"], inplace=True)\n",
    "df.drop_duplicates(inplace=True)\n",
    "df[\"IC50\"] = pd.to_numeric(df[\"IC50\"], errors=\"coerce\")\n",
    "print(f\"✅ Cleaned in {round(time.time() - start, 2)} seconds\")\n",
    "\n",
    "# Step 3: Save to CSV + Parquet\n",
    "print(\"💾 Step 3: Saving to disk...\")\n",
    "start = time.time()\n",
    "\n",
    "# CSV with tqdm\n",
    "with tqdm(total=len(df), desc=\"Saving CSV\", unit=\"rows\") as pbar:\n",
    "    df.to_csv(\"test_clean_chembl.csv\", index=False)\n",
    "    pbar.update(len(df))\n",
    "\n",
    "# Parquet (fast!)\n",
    "df.to_parquet(\"test_clean_chembl.parquet\", index=False)\n",
    "print(f\"✅ Saved CSV + Parquet in {round(time.time() - start, 2)} seconds\")\n",
    "\n",
    "print(f\"🏁 ALL DONE in {round(time.time() - start_total, 2)} seconds\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "92a5f27c-435e-47a9-a32d-aa07ddbe85aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🔹 Preview:\n",
      "        drug_id                                             SMILES  logD  \\\n",
      "0  CHEMBL113081           c1ccc(-c2nc3c(-c4nc5ccccc5o4)cccc3o2)cc1  4.72   \n",
      "1  CHEMBL324340  Cc1ccc2oc(-c3cccc(N4C(=O)c5ccc(C(=O)O)cc5C4=O)...  0.74   \n",
      "2  CHEMBL109600  COc1ccccc1-c1ccc2oc(-c3ccc(OC)c(N4C(=O)c5ccc(C...  1.56   \n",
      "3  CHEMBL357278  Cc1nc2cc(OC[C@H](O)CN3CCN(CC(=O)Nc4ccc(Cl)c(C(...  3.62   \n",
      "4  CHEMBL357119  Cc1nc2cc(OC[C@H](O)CN3CCN(CC(=O)NCCc4ccccc4)CC...  2.12   \n",
      "\n",
      "   drug_likeness     psa  logP      IC50                      target  \\\n",
      "0           0.44   52.06  4.72  100000.0  DNA topoisomerase II alpha   \n",
      "1           0.52  100.71  4.12    2500.0                  Heparanase   \n",
      "2           0.28  119.17  4.94    9000.0                  Heparanase   \n",
      "3           0.44   77.93  3.71    4000.0       Palmitoyl-CoA oxidase   \n",
      "4           0.48   77.93  2.22   17000.0       Palmitoyl-CoA oxidase   \n",
      "\n",
      "  indication toxicity_alert  warning metabolism_enzyme metabolism_conversion  \n",
      "0        NaN            NaN      NaN               NaN                   NaN  \n",
      "1        NaN    phthalimide      NaN               NaN                   NaN  \n",
      "2        NaN    phthalimide      NaN               NaN                   NaN  \n",
      "3        NaN            NaN      NaN               NaN                   NaN  \n",
      "4        NaN            NaN      NaN               NaN                   NaN  \n",
      "\n",
      "🔹 Data Info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1871 entries, 0 to 1870\n",
      "Data columns (total 13 columns):\n",
      " #   Column                 Non-Null Count  Dtype  \n",
      "---  ------                 --------------  -----  \n",
      " 0   drug_id                1871 non-null   object \n",
      " 1   SMILES                 1871 non-null   object \n",
      " 2   logD                   1864 non-null   float64\n",
      " 3   drug_likeness          1864 non-null   float64\n",
      " 4   psa                    1864 non-null   float64\n",
      " 5   logP                   1864 non-null   float64\n",
      " 6   IC50                   1871 non-null   float64\n",
      " 7   target                 1871 non-null   object \n",
      " 8   indication             186 non-null    object \n",
      " 9   toxicity_alert         1558 non-null   object \n",
      " 10  warning                0 non-null      float64\n",
      " 11  metabolism_enzyme      1 non-null      object \n",
      " 12  metabolism_conversion  75 non-null     object \n",
      "dtypes: float64(6), object(7)\n",
      "memory usage: 190.2+ KB\n",
      "None\n",
      "\n",
      "🔹 Null Values:\n",
      "drug_id                     0\n",
      "SMILES                      0\n",
      "logD                        7\n",
      "drug_likeness               7\n",
      "psa                         7\n",
      "logP                        7\n",
      "IC50                        0\n",
      "target                      0\n",
      "indication               1685\n",
      "toxicity_alert            313\n",
      "warning                  1871\n",
      "metabolism_enzyme        1870\n",
      "metabolism_conversion    1796\n",
      "dtype: int64\n",
      "\n",
      "🔹 Unique targets: 147\n",
      "🔹 Unique indications: 52\n",
      "\n",
      "🔹 IC50 (nM) stats:\n",
      "count    1.871000e+03\n",
      "mean     8.184043e+04\n",
      "std      4.477809e+05\n",
      "min      1.000000e-02\n",
      "25%      1.700000e+01\n",
      "50%      4.160000e+02\n",
      "75%      9.100000e+03\n",
      "max      1.200000e+07\n",
      "Name: IC50, dtype: float64\n",
      "\n",
      "🔹 Top 10 Targets:\n",
      "target\n",
      "Epidermal growth factor receptor erbB1          109\n",
      "Cyclooxygenase-2                                109\n",
      "Phospholipase A2 group 1B                        67\n",
      "Matrix metalloproteinase 3                       57\n",
      "Matrix metalloproteinase-2                       48\n",
      "Renin                                            46\n",
      "Matrix metalloproteinase-1                       42\n",
      "Matrix metalloproteinase 13                      42\n",
      "ADAM17                                           40\n",
      "Human immunodeficiency virus type 1 protease     40\n",
      "Name: count, dtype: int64\n",
      "\n",
      "🔍 Drugs targeting EGFR:\n",
      "Empty DataFrame\n",
      "Columns: [drug_id, SMILES, IC50, indication]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "# Load dataset (choose either CSV or Parquet)\n",
    "df = pd.read_csv(\"test_clean_chembl.csv\")  # OR\n",
    "# df = pd.read_parquet(\"test_clean_chembl.parquet\")\n",
    "\n",
    "# 🔍 Preview top rows\n",
    "print(\"🔹 Preview:\")\n",
    "print(df.head())\n",
    "\n",
    "# 📊 Summary info\n",
    "print(\"\\n🔹 Data Info:\")\n",
    "print(df.info())\n",
    "\n",
    "# 🧼 Missing values check\n",
    "print(\"\\n🔹 Null Values:\")\n",
    "print(df.isnull().sum())\n",
    "\n",
    "# 🔢 Unique targets and indications\n",
    "print(\"\\n🔹 Unique targets:\", df['target'].nunique())\n",
    "print(\"🔹 Unique indications:\", df['indication'].nunique())\n",
    "\n",
    "# 📈 IC50 distribution\n",
    "print(\"\\n🔹 IC50 (nM) stats:\")\n",
    "print(df['IC50'].describe())\n",
    "\n",
    "# 📊 Most common targets\n",
    "print(\"\\n🔹 Top 10 Targets:\")\n",
    "print(df['target'].value_counts().head(10))\n",
    "\n",
    "# 🧪 Optional: Convert IC50 to pIC50 (–log10(IC50 in molar units))\n",
    "df[\"pIC50\"] = -df[\"IC50\"].astype(float).div(1e9).apply(lambda x: np.log10(x) if x > 0 else None)\n",
    "\n",
    "# 🎯 Filter: Only rows targeting EGFR\n",
    "egfr_df = df[df[\"target\"].str.contains(\"EGFR\", na=False)]\n",
    "print(\"\\n🔍 Drugs targeting EGFR:\")\n",
    "print(egfr_df[[\"drug_id\", \"SMILES\", \"IC50\", \"indication\"]].head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b4b9ae0b-39e2-4716-9b7f-3fdf8733e8d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "egfr_df = df[df[\"target\"].str.contains(\"epidermal growth factor\", case=False, na=False)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0d918542-ca06-4db6-b76c-d3e8b2352b8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Rows with all critical data: 186\n"
     ]
    }
   ],
   "source": [
    "# Keep rows with non-null SMILES, IC50, and indication\n",
    "filtered_df = df.dropna(subset=[\"SMILES\", \"IC50\", \"indication\"])\n",
    "print(f\"✅ Rows with all critical data: {len(filtered_df)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "724999c0-0ee1-4315-96cd-2728e24e24ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "df[\"pIC50\"] = -df[\"IC50\"].div(1e9).apply(lambda x: np.log10(x) if x > 0 else None)\n",
    "strong_binders = df[df[\"pIC50\"] > 7]\n",
    "strong_binders.to_csv(\"strong_binders.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5968f3b6-c077-4d29-96c3-43cdb16b45b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug_id</th>\n",
       "      <th>SMILES</th>\n",
       "      <th>logD</th>\n",
       "      <th>drug_likeness</th>\n",
       "      <th>psa</th>\n",
       "      <th>logP</th>\n",
       "      <th>IC50</th>\n",
       "      <th>target</th>\n",
       "      <th>indication</th>\n",
       "      <th>toxicity_alert</th>\n",
       "      <th>warning</th>\n",
       "      <th>metabolism_enzyme</th>\n",
       "      <th>metabolism_conversion</th>\n",
       "      <th>pIC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL305153</td>\n",
       "      <td>CC(C)(C)NC[C@H](O)CON=C1c2ccccc2-c2ccccc21</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.71</td>\n",
       "      <td>53.85</td>\n",
       "      <td>3.55</td>\n",
       "      <td>29.4</td>\n",
       "      <td>Beta-1 adrenergic receptor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>imine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.531653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL305153</td>\n",
       "      <td>CC(C)(C)NC[C@H](O)CON=C1c2ccccc2-c2ccccc21</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.71</td>\n",
       "      <td>53.85</td>\n",
       "      <td>3.55</td>\n",
       "      <td>29.4</td>\n",
       "      <td>Beta-1 adrenergic receptor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxygen-nitrogen single bond</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.531653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL305153</td>\n",
       "      <td>CC(C)(C)NC[C@H](O)CON=C1c2ccccc2-c2ccccc21</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.71</td>\n",
       "      <td>53.85</td>\n",
       "      <td>3.55</td>\n",
       "      <td>29.4</td>\n",
       "      <td>Beta-1 adrenergic receptor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long aliphatic chain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.531653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL305153</td>\n",
       "      <td>CC(C)(C)NC[C@H](O)CON=C1c2ccccc2-c2ccccc21</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.71</td>\n",
       "      <td>53.85</td>\n",
       "      <td>3.55</td>\n",
       "      <td>30.8</td>\n",
       "      <td>Beta-2 adrenergic receptor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>imine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.511449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL305153</td>\n",
       "      <td>CC(C)(C)NC[C@H](O)CON=C1c2ccccc2-c2ccccc21</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.71</td>\n",
       "      <td>53.85</td>\n",
       "      <td>3.55</td>\n",
       "      <td>30.8</td>\n",
       "      <td>Beta-2 adrenergic receptor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oxygen-nitrogen single bond</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.511449</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        drug_id                                      SMILES  logD  \\\n",
       "0  CHEMBL305153  CC(C)(C)NC[C@H](O)CON=C1c2ccccc2-c2ccccc21  1.08   \n",
       "1  CHEMBL305153  CC(C)(C)NC[C@H](O)CON=C1c2ccccc2-c2ccccc21  1.08   \n",
       "2  CHEMBL305153  CC(C)(C)NC[C@H](O)CON=C1c2ccccc2-c2ccccc21  1.08   \n",
       "3  CHEMBL305153  CC(C)(C)NC[C@H](O)CON=C1c2ccccc2-c2ccccc21  1.08   \n",
       "4  CHEMBL305153  CC(C)(C)NC[C@H](O)CON=C1c2ccccc2-c2ccccc21  1.08   \n",
       "\n",
       "   drug_likeness    psa  logP  IC50                      target indication  \\\n",
       "0           0.71  53.85  3.55  29.4  Beta-1 adrenergic receptor        NaN   \n",
       "1           0.71  53.85  3.55  29.4  Beta-1 adrenergic receptor        NaN   \n",
       "2           0.71  53.85  3.55  29.4  Beta-1 adrenergic receptor        NaN   \n",
       "3           0.71  53.85  3.55  30.8  Beta-2 adrenergic receptor        NaN   \n",
       "4           0.71  53.85  3.55  30.8  Beta-2 adrenergic receptor        NaN   \n",
       "\n",
       "                toxicity_alert  warning  metabolism_enzyme  \\\n",
       "0                        imine      NaN                NaN   \n",
       "1  Oxygen-nitrogen single bond      NaN                NaN   \n",
       "2         Long aliphatic chain      NaN                NaN   \n",
       "3                        imine      NaN                NaN   \n",
       "4  Oxygen-nitrogen single bond      NaN                NaN   \n",
       "\n",
       "  metabolism_conversion     pIC50  \n",
       "0                   NaN  7.531653  \n",
       "1                   NaN  7.531653  \n",
       "2                   NaN  7.531653  \n",
       "3                   NaN  7.511449  \n",
       "4                   NaN  7.511449  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.read_csv(\"strong_binders.csv\")\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0ff282d3-610d-4bb8-9893-7661ba91b6ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug_id</th>\n",
       "      <th>drug_name</th>\n",
       "      <th>SMILES</th>\n",
       "      <th>logD</th>\n",
       "      <th>logP</th>\n",
       "      <th>psa</th>\n",
       "      <th>drug_likeness</th>\n",
       "      <th>max_phase</th>\n",
       "      <th>IC50</th>\n",
       "      <th>target</th>\n",
       "      <th>organism</th>\n",
       "      <th>target_type</th>\n",
       "      <th>mechanism_of_action</th>\n",
       "      <th>efo_term</th>\n",
       "      <th>efo_id</th>\n",
       "      <th>mesh_heading</th>\n",
       "      <th>mesh_id</th>\n",
       "      <th>toxicity_alert</th>\n",
       "      <th>pIC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL113081</td>\n",
       "      <td>None</td>\n",
       "      <td>c1ccc(-c2nc3c(-c4nc5ccccc5o4)cccc3o2)cc1</td>\n",
       "      <td>4.72</td>\n",
       "      <td>4.72</td>\n",
       "      <td>52.06</td>\n",
       "      <td>0.44</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>DNA topoisomerase II alpha</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL324340</td>\n",
       "      <td>None</td>\n",
       "      <td>Cc1ccc2oc(-c3cccc(N4C(=O)c5ccc(C(=O)O)cc5C4=O)...</td>\n",
       "      <td>0.74</td>\n",
       "      <td>4.12</td>\n",
       "      <td>100.71</td>\n",
       "      <td>0.52</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2500.0</td>\n",
       "      <td>Heparanase</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>phthalimide</td>\n",
       "      <td>5.602060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL109600</td>\n",
       "      <td>None</td>\n",
       "      <td>COc1ccccc1-c1ccc2oc(-c3ccc(OC)c(N4C(=O)c5ccc(C...</td>\n",
       "      <td>1.56</td>\n",
       "      <td>4.94</td>\n",
       "      <td>119.17</td>\n",
       "      <td>0.28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9000.0</td>\n",
       "      <td>Heparanase</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>phthalimide</td>\n",
       "      <td>5.045757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL357278</td>\n",
       "      <td>None</td>\n",
       "      <td>Cc1nc2cc(OC[C@H](O)CN3CCN(CC(=O)Nc4ccc(Cl)c(C(...</td>\n",
       "      <td>3.62</td>\n",
       "      <td>3.71</td>\n",
       "      <td>77.93</td>\n",
       "      <td>0.44</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>Palmitoyl-CoA oxidase</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5.397940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL357119</td>\n",
       "      <td>None</td>\n",
       "      <td>Cc1nc2cc(OC[C@H](O)CN3CCN(CC(=O)NCCc4ccccc4)CC...</td>\n",
       "      <td>2.12</td>\n",
       "      <td>2.22</td>\n",
       "      <td>77.93</td>\n",
       "      <td>0.48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17000.0</td>\n",
       "      <td>Palmitoyl-CoA oxidase</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4.769551</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        drug_id drug_name                                             SMILES  \\\n",
       "0  CHEMBL113081      None           c1ccc(-c2nc3c(-c4nc5ccccc5o4)cccc3o2)cc1   \n",
       "1  CHEMBL324340      None  Cc1ccc2oc(-c3cccc(N4C(=O)c5ccc(C(=O)O)cc5C4=O)...   \n",
       "2  CHEMBL109600      None  COc1ccccc1-c1ccc2oc(-c3ccc(OC)c(N4C(=O)c5ccc(C...   \n",
       "3  CHEMBL357278      None  Cc1nc2cc(OC[C@H](O)CN3CCN(CC(=O)Nc4ccc(Cl)c(C(...   \n",
       "4  CHEMBL357119      None  Cc1nc2cc(OC[C@H](O)CN3CCN(CC(=O)NCCc4ccccc4)CC...   \n",
       "\n",
       "   logD  logP     psa  drug_likeness  max_phase      IC50  \\\n",
       "0  4.72  4.72   52.06           0.44        NaN  100000.0   \n",
       "1  0.74  4.12  100.71           0.52        NaN    2500.0   \n",
       "2  1.56  4.94  119.17           0.28        NaN    9000.0   \n",
       "3  3.62  3.71   77.93           0.44        NaN    4000.0   \n",
       "4  2.12  2.22   77.93           0.48        NaN   17000.0   \n",
       "\n",
       "                       target           organism     target_type  \\\n",
       "0  DNA topoisomerase II alpha       Homo sapiens  SINGLE PROTEIN   \n",
       "1                  Heparanase       Homo sapiens  SINGLE PROTEIN   \n",
       "2                  Heparanase       Homo sapiens  SINGLE PROTEIN   \n",
       "3       Palmitoyl-CoA oxidase  Rattus norvegicus  SINGLE PROTEIN   \n",
       "4       Palmitoyl-CoA oxidase  Rattus norvegicus  SINGLE PROTEIN   \n",
       "\n",
       "  mechanism_of_action efo_term efo_id mesh_heading mesh_id toxicity_alert  \\\n",
       "0                None     None   None         None    None           None   \n",
       "1                None     None   None         None    None    phthalimide   \n",
       "2                None     None   None         None    None    phthalimide   \n",
       "3                None     None   None         None    None           None   \n",
       "4                None     None   None         None    None           None   \n",
       "\n",
       "      pIC50  \n",
       "0  4.000000  \n",
       "1  5.602060  \n",
       "2  5.045757  \n",
       "3  5.397940  \n",
       "4  4.769551  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import sqlite3\n",
    "\n",
    "# Connect to your ChEMBL SQLite database\n",
    "conn = sqlite3.connect(\"chembl_35/chembl_35_sqlite/chembl_35.db\")\n",
    "\n",
    "# Define the solid SQL query\n",
    "query = \"\"\"\n",
    "SELECT DISTINCT\n",
    "    md.chembl_id                    AS drug_id,\n",
    "    md.pref_name                    AS drug_name,\n",
    "    cs.canonical_smiles             AS SMILES,\n",
    "    cp.cx_logd                      AS logD,\n",
    "    cp.cx_logp                      AS logP,\n",
    "    cp.psa                          AS psa,\n",
    "    cp.qed_weighted                 AS drug_likeness,\n",
    "    md.max_phase                    AS max_phase,\n",
    "    act.standard_value              AS IC50,\n",
    "    td.pref_name                    AS target,\n",
    "    td.organism                     AS organism,\n",
    "    td.target_type                  AS target_type,\n",
    "    dm.mechanism_of_action          AS mechanism_of_action,\n",
    "    di.efo_term                     AS efo_term,\n",
    "    di.efo_id                       AS efo_id,\n",
    "    di.mesh_heading                 AS mesh_heading,\n",
    "    di.mesh_id                      AS mesh_id,\n",
    "    sa.alert_name                   AS toxicity_alert\n",
    "\n",
    "FROM molecule_dictionary md\n",
    "INNER JOIN compound_structures cs ON md.molregno = cs.molregno\n",
    "INNER JOIN compound_properties cp ON md.molregno = cp.molregno\n",
    "INNER JOIN activities act ON md.molregno = act.molregno\n",
    "    AND act.standard_type = 'IC50'\n",
    "    AND act.standard_value IS NOT NULL\n",
    "INNER JOIN assays a ON act.assay_id = a.assay_id\n",
    "    AND a.confidence_score >= 8\n",
    "INNER JOIN target_dictionary td ON a.tid = td.tid\n",
    "LEFT JOIN drug_mechanism dm ON md.molregno = dm.molregno AND dm.tid = td.tid\n",
    "LEFT JOIN drug_indication di ON md.molregno = di.molregno\n",
    "LEFT JOIN compound_structural_alerts csa ON md.molregno = csa.molregno\n",
    "LEFT JOIN structural_alerts sa ON csa.alert_id = sa.alert_id\n",
    "LIMIT 50000;\n",
    "\"\"\"\n",
    "\n",
    "# Run the query and load into DataFrame\n",
    "df = pd.read_sql_query(query, conn)\n",
    "\n",
    "# Clean: drop any rows with NULL SMILES or IC50 just in case\n",
    "df.dropna(subset=[\"SMILES\", \"IC50\"], inplace=True)\n",
    "\n",
    "# Convert IC50 to pIC50 (optional)\n",
    "import numpy as np\n",
    "df[\"pIC50\"] = -df[\"IC50\"].astype(float).div(1e9).apply(lambda x: np.log10(x) if x > 0 else None)\n",
    "\n",
    "# Preview\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "de5b4ce2-97fa-46e6-9125-d057ac149319",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3761"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[df[\"max_phase\"] >= 2]             # Keep only clinical-stage drugs\n",
    "df = df[df[\"mechanism_of_action\"].notna()]  # Only drugs with known MoA\n",
    "df = df[df[\"efo_term\"].notna()]             # Keep drugs with disease linkage\n",
    "df.head()\n",
    "df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6ff2c190-b0bd-4235-882d-f615d47514cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🔄 Loading data...\n",
      "✅ Final dataset shape: (553992, 19)\n",
      "🧪 Columns: ['drug_id', 'drug_name', 'SMILES', 'logD', 'logP', 'psa', 'drug_likeness', 'max_phase', 'IC50', 'target', 'organism', 'target_type', 'mechanism_of_action', 'efo_term', 'efo_id', 'mesh_heading', 'mesh_id', 'toxicity_alert', 'pIC50']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import sqlite3\n",
    "import numpy as np\n",
    "\n",
    "# Connect to your ChEMBL SQLite database\n",
    "conn = sqlite3.connect(\"chembl_35/chembl_35_sqlite/chembl_35.db\")\n",
    "\n",
    "# Optimized SQL query (no LIMIT)\n",
    "query = \"\"\"\n",
    "SELECT DISTINCT\n",
    "    md.chembl_id                    AS drug_id,\n",
    "    md.pref_name                    AS drug_name,\n",
    "    cs.canonical_smiles             AS SMILES,\n",
    "    cp.cx_logd                      AS logD,\n",
    "    cp.cx_logp                      AS logP,\n",
    "    cp.psa                          AS psa,\n",
    "    cp.qed_weighted                 AS drug_likeness,\n",
    "    md.max_phase                    AS max_phase,\n",
    "    act.standard_value              AS IC50,\n",
    "    td.pref_name                    AS target,\n",
    "    td.organism                     AS organism,\n",
    "    td.target_type                  AS target_type,\n",
    "    dm.mechanism_of_action          AS mechanism_of_action,\n",
    "    di.efo_term                     AS efo_term,\n",
    "    di.efo_id                       AS efo_id,\n",
    "    di.mesh_heading                 AS mesh_heading,\n",
    "    di.mesh_id                      AS mesh_id,\n",
    "    sa.alert_name                   AS toxicity_alert\n",
    "FROM molecule_dictionary md\n",
    "INNER JOIN compound_structures cs ON md.molregno = cs.molregno\n",
    "INNER JOIN compound_properties cp ON md.molregno = cp.molregno\n",
    "INNER JOIN activities act ON md.molregno = act.molregno\n",
    "    AND act.standard_type = 'IC50'\n",
    "    AND act.standard_value IS NOT NULL\n",
    "INNER JOIN assays a ON act.assay_id = a.assay_id\n",
    "    AND a.confidence_score >= 8\n",
    "INNER JOIN target_dictionary td ON a.tid = td.tid\n",
    "LEFT JOIN drug_mechanism dm ON md.molregno = dm.molregno AND dm.tid = td.tid\n",
    "LEFT JOIN drug_indication di ON md.molregno = di.molregno\n",
    "LEFT JOIN compound_structural_alerts csa ON md.molregno = csa.molregno\n",
    "LEFT JOIN structural_alerts sa ON csa.alert_id = sa.alert_id\n",
    "\"\"\"\n",
    "\n",
    "# Run the query\n",
    "print(\"🔄 Loading data...\")\n",
    "df = pd.read_sql_query(query, conn)\n",
    "\n",
    "# Drop rows missing SMILES or IC50\n",
    "df.dropna(subset=[\"SMILES\", \"IC50\"], inplace=True)\n",
    "\n",
    "# Filter to clinically relevant + annotated data\n",
    "df = df[df[\"max_phase\"] >= 2]\n",
    "df = df[df[\"mechanism_of_action\"].notna()]\n",
    "df = df[df[\"efo_term\"].notna()]\n",
    "\n",
    "# Drop duplicates\n",
    "df.drop_duplicates(inplace=True)\n",
    "\n",
    "# Compute pIC50\n",
    "df[\"pIC50\"] = -df[\"IC50\"].astype(float).div(1e9).apply(lambda x: np.log10(x) if x > 0 else None)\n",
    "\n",
    "# Save to CSV (added line)\n",
    "df.to_csv(\"cleaned_clinical_drugs_dataset.csv\", index=False)\n",
    "\n",
    "# Summary\n",
    "print(\"✅ Final dataset shape:\", df.shape)\n",
    "print(\"🧪 Columns:\", list(df.columns))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f919505a-be30-49fe-9f2a-8c9e2371d4b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug_id</th>\n",
       "      <th>drug_name</th>\n",
       "      <th>SMILES</th>\n",
       "      <th>logD</th>\n",
       "      <th>logP</th>\n",
       "      <th>psa</th>\n",
       "      <th>drug_likeness</th>\n",
       "      <th>max_phase</th>\n",
       "      <th>IC50</th>\n",
       "      <th>target</th>\n",
       "      <th>organism</th>\n",
       "      <th>target_type</th>\n",
       "      <th>mechanism_of_action</th>\n",
       "      <th>efo_term</th>\n",
       "      <th>efo_id</th>\n",
       "      <th>mesh_heading</th>\n",
       "      <th>mesh_id</th>\n",
       "      <th>toxicity_alert</th>\n",
       "      <th>pIC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>AIDS</td>\n",
       "      <td>EFO:0000765</td>\n",
       "      <td>Acquired Immunodeficiency Syndrome</td>\n",
       "      <td>D000163</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>immunodeficiency disease</td>\n",
       "      <td>MONDO:0021094</td>\n",
       "      <td>Autoimmune Diseases</td>\n",
       "      <td>D001327</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>type 2 diabetes mellitus</td>\n",
       "      <td>MONDO:0005148</td>\n",
       "      <td>Diabetes Mellitus, Type 2</td>\n",
       "      <td>D003924</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>hepatitis C virus infection</td>\n",
       "      <td>EFO:0003047</td>\n",
       "      <td>Hepatitis C</td>\n",
       "      <td>D006526</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>infection</td>\n",
       "      <td>EFO:0000544</td>\n",
       "      <td>Infections</td>\n",
       "      <td>D007239</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>non-Hodgkins lymphoma</td>\n",
       "      <td>EFO:0005952</td>\n",
       "      <td>Lymphoma, Non-Hodgkin</td>\n",
       "      <td>D008228</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>malaria</td>\n",
       "      <td>EFO:0001068</td>\n",
       "      <td>Malaria</td>\n",
       "      <td>D008288</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>neoplasm</td>\n",
       "      <td>EFO:0000616</td>\n",
       "      <td>Neoplasms</td>\n",
       "      <td>D009369</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>pancreatic carcinoma</td>\n",
       "      <td>EFO:0002618</td>\n",
       "      <td>Pancreatic Neoplasms</td>\n",
       "      <td>D010190</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>tuberculosis</td>\n",
       "      <td>MONDO:0018076</td>\n",
       "      <td>Tuberculosis</td>\n",
       "      <td>D014376</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>viral disease</td>\n",
       "      <td>EFO:0000763</td>\n",
       "      <td>Virus Diseases</td>\n",
       "      <td>D014777</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV-1 infection</td>\n",
       "      <td>EFO:0000180</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV infection</td>\n",
       "      <td>EFO:0000764</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>MONDO:0008315</td>\n",
       "      <td>Prostatic Neoplasms, Castration-Resistant</td>\n",
       "      <td>D064129</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>triple-negative breast cancer</td>\n",
       "      <td>EFO:0005537</td>\n",
       "      <td>Triple Negative Breast Neoplasms</td>\n",
       "      <td>D064726</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>AIDS</td>\n",
       "      <td>EFO:0000765</td>\n",
       "      <td>Acquired Immunodeficiency Syndrome</td>\n",
       "      <td>D000163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>infection</td>\n",
       "      <td>EFO:0000544</td>\n",
       "      <td>Infections</td>\n",
       "      <td>D007239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>tuberculosis</td>\n",
       "      <td>MONDO:0018076</td>\n",
       "      <td>Tuberculosis</td>\n",
       "      <td>D014376</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>viral disease</td>\n",
       "      <td>EFO:0000763</td>\n",
       "      <td>Virus Diseases</td>\n",
       "      <td>D014777</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV-1 infection</td>\n",
       "      <td>EFO:0000180</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV infection</td>\n",
       "      <td>EFO:0000764</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>AIDS</td>\n",
       "      <td>EFO:0000765</td>\n",
       "      <td>Acquired Immunodeficiency Syndrome</td>\n",
       "      <td>D000163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>infection</td>\n",
       "      <td>EFO:0000544</td>\n",
       "      <td>Infections</td>\n",
       "      <td>D007239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>tuberculosis</td>\n",
       "      <td>MONDO:0018076</td>\n",
       "      <td>Tuberculosis</td>\n",
       "      <td>D014376</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>viral disease</td>\n",
       "      <td>EFO:0000763</td>\n",
       "      <td>Virus Diseases</td>\n",
       "      <td>D014777</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV-1 infection</td>\n",
       "      <td>EFO:0000180</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV infection</td>\n",
       "      <td>EFO:0000764</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CHEMBL440498</td>\n",
       "      <td>CTS-1027</td>\n",
       "      <td>O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...</td>\n",
       "      <td>2.24</td>\n",
       "      <td>2.25</td>\n",
       "      <td>101.93</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.00</td>\n",
       "      <td>Matrix metalloproteinase 9</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Matrix metalloproteinase 9 inhibitor</td>\n",
       "      <td>hepatitis C virus infection</td>\n",
       "      <td>EFO:0003047</td>\n",
       "      <td>Hepatitis C</td>\n",
       "      <td>D006526</td>\n",
       "      <td>hydroxamic acid</td>\n",
       "      <td>8.522879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CHEMBL440498</td>\n",
       "      <td>CTS-1027</td>\n",
       "      <td>O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...</td>\n",
       "      <td>2.24</td>\n",
       "      <td>2.25</td>\n",
       "      <td>101.93</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.00</td>\n",
       "      <td>Matrix metalloproteinase 9</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Matrix metalloproteinase 9 inhibitor</td>\n",
       "      <td>hepatitis C virus infection</td>\n",
       "      <td>EFO:0003047</td>\n",
       "      <td>Hepatitis C</td>\n",
       "      <td>D006526</td>\n",
       "      <td>Oxygen-nitrogen single bond</td>\n",
       "      <td>8.522879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CHEMBL440498</td>\n",
       "      <td>CTS-1027</td>\n",
       "      <td>O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...</td>\n",
       "      <td>2.24</td>\n",
       "      <td>2.25</td>\n",
       "      <td>101.93</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.00</td>\n",
       "      <td>Matrix metalloproteinase 9</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Matrix metalloproteinase 9 inhibitor</td>\n",
       "      <td>hepatitis C virus infection</td>\n",
       "      <td>EFO:0003047</td>\n",
       "      <td>Hepatitis C</td>\n",
       "      <td>D006526</td>\n",
       "      <td>hydroxamate_warhead</td>\n",
       "      <td>8.522879</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         drug_id   drug_name  \\\n",
       "0   CHEMBL223228   EFAVIRENZ   \n",
       "1   CHEMBL223228   EFAVIRENZ   \n",
       "2   CHEMBL223228   EFAVIRENZ   \n",
       "3   CHEMBL223228   EFAVIRENZ   \n",
       "4   CHEMBL223228   EFAVIRENZ   \n",
       "5   CHEMBL223228   EFAVIRENZ   \n",
       "6   CHEMBL223228   EFAVIRENZ   \n",
       "7   CHEMBL223228   EFAVIRENZ   \n",
       "8   CHEMBL223228   EFAVIRENZ   \n",
       "9   CHEMBL223228   EFAVIRENZ   \n",
       "10  CHEMBL223228   EFAVIRENZ   \n",
       "11  CHEMBL223228   EFAVIRENZ   \n",
       "12  CHEMBL223228   EFAVIRENZ   \n",
       "13  CHEMBL223228   EFAVIRENZ   \n",
       "14  CHEMBL223228   EFAVIRENZ   \n",
       "15      CHEMBL57  NEVIRAPINE   \n",
       "16      CHEMBL57  NEVIRAPINE   \n",
       "17      CHEMBL57  NEVIRAPINE   \n",
       "18      CHEMBL57  NEVIRAPINE   \n",
       "19      CHEMBL57  NEVIRAPINE   \n",
       "20      CHEMBL57  NEVIRAPINE   \n",
       "21      CHEMBL57  NEVIRAPINE   \n",
       "22      CHEMBL57  NEVIRAPINE   \n",
       "23      CHEMBL57  NEVIRAPINE   \n",
       "24      CHEMBL57  NEVIRAPINE   \n",
       "25      CHEMBL57  NEVIRAPINE   \n",
       "26      CHEMBL57  NEVIRAPINE   \n",
       "27  CHEMBL440498    CTS-1027   \n",
       "28  CHEMBL440498    CTS-1027   \n",
       "29  CHEMBL440498    CTS-1027   \n",
       "\n",
       "                                               SMILES  logD  logP     psa  \\\n",
       "0        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "1        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "2        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "3        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "4        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "5        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "6        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "7        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "8        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "9        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "10       O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "11       O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "12       O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "13       O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "14       O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "15                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "16                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "17                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "18                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "19                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "20                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "21                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "22                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "23                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "24                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "25                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "26                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "27  O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...  2.24  2.25  101.93   \n",
       "28  O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...  2.24  2.25  101.93   \n",
       "29  O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...  2.24  2.25  101.93   \n",
       "\n",
       "    drug_likeness  max_phase      IC50  \\\n",
       "0            0.73        4.0      0.75   \n",
       "1            0.73        4.0      0.75   \n",
       "2            0.73        4.0      0.75   \n",
       "3            0.73        4.0      0.75   \n",
       "4            0.73        4.0      0.75   \n",
       "5            0.73        4.0      0.75   \n",
       "6            0.73        4.0      0.75   \n",
       "7            0.73        4.0      0.75   \n",
       "8            0.73        4.0      0.75   \n",
       "9            0.73        4.0      0.75   \n",
       "10           0.73        4.0      0.75   \n",
       "11           0.73        4.0      0.75   \n",
       "12           0.73        4.0      0.75   \n",
       "13           0.73        4.0      0.75   \n",
       "14           0.73        4.0      0.75   \n",
       "15           0.86        4.0  23000.00   \n",
       "16           0.86        4.0  23000.00   \n",
       "17           0.86        4.0  23000.00   \n",
       "18           0.86        4.0  23000.00   \n",
       "19           0.86        4.0  23000.00   \n",
       "20           0.86        4.0  23000.00   \n",
       "21           0.86        4.0    230.00   \n",
       "22           0.86        4.0    230.00   \n",
       "23           0.86        4.0    230.00   \n",
       "24           0.86        4.0    230.00   \n",
       "25           0.86        4.0    230.00   \n",
       "26           0.86        4.0    230.00   \n",
       "27           0.54        2.0      3.00   \n",
       "28           0.54        2.0      3.00   \n",
       "29           0.54        2.0      3.00   \n",
       "\n",
       "                                               target  \\\n",
       "0   Human immunodeficiency virus type 1 reverse tr...   \n",
       "1   Human immunodeficiency virus type 1 reverse tr...   \n",
       "2   Human immunodeficiency virus type 1 reverse tr...   \n",
       "3   Human immunodeficiency virus type 1 reverse tr...   \n",
       "4   Human immunodeficiency virus type 1 reverse tr...   \n",
       "5   Human immunodeficiency virus type 1 reverse tr...   \n",
       "6   Human immunodeficiency virus type 1 reverse tr...   \n",
       "7   Human immunodeficiency virus type 1 reverse tr...   \n",
       "8   Human immunodeficiency virus type 1 reverse tr...   \n",
       "9   Human immunodeficiency virus type 1 reverse tr...   \n",
       "10  Human immunodeficiency virus type 1 reverse tr...   \n",
       "11  Human immunodeficiency virus type 1 reverse tr...   \n",
       "12  Human immunodeficiency virus type 1 reverse tr...   \n",
       "13  Human immunodeficiency virus type 1 reverse tr...   \n",
       "14  Human immunodeficiency virus type 1 reverse tr...   \n",
       "15  Human immunodeficiency virus type 1 reverse tr...   \n",
       "16  Human immunodeficiency virus type 1 reverse tr...   \n",
       "17  Human immunodeficiency virus type 1 reverse tr...   \n",
       "18  Human immunodeficiency virus type 1 reverse tr...   \n",
       "19  Human immunodeficiency virus type 1 reverse tr...   \n",
       "20  Human immunodeficiency virus type 1 reverse tr...   \n",
       "21  Human immunodeficiency virus type 1 reverse tr...   \n",
       "22  Human immunodeficiency virus type 1 reverse tr...   \n",
       "23  Human immunodeficiency virus type 1 reverse tr...   \n",
       "24  Human immunodeficiency virus type 1 reverse tr...   \n",
       "25  Human immunodeficiency virus type 1 reverse tr...   \n",
       "26  Human immunodeficiency virus type 1 reverse tr...   \n",
       "27                         Matrix metalloproteinase 9   \n",
       "28                         Matrix metalloproteinase 9   \n",
       "29                         Matrix metalloproteinase 9   \n",
       "\n",
       "                          organism     target_type  \\\n",
       "0   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "1   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "2   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "3   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "4   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "5   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "6   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "7   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "8   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "9   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "10  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "11  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "12  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "13  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "14  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "15  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "16  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "17  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "18  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "19  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "20  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "21  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "22  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "23  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "24  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "25  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "26  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "27                    Homo sapiens  SINGLE PROTEIN   \n",
       "28                    Homo sapiens  SINGLE PROTEIN   \n",
       "29                    Homo sapiens  SINGLE PROTEIN   \n",
       "\n",
       "                                  mechanism_of_action  \\\n",
       "0   Human immunodeficiency virus type 1 reverse tr...   \n",
       "1   Human immunodeficiency virus type 1 reverse tr...   \n",
       "2   Human immunodeficiency virus type 1 reverse tr...   \n",
       "3   Human immunodeficiency virus type 1 reverse tr...   \n",
       "4   Human immunodeficiency virus type 1 reverse tr...   \n",
       "5   Human immunodeficiency virus type 1 reverse tr...   \n",
       "6   Human immunodeficiency virus type 1 reverse tr...   \n",
       "7   Human immunodeficiency virus type 1 reverse tr...   \n",
       "8   Human immunodeficiency virus type 1 reverse tr...   \n",
       "9   Human immunodeficiency virus type 1 reverse tr...   \n",
       "10  Human immunodeficiency virus type 1 reverse tr...   \n",
       "11  Human immunodeficiency virus type 1 reverse tr...   \n",
       "12  Human immunodeficiency virus type 1 reverse tr...   \n",
       "13  Human immunodeficiency virus type 1 reverse tr...   \n",
       "14  Human immunodeficiency virus type 1 reverse tr...   \n",
       "15  Human immunodeficiency virus type 1 reverse tr...   \n",
       "16  Human immunodeficiency virus type 1 reverse tr...   \n",
       "17  Human immunodeficiency virus type 1 reverse tr...   \n",
       "18  Human immunodeficiency virus type 1 reverse tr...   \n",
       "19  Human immunodeficiency virus type 1 reverse tr...   \n",
       "20  Human immunodeficiency virus type 1 reverse tr...   \n",
       "21  Human immunodeficiency virus type 1 reverse tr...   \n",
       "22  Human immunodeficiency virus type 1 reverse tr...   \n",
       "23  Human immunodeficiency virus type 1 reverse tr...   \n",
       "24  Human immunodeficiency virus type 1 reverse tr...   \n",
       "25  Human immunodeficiency virus type 1 reverse tr...   \n",
       "26  Human immunodeficiency virus type 1 reverse tr...   \n",
       "27               Matrix metalloproteinase 9 inhibitor   \n",
       "28               Matrix metalloproteinase 9 inhibitor   \n",
       "29               Matrix metalloproteinase 9 inhibitor   \n",
       "\n",
       "                         efo_term         efo_id  \\\n",
       "0                            AIDS    EFO:0000765   \n",
       "1        immunodeficiency disease  MONDO:0021094   \n",
       "2        type 2 diabetes mellitus  MONDO:0005148   \n",
       "3     hepatitis C virus infection    EFO:0003047   \n",
       "4                       infection    EFO:0000544   \n",
       "5           non-Hodgkins lymphoma    EFO:0005952   \n",
       "6                         malaria    EFO:0001068   \n",
       "7                        neoplasm    EFO:0000616   \n",
       "8            pancreatic carcinoma    EFO:0002618   \n",
       "9                    tuberculosis  MONDO:0018076   \n",
       "10                  viral disease    EFO:0000763   \n",
       "11                HIV-1 infection    EFO:0000180   \n",
       "12                  HIV infection    EFO:0000764   \n",
       "13                prostate cancer  MONDO:0008315   \n",
       "14  triple-negative breast cancer    EFO:0005537   \n",
       "15                           AIDS    EFO:0000765   \n",
       "16                      infection    EFO:0000544   \n",
       "17                   tuberculosis  MONDO:0018076   \n",
       "18                  viral disease    EFO:0000763   \n",
       "19                HIV-1 infection    EFO:0000180   \n",
       "20                  HIV infection    EFO:0000764   \n",
       "21                           AIDS    EFO:0000765   \n",
       "22                      infection    EFO:0000544   \n",
       "23                   tuberculosis  MONDO:0018076   \n",
       "24                  viral disease    EFO:0000763   \n",
       "25                HIV-1 infection    EFO:0000180   \n",
       "26                  HIV infection    EFO:0000764   \n",
       "27    hepatitis C virus infection    EFO:0003047   \n",
       "28    hepatitis C virus infection    EFO:0003047   \n",
       "29    hepatitis C virus infection    EFO:0003047   \n",
       "\n",
       "                                 mesh_heading  mesh_id  \\\n",
       "0          Acquired Immunodeficiency Syndrome  D000163   \n",
       "1                         Autoimmune Diseases  D001327   \n",
       "2                   Diabetes Mellitus, Type 2  D003924   \n",
       "3                                 Hepatitis C  D006526   \n",
       "4                                  Infections  D007239   \n",
       "5                       Lymphoma, Non-Hodgkin  D008228   \n",
       "6                                     Malaria  D008288   \n",
       "7                                   Neoplasms  D009369   \n",
       "8                        Pancreatic Neoplasms  D010190   \n",
       "9                                Tuberculosis  D014376   \n",
       "10                             Virus Diseases  D014777   \n",
       "11                             HIV Infections  D015658   \n",
       "12                             HIV Infections  D015658   \n",
       "13  Prostatic Neoplasms, Castration-Resistant  D064129   \n",
       "14           Triple Negative Breast Neoplasms  D064726   \n",
       "15         Acquired Immunodeficiency Syndrome  D000163   \n",
       "16                                 Infections  D007239   \n",
       "17                               Tuberculosis  D014376   \n",
       "18                             Virus Diseases  D014777   \n",
       "19                             HIV Infections  D015658   \n",
       "20                             HIV Infections  D015658   \n",
       "21         Acquired Immunodeficiency Syndrome  D000163   \n",
       "22                                 Infections  D007239   \n",
       "23                               Tuberculosis  D014376   \n",
       "24                             Virus Diseases  D014777   \n",
       "25                             HIV Infections  D015658   \n",
       "26                             HIV Infections  D015658   \n",
       "27                                Hepatitis C  D006526   \n",
       "28                                Hepatitis C  D006526   \n",
       "29                                Hepatitis C  D006526   \n",
       "\n",
       "                 toxicity_alert     pIC50  \n",
       "0                   triple bond  9.124939  \n",
       "1                   triple bond  9.124939  \n",
       "2                   triple bond  9.124939  \n",
       "3                   triple bond  9.124939  \n",
       "4                   triple bond  9.124939  \n",
       "5                   triple bond  9.124939  \n",
       "6                   triple bond  9.124939  \n",
       "7                   triple bond  9.124939  \n",
       "8                   triple bond  9.124939  \n",
       "9                   triple bond  9.124939  \n",
       "10                  triple bond  9.124939  \n",
       "11                  triple bond  9.124939  \n",
       "12                  triple bond  9.124939  \n",
       "13                  triple bond  9.124939  \n",
       "14                  triple bond  9.124939  \n",
       "15                          NaN  4.638272  \n",
       "16                          NaN  4.638272  \n",
       "17                          NaN  4.638272  \n",
       "18                          NaN  4.638272  \n",
       "19                          NaN  4.638272  \n",
       "20                          NaN  4.638272  \n",
       "21                          NaN  6.638272  \n",
       "22                          NaN  6.638272  \n",
       "23                          NaN  6.638272  \n",
       "24                          NaN  6.638272  \n",
       "25                          NaN  6.638272  \n",
       "26                          NaN  6.638272  \n",
       "27              hydroxamic acid  8.522879  \n",
       "28  Oxygen-nitrogen single bond  8.522879  \n",
       "29          hydroxamate_warhead  8.522879  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af75f155-9812-4d95-9519-dd85b255caa3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🔢 Unique drugs: 850\n"
     ]
    }
   ],
   "source": [
    "# Number of unique drugs\n",
    "num_unique_drugs = df['drug_id'].nunique()\n",
    "print(f\"🔢 Unique drugs: {num_unique_drugs}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "73234c02-5817-4ad8-b5b5-c4f36e24c1f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Human immunodeficiency virus type 1 reverse transcriptase',\n",
       "       'Matrix metalloproteinase 9', 'Matrix metalloproteinase 13',\n",
       "       'Steroid 5-alpha-reductase 2', 'Androgen Receptor',\n",
       "       'Xanthine dehydrogenase', 'FK506-binding protein 1A',\n",
       "       'Human immunodeficiency virus type 1 protease', 'Cyclophilin A',\n",
       "       'Ileal bile acid transporter', 'Nitric oxide synthase, inducible',\n",
       "       'Matrix metalloproteinase-1', 'Matrix metalloproteinase-2',\n",
       "       'Matrix metalloproteinase 3', 'Tyrosine-protein kinase SRC',\n",
       "       'Serotonin 1d (5-HT1d) receptor', 'Carbonic anhydrase II',\n",
       "       'Dihydrofolate reductase', 'Cyclooxygenase-2', 'Aldose reductase',\n",
       "       'Thymidylate synthase', 'Cytochrome P450 17A1',\n",
       "       'Muscarinic acetylcholine receptor M3',\n",
       "       'Farnesyl diphosphate synthase', 'HMG-CoA reductase',\n",
       "       'Epidermal growth factor receptor erbB1',\n",
       "       'Receptor protein-tyrosine kinase erbB-2',\n",
       "       'Angiotensin-converting enzyme', 'Neuraminidase',\n",
       "       'Mu opioid receptor', 'Plasminogen activator inhibitor-1',\n",
       "       'Kappa opioid receptor', 'LDL-associated phospholipase A2',\n",
       "       'Histone deacetylase 1', 'DNA topoisomerase I',\n",
       "       'Acetylcholinesterase', 'Neurokinin 1 receptor',\n",
       "       'Type-1 angiotensin II receptor', 'Dihydroorotate dehydrogenase',\n",
       "       'Endothelin receptor ET-A', 'Estrogen receptor alpha',\n",
       "       'Progesterone receptor', 'Cytochrome P450 19A1',\n",
       "       'Thromboxane-A synthase', 'Serotonin 1b (5-HT1b) receptor',\n",
       "       'Mitogen-activated protein kinase kinase kinase 9',\n",
       "       'Mitogen-activated protein kinase kinase kinase 10',\n",
       "       'Mitogen-activated protein kinase kinase kinase 11',\n",
       "       'Isoleucyl-tRNA synthetase', 'Vasopressin V2 receptor',\n",
       "       'Melanocortin receptor 1', 'Arachidonate 5-lipoxygenase',\n",
       "       'Protein kinase C beta', 'Pancreatic lipase',\n",
       "       'Matrix metalloproteinase 7', 'Dopamine D2 receptor',\n",
       "       'Steryl-sulfatase', 'Glucocorticoid receptor',\n",
       "       'Estrogen receptor beta', 'Cyclin-dependent kinase 1', 'HERG',\n",
       "       'Dipeptidyl peptidase IV', 'Phosphodiesterase 5A', 'Thrombin',\n",
       "       'Bacterial dihydrofolate reductase', 'Histamine H1 receptor',\n",
       "       'Neurokinin 2 receptor', 'Muscarinic acetylcholine receptor M4',\n",
       "       'Angiotensin II type 2 (AT-2) receptor',\n",
       "       'Retinoic acid receptor alpha',\n",
       "       'Muscarinic acetylcholine receptor M1',\n",
       "       'Dual specificity mitogen-activated protein kinase kinase 1',\n",
       "       'Vascular endothelial growth factor receptor 2',\n",
       "       'MAP kinase p38 alpha', 'Cyclin-dependent kinase 2',\n",
       "       'Serotonin 1a (5-HT1a) receptor', 'Ghrelin receptor',\n",
       "       'Carbonic anhydrase I', 'Dihydropyrimidine dehydrogenase',\n",
       "       'Phospholipase A2 group IIA',\n",
       "       'Platelet-derived growth factor receptor beta',\n",
       "       'Prostanoid IP receptor', 'Endothelin receptor ET-B',\n",
       "       'Stem cell growth factor receptor',\n",
       "       'Tyrosine-protein kinase receptor FLT3',\n",
       "       'Cholesteryl ester transfer protein',\n",
       "       'Peroxisome proliferator-activated receptor alpha', 'Tyrosinase',\n",
       "       'Thymidine phosphorylase',\n",
       "       'Peroxisome proliferator-activated receptor gamma',\n",
       "       'Cytochrome P450 11A1', 'Squalene synthetase',\n",
       "       'Human herpesvirus 1 DNA polymerase', 'Dopamine transporter',\n",
       "       'Tyrosine-protein kinase TIE-2', 'Serotonin 2a (5-HT2a) receptor',\n",
       "       'Thromboxane A2 receptor',\n",
       "       'Peroxisome proliferator-activated receptor delta',\n",
       "       'Serotonin 4 (5-HT4) receptor',\n",
       "       'Gonadotropin-releasing hormone receptor', 'Histone deacetylase 2',\n",
       "       'Histone deacetylase 6', 'Alpha-galactosidase A',\n",
       "       'Cyclin-dependent kinase 6', 'Cannabinoid CB1 receptor',\n",
       "       'Glycogen synthase kinase-3 beta', 'Adenosine A3 receptor',\n",
       "       'Voltage-gated N-type calcium channel alpha-1B subunit',\n",
       "       'Norepinephrine transporter', 'G protein-coupled receptor 44',\n",
       "       'Coagulation factor X', 'Tyrosine-protein kinase ABL',\n",
       "       'Beta-1 adrenergic receptor',\n",
       "       'Platelet activating factor receptor', 'Cathepsin K',\n",
       "       'Oxytocin receptor', 'Receptor protein-tyrosine kinase erbB-4',\n",
       "       'Human immunodeficiency virus type 1 integrase', 'ADAM17',\n",
       "       'Vanilloid receptor', 'Vitamin D receptor',\n",
       "       'Interleukin-8 receptor B', 'Interleukin-8 receptor A',\n",
       "       'Bradykinin B1 receptor', 'Neurokinin 3 receptor',\n",
       "       'Prostanoid DP receptor',\n",
       "       'Macrophage colony stimulating factor receptor',\n",
       "       'Equilibrative nucleoside transporter 1',\n",
       "       'PI3-kinase p110-delta subunit', 'PI3-kinase p110-gamma subunit',\n",
       "       'Tyrosine-protein kinase Lyn',\n",
       "       'Calcitonin gene-related peptide type 1 receptor',\n",
       "       'Adenosine A1 receptor', 'Histone deacetylase 3',\n",
       "       'Monoamine oxidase B',\n",
       "       'Intermediate conductance calcium-activated potassium channel protein 4',\n",
       "       'Histone deacetylase 8', 'Purinergic receptor P2Y12',\n",
       "       'P-glycoprotein 1', 'Solute carrier family 22 member 12',\n",
       "       'C-X-C chemokine receptor type 4',\n",
       "       'Hepatocyte growth factor receptor',\n",
       "       'ALK tyrosine kinase receptor', 'GAR transformylase',\n",
       "       'Monoamine oxidase A', 'Fibroblast growth factor receptor 3',\n",
       "       'Tyrosine-protein kinase JAK2', 'Tyrosine-protein kinase JAK3',\n",
       "       'Neurotensin receptor 1', 'Glucagon-like peptide 1 receptor',\n",
       "       'Corticotropin releasing factor receptor 1',\n",
       "       'Histamine H2 receptor', 'Histone deacetylase 11',\n",
       "       'Histamine H3 receptor', 'Serine/threonine-protein kinase PLK1',\n",
       "       'Serine/threonine-protein kinase Aurora-A',\n",
       "       'Serine/threonine-protein kinase Aurora-B',\n",
       "       'Lysosomal alpha-glucosidase',\n",
       "       'Acyl coenzyme A:cholesterol acyltransferase 1',\n",
       "       'Beta-2 adrenergic receptor', 'Matrix metalloproteinase 14',\n",
       "       'Smoothened homolog', 'Nerve growth factor receptor Trk-A',\n",
       "       'Sodium/glucose cotransporter 2',\n",
       "       'Insulin-like growth factor I receptor',\n",
       "       'DNA-(apurinic or apyrimidinic site) lyase',\n",
       "       'Neuropeptide Y receptor type 5',\n",
       "       'Tyrosine-protein kinase receptor RET',\n",
       "       'Serine/threonine-protein kinase mTOR',\n",
       "       'Apoptosis regulator Bcl-X', 'Ceramide glucosyltransferase',\n",
       "       'Maltase-glucoamylase', 'Mineralocorticoid receptor',\n",
       "       'Tyrosine-protein kinase JAK1', 'Adenosine A2a receptor',\n",
       "       'Poly [ADP-ribose] polymerase 2', 'Poly [ADP-ribose] polymerase-1',\n",
       "       'Phosphodiesterase 10A',\n",
       "       'Vascular endothelial growth factor receptor 3',\n",
       "       'Cysteinyl leukotriene receptor 1', 'Cyclin-dependent kinase 7',\n",
       "       'Cyclin-dependent kinase 9',\n",
       "       'Serine/threonine-protein kinase Chk1',\n",
       "       'Transient receptor potential cation channel subfamily M member 8',\n",
       "       'Tyrosine-protein kinase HCK', 'Cyclin-dependent kinase 5',\n",
       "       'Cyclin-dependent kinase 4', 'Cholecystokinin A receptor',\n",
       "       'Anandamide amidohydrolase', 'C-C chemokine receptor type 5',\n",
       "       'Casein kinase II alpha', 'Nicotinamide phosphoribosyltransferase',\n",
       "       '11-beta-hydroxysteroid dehydrogenase 1', 'P2X purinoceptor 7',\n",
       "       'Phosphodiesterase 3A', 'Muscarinic acetylcholine receptor M2',\n",
       "       'Calcium sensing receptor', 'Kinesin-like protein 1',\n",
       "       'Serotonin transporter', 'Serotonin 2c (5-HT2c) receptor',\n",
       "       'Dopamine D3 receptor', 'Histamine H4 receptor',\n",
       "       'Tyrosine-protein kinase BTK', '5-lipoxygenase activating protein',\n",
       "       'Serine/threonine-protein kinase B-raf', 'Delta opioid receptor',\n",
       "       'c-Jun N-terminal kinase 1', 'c-Jun N-terminal kinase 2',\n",
       "       'c-Jun N-terminal kinase 3', 'Cannabinoid CB2 receptor',\n",
       "       'Solute carrier family 22 member 6',\n",
       "       'Solute carrier family 22 member 11',\n",
       "       'Solute carrier family 22 member 8',\n",
       "       'Vascular endothelial growth factor receptor 1', 'Renin',\n",
       "       'Glucagon receptor', 'Beta-glucocerebrosidase',\n",
       "       'Tyrosine-protein kinase SYK', 'C-C chemokine receptor type 9',\n",
       "       'C-C chemokine receptor type 2', 'Phosphodiesterase 9A',\n",
       "       'Serotonin 3a (5-HT3a) receptor', 'Thioredoxin',\n",
       "       'Sodium channel protein type IX alpha subunit', 'Plasminogen',\n",
       "       'Metabotropic glutamate receptor 5', 'Amyloid-beta A4 protein',\n",
       "       'PI3-kinase p110-alpha subunit',\n",
       "       'Acyl coenzyme A:cholesterol acyltransferase 2',\n",
       "       'Receptor tyrosine-protein kinase erbB-3', 'GABA transporter 1',\n",
       "       'Glutathione reductase', 'Cytidine deaminase',\n",
       "       'Carbonic anhydrase XII', 'Insulin receptor',\n",
       "       'TGF-beta receptor type I', 'Tyrosine-protein kinase TYK2',\n",
       "       'Sphingosine 1-phosphate receptor Edg-1',\n",
       "       'Metabotropic glutamate receptor 2', 'Apoptosis regulator Bcl-2',\n",
       "       'C-C chemokine receptor type 1', 'Matrix metalloproteinase 8',\n",
       "       'Free fatty acid receptor 2', 'PI3-kinase p110-beta subunit',\n",
       "       'Serine/threonine-protein kinase PLK4',\n",
       "       'Sodium channel protein type V alpha subunit',\n",
       "       'Glycogen synthase kinase-3 alpha', 'DNA-dependent protein kinase',\n",
       "       'MAP kinase ERK2', 'Sodium channel protein type X alpha subunit',\n",
       "       'Prostanoid EP4 receptor', 'MAP kinase ERK1',\n",
       "       'Indoleamine 2,3-dioxygenase', 'Thyroid peroxidase',\n",
       "       'Probable protein-cysteine N-palmitoyltransferase porcupine',\n",
       "       'Egl nine homolog 1', 'Egl nine homolog 3',\n",
       "       'G-protein coupled receptor 84', 'Proteinase activated receptor 4',\n",
       "       'Alpha-1d adrenergic receptor', 'Matrix metalloproteinase 12',\n",
       "       'Glutaminase kidney isoform, mitochondrial',\n",
       "       'Cytochrome P450 26A1',\n",
       "       'Proto-oncogene tyrosine-protein kinase ROS',\n",
       "       'Tyrosine-protein kinase receptor UFO',\n",
       "       'Alpha-1a adrenergic receptor', 'Catechol O-methyltransferase',\n",
       "       'Epoxide hydratase', 'Autotaxin', 'Dipeptidyl peptidase I',\n",
       "       'Fibroblast growth factor receptor 2',\n",
       "       'Fibroblast growth factor receptor 1', 'Transthyretin',\n",
       "       'Mitogen-activated protein kinase kinase kinase 5',\n",
       "       'Voltage-gated potassium channel subunit Kv1.5',\n",
       "       'Receptor-type tyrosine-protein phosphatase beta',\n",
       "       'Sodium/hydrogen exchanger 3',\n",
       "       'Platelet-derived growth factor receptor alpha',\n",
       "       'Ephrin type-B receptor 4', 'Beta-secretase 1',\n",
       "       'Hypoxia-inducible factor prolyl hydroxylase 1',\n",
       "       'Focal adhesion kinase 1', 'Fibroblast growth factor receptor 4',\n",
       "       'Receptor-interacting serine/threonine-protein kinase 1',\n",
       "       'Sodium/glucose cotransporter 1',\n",
       "       'Interleukin-1 receptor-associated kinase 4',\n",
       "       'Sphingosine kinase 2',\n",
       "       'Isocitrate dehydrogenase [NADP] cytoplasmic',\n",
       "       'DNA (cytosine-5)-methyltransferase 1',\n",
       "       'Serotonin 6 (5-HT6) receptor', 'Serine-protein kinase ATR',\n",
       "       'Parathyroid hormone receptor',\n",
       "       'Macrophage migration inhibitory factor',\n",
       "       'Synaptic vesicular amine transporter', 'Glycine transporter 1',\n",
       "       'Toll-like receptor 4', 'Proteinase-activated receptor 1',\n",
       "       'Complement factor D', 'Serine/threonine-protein kinase WEE1',\n",
       "       'Alcohol dehydrogenase class III', 'Leukocyte elastase',\n",
       "       'Rho-associated protein kinase 2',\n",
       "       'Diacylglycerol O-acyltransferase 1', 'Monoglyceride lipase',\n",
       "       'Myeloperoxidase', 'Endothelial PAS domain-containing protein 1',\n",
       "       'Nonstructural protein 5A',\n",
       "       'Dual specificity mitogen-activated protein kinase kinase 2',\n",
       "       'Cell division cycle 7-related protein kinase',\n",
       "       'Prostanoid EP2 receptor',\n",
       "       'Protein arginine N-methyltransferase 5',\n",
       "       'Dual specificity protein kinase TTK', 'GTPase KRas', 'Nischarin',\n",
       "       'Serine/threonine-protein kinase Chk2', 'Phosphodiesterase 4D',\n",
       "       'Serine/threonine-protein kinase RAF', 'Apoptosis regulator Bcl-W',\n",
       "       'P2X purinoceptor 3', 'Enoyl-[acyl-carrier-protein] reductase',\n",
       "       'Complement factor B', 'DNA topoisomerase II alpha',\n",
       "       'Vasopressin V1a receptor', 'Protein-tyrosine phosphatase 2C',\n",
       "       'Isocitrate dehydrogenase [NADP], mitochondrial',\n",
       "       'FAD-dependent decaprenylphosphoryl-beta-D-ribofuranose 2-oxidase',\n",
       "       'Cytochrome P450 51', 'Aldehyde dehydrogenase',\n",
       "       'Protein tyrosine kinase 2 beta', 'Serotonin 1f (5-HT1f) receptor',\n",
       "       'Replicase polyprotein 1ab', 'Leukotriene A4 hydrolase', 'ADAMTS5',\n",
       "       'NT-3 growth factor receptor',\n",
       "       'Neurotrophic tyrosine kinase receptor type 2', 'Galectin-3',\n",
       "       'Cholesterol 24-hydroxylase',\n",
       "       '1-phosphatidylinositol 3-phosphate 5-kinase',\n",
       "       'Niemann-Pick C1-like protein 1',\n",
       "       'Advanced glycosylation end product-specific receptor',\n",
       "       'Polymerase acidic protein',\n",
       "       'Lysine-specific histone demethylase 1', 'Exportin-1',\n",
       "       'Hepatitis C virus NS5B RNA-dependent RNA polymerase',\n",
       "       'Polymerase basic protein 2'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['target'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "eb5d202d-0860-4895-b832-d8171d1e0de3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ BioGPT-Large loaded and ready.\n"
     ]
    }
   ],
   "source": [
    "from transformers import AutoTokenizer, AutoModelForCausalLM\n",
    "\n",
    "# Load BioGPT model from Microsoft\n",
    "model_name = \"microsoft/BioGPT-Large\"\n",
    "biogpt_tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "biogpt_model = AutoModelForCausalLM.from_pretrained(model_name)\n",
    "\n",
    "print(\"✅ BioGPT-Large loaded and ready.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "25402df1-7196-4391-9fff-4cb4d934ad15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "📦 BioGPT-Large saved to: ./saved_models/biogpt-large\n"
     ]
    }
   ],
   "source": [
    "save_directory = \"./saved_models/biogpt-large\"\n",
    "\n",
    "# Save model\n",
    "biogpt_model.save_pretrained(save_directory)\n",
    "\n",
    "# Save tokenizer\n",
    "biogpt_tokenizer.save_pretrained(save_directory)\n",
    "\n",
    "print(\"📦 BioGPT-Large saved to:\", save_directory)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1ca67537-62a2-47b9-92b6-51778d9b7069",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: peft in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (0.15.2)Note: you may need to restart the kernel to use updated packages.\n",
      "\n",
      "Collecting bitsandbytes\n",
      "  Downloading bitsandbytes-0.46.1-py3-none-win_amd64.whl.metadata (10 kB)\n",
      "Requirement already satisfied: accelerate in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (1.8.1)\n",
      "Requirement already satisfied: transformers in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (4.41.0)\n",
      "Requirement already satisfied: datasets in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (3.6.0)\n",
      "Requirement already satisfied: numpy>=1.17 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from peft) (1.26.4)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from peft) (24.2)\n",
      "Requirement already satisfied: psutil in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from peft) (6.0.0)\n",
      "Requirement already satisfied: pyyaml in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from peft) (6.0.2)\n",
      "Requirement already satisfied: torch>=1.13.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from peft) (2.5.1)\n",
      "Requirement already satisfied: tqdm in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from peft) (4.67.1)\n",
      "Requirement already satisfied: safetensors in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from peft) (0.5.2)\n",
      "Requirement already satisfied: huggingface_hub>=0.25.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from peft) (0.27.1)\n",
      "Requirement already satisfied: filelock in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from torch>=1.13.0->peft) (3.15.4)\n",
      "Requirement already satisfied: typing-extensions>=4.8.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from torch>=1.13.0->peft) (4.12.2)\n",
      "Requirement already satisfied: networkx in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from torch>=1.13.0->peft) (3.2.1)\n",
      "Requirement already satisfied: jinja2 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from torch>=1.13.0->peft) (3.1.4)\n",
      "Requirement already satisfied: fsspec in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from torch>=1.13.0->peft) (2024.6.1)\n",
      "Requirement already satisfied: setuptools in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from torch>=1.13.0->peft) (80.9.0)\n",
      "Requirement already satisfied: sympy==1.13.1 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from torch>=1.13.0->peft) (1.13.1)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from sympy==1.13.1->torch>=1.13.0->peft) (1.3.0)\n",
      "Requirement already satisfied: regex!=2019.12.17 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from transformers) (2024.7.24)\n",
      "Requirement already satisfied: requests in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from transformers) (2.32.3)\n",
      "Requirement already satisfied: tokenizers<0.20,>=0.19 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from transformers) (0.19.1)\n",
      "Requirement already satisfied: pyarrow>=15.0.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from datasets) (15.0.2)\n",
      "Requirement already satisfied: dill<0.3.9,>=0.3.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from datasets) (0.3.8)\n",
      "Requirement already satisfied: pandas in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from datasets) (2.1.1)\n",
      "Requirement already satisfied: xxhash in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from datasets) (3.5.0)\n",
      "Requirement already satisfied: multiprocess<0.70.17 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from datasets) (0.70.16)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from fsspec[http]<=2025.3.0,>=2023.1.0->datasets) (3.10.5)\n",
      "Requirement already satisfied: aiohappyeyeballs>=2.3.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.3.0,>=2023.1.0->datasets) (2.4.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.3.0,>=2023.1.0->datasets) (1.3.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.3.0,>=2023.1.0->datasets) (24.2.0)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.3.0,>=2023.1.0->datasets) (1.4.1)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.3.0,>=2023.1.0->datasets) (6.0.5)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.3.0,>=2023.1.0->datasets) (1.9.8)\n",
      "Requirement already satisfied: idna>=2.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from yarl<2.0,>=1.0->aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<=2025.3.0,>=2023.1.0->datasets) (3.7)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from requests->transformers) (3.3.2)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from requests->transformers) (2.2.2)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from requests->transformers) (2024.7.4)\n",
      "Requirement already satisfied: colorama in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from tqdm->peft) (0.4.6)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from jinja2->torch>=1.13.0->peft) (2.1.5)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from pandas->datasets) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from pandas->datasets) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.1 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from pandas->datasets) (2024.1)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\konde\\appdata\\local\\programs\\python\\python312\\lib\\site-packages (from python-dateutil>=2.8.2->pandas->datasets) (1.16.0)\n",
      "Downloading bitsandbytes-0.46.1-py3-none-win_amd64.whl (72.2 MB)\n",
      "   ---------------------------------------- 0.0/72.2 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/72.2 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.3/72.2 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.3/72.2 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.5/72.2 MB 837.5 kB/s eta 0:01:26\n",
      "   ---------------------------------------- 0.8/72.2 MB 1.0 MB/s eta 0:01:11\n",
      "    --------------------------------------- 1.3/72.2 MB 1.2 MB/s eta 0:00:58\n",
      "    --------------------------------------- 1.6/72.2 MB 1.3 MB/s eta 0:00:54\n",
      "   - -------------------------------------- 1.8/72.2 MB 1.4 MB/s eta 0:00:50\n",
      "   - -------------------------------------- 2.1/72.2 MB 1.4 MB/s eta 0:00:51\n",
      "   - -------------------------------------- 2.4/72.2 MB 1.4 MB/s eta 0:00:52\n",
      "   - -------------------------------------- 2.6/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   - -------------------------------------- 2.9/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   - -------------------------------------- 3.1/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   - -------------------------------------- 3.4/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   -- ------------------------------------- 3.7/72.2 MB 1.3 MB/s eta 0:00:54\n",
      "   -- ------------------------------------- 3.9/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   -- ------------------------------------- 4.2/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   -- ------------------------------------- 4.5/72.2 MB 1.3 MB/s eta 0:00:54\n",
      "   -- ------------------------------------- 4.7/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   -- ------------------------------------- 5.0/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   -- ------------------------------------- 5.2/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   --- ------------------------------------ 5.5/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   --- ------------------------------------ 5.8/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   --- ------------------------------------ 6.0/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   --- ------------------------------------ 6.3/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   --- ------------------------------------ 6.6/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   --- ------------------------------------ 6.8/72.2 MB 1.3 MB/s eta 0:00:53\n",
      "   --- ------------------------------------ 7.1/72.2 MB 1.2 MB/s eta 0:00:53\n",
      "   ---- ----------------------------------- 7.3/72.2 MB 1.2 MB/s eta 0:00:52\n",
      "   ---- ----------------------------------- 7.6/72.2 MB 1.2 MB/s eta 0:00:52\n",
      "   ---- ----------------------------------- 7.9/72.2 MB 1.3 MB/s eta 0:00:52\n",
      "   ---- ----------------------------------- 8.1/72.2 MB 1.2 MB/s eta 0:00:52\n",
      "   ---- ----------------------------------- 8.4/72.2 MB 1.2 MB/s eta 0:00:52\n",
      "   ---- ----------------------------------- 8.7/72.2 MB 1.2 MB/s eta 0:00:52\n",
      "   ---- ----------------------------------- 8.9/72.2 MB 1.2 MB/s eta 0:00:51\n",
      "   ----- ---------------------------------- 9.2/72.2 MB 1.2 MB/s eta 0:00:51\n",
      "   ----- ---------------------------------- 9.4/72.2 MB 1.2 MB/s eta 0:00:51\n",
      "   ----- ---------------------------------- 9.7/72.2 MB 1.2 MB/s eta 0:00:51\n",
      "   ----- ---------------------------------- 10.0/72.2 MB 1.2 MB/s eta 0:00:51\n",
      "   ----- ---------------------------------- 10.2/72.2 MB 1.2 MB/s eta 0:00:50\n",
      "   ----- ---------------------------------- 10.5/72.2 MB 1.2 MB/s eta 0:00:50\n",
      "   ----- ---------------------------------- 10.7/72.2 MB 1.2 MB/s eta 0:00:50\n",
      "   ------ --------------------------------- 11.0/72.2 MB 1.2 MB/s eta 0:00:50\n",
      "   ------ --------------------------------- 11.3/72.2 MB 1.2 MB/s eta 0:00:50\n",
      "   ------ --------------------------------- 11.5/72.2 MB 1.2 MB/s eta 0:00:50\n",
      "   ------ --------------------------------- 11.8/72.2 MB 1.2 MB/s eta 0:00:49\n",
      "   ------ --------------------------------- 12.1/72.2 MB 1.2 MB/s eta 0:00:49\n",
      "   ------ --------------------------------- 12.3/72.2 MB 1.2 MB/s eta 0:00:49\n",
      "   ------ --------------------------------- 12.6/72.2 MB 1.2 MB/s eta 0:00:49\n",
      "   ------- -------------------------------- 12.8/72.2 MB 1.2 MB/s eta 0:00:49\n",
      "   ------- -------------------------------- 12.8/72.2 MB 1.2 MB/s eta 0:00:49\n",
      "   ------- -------------------------------- 13.4/72.2 MB 1.2 MB/s eta 0:00:48\n",
      "   ------- -------------------------------- 13.6/72.2 MB 1.2 MB/s eta 0:00:48\n",
      "   ------- -------------------------------- 13.9/72.2 MB 1.2 MB/s eta 0:00:48\n",
      "   ------- -------------------------------- 13.9/72.2 MB 1.2 MB/s eta 0:00:48\n",
      "   ------- -------------------------------- 14.2/72.2 MB 1.2 MB/s eta 0:00:48\n",
      "   ------- -------------------------------- 14.4/72.2 MB 1.2 MB/s eta 0:00:47\n",
      "   -------- ------------------------------- 14.7/72.2 MB 1.2 MB/s eta 0:00:47\n",
      "   -------- ------------------------------- 14.9/72.2 MB 1.2 MB/s eta 0:00:47\n",
      "   -------- ------------------------------- 15.2/72.2 MB 1.2 MB/s eta 0:00:47\n",
      "   -------- ------------------------------- 15.5/72.2 MB 1.2 MB/s eta 0:00:47\n",
      "   -------- ------------------------------- 15.7/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   -------- ------------------------------- 16.0/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   --------- ------------------------------ 16.3/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   --------- ------------------------------ 16.5/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   --------- ------------------------------ 16.5/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   --------- ------------------------------ 16.5/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   --------- ------------------------------ 16.5/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   --------- ------------------------------ 16.5/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   --------- ------------------------------ 16.8/72.2 MB 1.1 MB/s eta 0:00:49\n",
      "   --------- ------------------------------ 17.3/72.2 MB 1.2 MB/s eta 0:00:48\n",
      "   --------- ------------------------------ 17.8/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   ---------- ----------------------------- 18.1/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   ---------- ----------------------------- 18.4/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   ---------- ----------------------------- 18.6/72.2 MB 1.2 MB/s eta 0:00:46\n",
      "   ---------- ----------------------------- 18.9/72.2 MB 1.2 MB/s eta 0:00:45\n",
      "   ---------- ----------------------------- 19.1/72.2 MB 1.2 MB/s eta 0:00:45\n",
      "   ---------- ----------------------------- 19.4/72.2 MB 1.2 MB/s eta 0:00:45\n",
      "   ---------- ----------------------------- 19.7/72.2 MB 1.2 MB/s eta 0:00:45\n",
      "   ----------- ---------------------------- 19.9/72.2 MB 1.2 MB/s eta 0:00:44\n",
      "   ----------- ---------------------------- 20.2/72.2 MB 1.2 MB/s eta 0:00:44\n",
      "   ----------- ---------------------------- 20.4/72.2 MB 1.2 MB/s eta 0:00:44\n",
      "   ----------- ---------------------------- 20.7/72.2 MB 1.2 MB/s eta 0:00:44\n",
      "   ----------- ---------------------------- 21.0/72.2 MB 1.2 MB/s eta 0:00:44\n",
      "   ----------- ---------------------------- 21.2/72.2 MB 1.2 MB/s eta 0:00:43\n",
      "   ----------- ---------------------------- 21.5/72.2 MB 1.2 MB/s eta 0:00:43\n",
      "   ----------- ---------------------------- 21.5/72.2 MB 1.2 MB/s eta 0:00:43\n",
      "   ------------ --------------------------- 21.8/72.2 MB 1.2 MB/s eta 0:00:43\n",
      "   ------------ --------------------------- 22.0/72.2 MB 1.2 MB/s eta 0:00:43\n",
      "   ------------ --------------------------- 22.3/72.2 MB 1.2 MB/s eta 0:00:42\n",
      "   ------------ --------------------------- 22.8/72.2 MB 1.2 MB/s eta 0:00:42\n",
      "   ------------ --------------------------- 22.8/72.2 MB 1.2 MB/s eta 0:00:42\n",
      "   ------------ --------------------------- 23.1/72.2 MB 1.2 MB/s eta 0:00:42\n",
      "   ------------ --------------------------- 23.3/72.2 MB 1.2 MB/s eta 0:00:41\n",
      "   ------------- -------------------------- 23.9/72.2 MB 1.2 MB/s eta 0:00:41\n",
      "   ------------- -------------------------- 23.9/72.2 MB 1.2 MB/s eta 0:00:41\n",
      "   ------------- -------------------------- 24.1/72.2 MB 1.2 MB/s eta 0:00:41\n",
      "   ------------- -------------------------- 24.4/72.2 MB 1.2 MB/s eta 0:00:41\n",
      "   ------------- -------------------------- 24.4/72.2 MB 1.2 MB/s eta 0:00:41\n",
      "   ------------- -------------------------- 24.4/72.2 MB 1.2 MB/s eta 0:00:41\n",
      "   ------------- -------------------------- 24.4/72.2 MB 1.2 MB/s eta 0:00:41\n",
      "   ------------- -------------------------- 24.9/72.2 MB 1.2 MB/s eta 0:00:41\n",
      "   -------------- ------------------------- 25.4/72.2 MB 1.2 MB/s eta 0:00:40\n",
      "   -------------- ------------------------- 25.7/72.2 MB 1.2 MB/s eta 0:00:40\n",
      "   -------------- ------------------------- 26.0/72.2 MB 1.2 MB/s eta 0:00:40\n",
      "   -------------- ------------------------- 26.2/72.2 MB 1.2 MB/s eta 0:00:40\n",
      "   -------------- ------------------------- 26.5/72.2 MB 1.2 MB/s eta 0:00:39\n",
      "   -------------- ------------------------- 26.7/72.2 MB 1.2 MB/s eta 0:00:39\n",
      "   -------------- ------------------------- 27.0/72.2 MB 1.2 MB/s eta 0:00:39\n",
      "   --------------- ------------------------ 27.3/72.2 MB 1.2 MB/s eta 0:00:39\n",
      "   --------------- ------------------------ 27.5/72.2 MB 1.2 MB/s eta 0:00:38\n",
      "   --------------- ------------------------ 27.8/72.2 MB 1.2 MB/s eta 0:00:38\n",
      "   --------------- ------------------------ 28.0/72.2 MB 1.2 MB/s eta 0:00:38\n",
      "   --------------- ------------------------ 28.3/72.2 MB 1.2 MB/s eta 0:00:38\n",
      "   --------------- ------------------------ 28.6/72.2 MB 1.2 MB/s eta 0:00:37\n",
      "   --------------- ------------------------ 28.8/72.2 MB 1.2 MB/s eta 0:00:37\n",
      "   ---------------- ----------------------- 29.1/72.2 MB 1.2 MB/s eta 0:00:37\n",
      "   ---------------- ----------------------- 29.1/72.2 MB 1.2 MB/s eta 0:00:37\n",
      "   ---------------- ----------------------- 29.4/72.2 MB 1.2 MB/s eta 0:00:37\n",
      "   ---------------- ----------------------- 29.6/72.2 MB 1.2 MB/s eta 0:00:36\n",
      "   ---------------- ----------------------- 29.9/72.2 MB 1.2 MB/s eta 0:00:36\n",
      "   ---------------- ----------------------- 30.1/72.2 MB 1.2 MB/s eta 0:00:36\n",
      "   ---------------- ----------------------- 30.4/72.2 MB 1.2 MB/s eta 0:00:36\n",
      "   ---------------- ----------------------- 30.7/72.2 MB 1.2 MB/s eta 0:00:36\n",
      "   ----------------- ---------------------- 30.9/72.2 MB 1.2 MB/s eta 0:00:35\n",
      "   ----------------- ---------------------- 31.2/72.2 MB 1.2 MB/s eta 0:00:35\n",
      "   ----------------- ---------------------- 31.5/72.2 MB 1.2 MB/s eta 0:00:35\n",
      "   ----------------- ---------------------- 31.7/72.2 MB 1.2 MB/s eta 0:00:35\n",
      "   ----------------- ---------------------- 32.0/72.2 MB 1.2 MB/s eta 0:00:34\n",
      "   ----------------- ---------------------- 32.0/72.2 MB 1.2 MB/s eta 0:00:34\n",
      "   ----------------- ---------------------- 32.0/72.2 MB 1.2 MB/s eta 0:00:34\n",
      "   ----------------- ---------------------- 32.2/72.2 MB 1.2 MB/s eta 0:00:35\n",
      "   ------------------ --------------------- 32.8/72.2 MB 1.2 MB/s eta 0:00:34\n",
      "   ------------------ --------------------- 33.0/72.2 MB 1.2 MB/s eta 0:00:34\n",
      "   ------------------ --------------------- 33.3/72.2 MB 1.2 MB/s eta 0:00:34\n",
      "   ------------------ --------------------- 33.6/72.2 MB 1.2 MB/s eta 0:00:33\n",
      "   ------------------ --------------------- 33.8/72.2 MB 1.2 MB/s eta 0:00:33\n",
      "   ------------------ --------------------- 34.1/72.2 MB 1.2 MB/s eta 0:00:33\n",
      "   ------------------- -------------------- 34.3/72.2 MB 1.2 MB/s eta 0:00:33\n",
      "   ------------------- -------------------- 34.6/72.2 MB 1.2 MB/s eta 0:00:32\n",
      "   ------------------- -------------------- 34.9/72.2 MB 1.2 MB/s eta 0:00:32\n",
      "   ------------------- -------------------- 35.1/72.2 MB 1.2 MB/s eta 0:00:32\n",
      "   ------------------- -------------------- 35.4/72.2 MB 1.2 MB/s eta 0:00:32\n",
      "   ------------------- -------------------- 35.7/72.2 MB 1.2 MB/s eta 0:00:31\n",
      "   ------------------- -------------------- 35.9/72.2 MB 1.2 MB/s eta 0:00:31\n",
      "   -------------------- ------------------- 36.2/72.2 MB 1.2 MB/s eta 0:00:31\n",
      "   -------------------- ------------------- 36.4/72.2 MB 1.2 MB/s eta 0:00:31\n",
      "   -------------------- ------------------- 36.7/72.2 MB 1.2 MB/s eta 0:00:31\n",
      "   -------------------- ------------------- 37.0/72.2 MB 1.2 MB/s eta 0:00:31\n",
      "   -------------------- ------------------- 37.2/72.2 MB 1.2 MB/s eta 0:00:30\n",
      "   -------------------- ------------------- 37.5/72.2 MB 1.2 MB/s eta 0:00:30\n",
      "   -------------------- ------------------- 37.7/72.2 MB 1.2 MB/s eta 0:00:30\n",
      "   --------------------- ------------------ 38.0/72.2 MB 1.2 MB/s eta 0:00:30\n",
      "   --------------------- ------------------ 38.3/72.2 MB 1.2 MB/s eta 0:00:29\n",
      "   --------------------- ------------------ 38.5/72.2 MB 1.2 MB/s eta 0:00:29\n",
      "   --------------------- ------------------ 38.8/72.2 MB 1.2 MB/s eta 0:00:29\n",
      "   --------------------- ------------------ 39.1/72.2 MB 1.2 MB/s eta 0:00:29\n",
      "   --------------------- ------------------ 39.3/72.2 MB 1.2 MB/s eta 0:00:29\n",
      "   --------------------- ------------------ 39.6/72.2 MB 1.2 MB/s eta 0:00:28\n",
      "   ---------------------- ----------------- 39.8/72.2 MB 1.2 MB/s eta 0:00:28\n",
      "   ---------------------- ----------------- 40.1/72.2 MB 1.2 MB/s eta 0:00:28\n",
      "   ---------------------- ----------------- 40.4/72.2 MB 1.2 MB/s eta 0:00:28\n",
      "   ---------------------- ----------------- 40.6/72.2 MB 1.2 MB/s eta 0:00:27\n",
      "   ---------------------- ----------------- 40.9/72.2 MB 1.2 MB/s eta 0:00:27\n",
      "   ---------------------- ----------------- 40.9/72.2 MB 1.2 MB/s eta 0:00:27\n",
      "   ---------------------- ----------------- 41.2/72.2 MB 1.2 MB/s eta 0:00:27\n",
      "   ---------------------- ----------------- 41.4/72.2 MB 1.2 MB/s eta 0:00:27\n",
      "   ----------------------- ---------------- 41.7/72.2 MB 1.2 MB/s eta 0:00:27\n",
      "   ----------------------- ---------------- 41.9/72.2 MB 1.2 MB/s eta 0:00:26\n",
      "   ----------------------- ---------------- 42.2/72.2 MB 1.2 MB/s eta 0:00:26\n",
      "   ----------------------- ---------------- 42.5/72.2 MB 1.2 MB/s eta 0:00:26\n",
      "   ----------------------- ---------------- 42.7/72.2 MB 1.2 MB/s eta 0:00:26\n",
      "   ----------------------- ---------------- 43.0/72.2 MB 1.2 MB/s eta 0:00:25\n",
      "   ----------------------- ---------------- 43.3/72.2 MB 1.2 MB/s eta 0:00:25\n",
      "   ------------------------ --------------- 43.5/72.2 MB 1.2 MB/s eta 0:00:25\n",
      "   ------------------------ --------------- 43.8/72.2 MB 1.2 MB/s eta 0:00:25\n",
      "   ------------------------ --------------- 44.0/72.2 MB 1.2 MB/s eta 0:00:25\n",
      "   ------------------------ --------------- 44.3/72.2 MB 1.2 MB/s eta 0:00:24\n",
      "   ------------------------ --------------- 44.6/72.2 MB 1.2 MB/s eta 0:00:24\n",
      "   ------------------------ --------------- 44.8/72.2 MB 1.2 MB/s eta 0:00:24\n",
      "   ------------------------ --------------- 45.1/72.2 MB 1.2 MB/s eta 0:00:24\n",
      "   ------------------------- -------------- 45.4/72.2 MB 1.2 MB/s eta 0:00:23\n",
      "   ------------------------- -------------- 45.6/72.2 MB 1.2 MB/s eta 0:00:23\n",
      "   ------------------------- -------------- 45.9/72.2 MB 1.2 MB/s eta 0:00:23\n",
      "   ------------------------- -------------- 46.1/72.2 MB 1.2 MB/s eta 0:00:23\n",
      "   ------------------------- -------------- 46.4/72.2 MB 1.2 MB/s eta 0:00:23\n",
      "   ------------------------- -------------- 46.7/72.2 MB 1.2 MB/s eta 0:00:22\n",
      "   ------------------------- -------------- 46.9/72.2 MB 1.2 MB/s eta 0:00:22\n",
      "   -------------------------- ------------- 47.2/72.2 MB 1.2 MB/s eta 0:00:22\n",
      "   -------------------------- ------------- 47.4/72.2 MB 1.2 MB/s eta 0:00:22\n",
      "   -------------------------- ------------- 47.7/72.2 MB 1.2 MB/s eta 0:00:21\n",
      "   -------------------------- ------------- 47.7/72.2 MB 1.2 MB/s eta 0:00:21\n",
      "   -------------------------- ------------- 47.7/72.2 MB 1.2 MB/s eta 0:00:21\n",
      "   -------------------------- ------------- 48.2/72.2 MB 1.2 MB/s eta 0:00:21\n",
      "   -------------------------- ------------- 48.5/72.2 MB 1.2 MB/s eta 0:00:21\n",
      "   --------------------------- ------------ 48.8/72.2 MB 1.2 MB/s eta 0:00:21\n",
      "   --------------------------- ------------ 49.0/72.2 MB 1.2 MB/s eta 0:00:20\n",
      "   --------------------------- ------------ 49.3/72.2 MB 1.2 MB/s eta 0:00:20\n",
      "   --------------------------- ------------ 49.5/72.2 MB 1.2 MB/s eta 0:00:20\n",
      "   --------------------------- ------------ 49.8/72.2 MB 1.2 MB/s eta 0:00:20\n",
      "   --------------------------- ------------ 50.1/72.2 MB 1.2 MB/s eta 0:00:19\n",
      "   --------------------------- ------------ 50.3/72.2 MB 1.2 MB/s eta 0:00:19\n",
      "   ---------------------------- ----------- 50.6/72.2 MB 1.2 MB/s eta 0:00:19\n",
      "   ---------------------------- ----------- 50.9/72.2 MB 1.2 MB/s eta 0:00:19\n",
      "   ---------------------------- ----------- 51.1/72.2 MB 1.2 MB/s eta 0:00:19\n",
      "   ---------------------------- ----------- 51.4/72.2 MB 1.2 MB/s eta 0:00:18\n",
      "   ---------------------------- ----------- 51.4/72.2 MB 1.2 MB/s eta 0:00:18\n",
      "   ---------------------------- ----------- 51.9/72.2 MB 1.2 MB/s eta 0:00:17\n",
      "   ---------------------------- ----------- 52.2/72.2 MB 1.2 MB/s eta 0:00:17\n",
      "   ----------------------------- ---------- 52.4/72.2 MB 1.2 MB/s eta 0:00:17\n",
      "   ----------------------------- ---------- 52.7/72.2 MB 1.2 MB/s eta 0:00:17\n",
      "   ----------------------------- ---------- 53.0/72.2 MB 1.2 MB/s eta 0:00:17\n",
      "   ----------------------------- ---------- 53.2/72.2 MB 1.2 MB/s eta 0:00:16\n",
      "   ----------------------------- ---------- 53.5/72.2 MB 1.2 MB/s eta 0:00:16\n",
      "   ----------------------------- ---------- 53.7/72.2 MB 1.2 MB/s eta 0:00:16\n",
      "   ----------------------------- ---------- 54.0/72.2 MB 1.2 MB/s eta 0:00:16\n",
      "   ------------------------------ --------- 54.3/72.2 MB 1.2 MB/s eta 0:00:16\n",
      "   ------------------------------ --------- 54.5/72.2 MB 1.2 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 54.8/72.2 MB 1.2 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 55.1/72.2 MB 1.2 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 55.1/72.2 MB 1.2 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 55.3/72.2 MB 1.2 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 55.6/72.2 MB 1.2 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 55.8/72.2 MB 1.2 MB/s eta 0:00:14\n",
      "   ------------------------------- -------- 56.1/72.2 MB 1.2 MB/s eta 0:00:14\n",
      "   ------------------------------- -------- 56.4/72.2 MB 1.2 MB/s eta 0:00:14\n",
      "   ------------------------------- -------- 56.6/72.2 MB 1.2 MB/s eta 0:00:14\n",
      "   ------------------------------- -------- 56.9/72.2 MB 1.2 MB/s eta 0:00:13\n",
      "   ------------------------------- -------- 57.1/72.2 MB 1.2 MB/s eta 0:00:13\n",
      "   ------------------------------- -------- 57.4/72.2 MB 1.2 MB/s eta 0:00:13\n",
      "   ------------------------------- -------- 57.7/72.2 MB 1.2 MB/s eta 0:00:13\n",
      "   -------------------------------- ------- 57.9/72.2 MB 1.2 MB/s eta 0:00:13\n",
      "   -------------------------------- ------- 58.2/72.2 MB 1.2 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 58.5/72.2 MB 1.2 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 58.7/72.2 MB 1.2 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 59.0/72.2 MB 1.2 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 59.2/72.2 MB 1.2 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 59.5/72.2 MB 1.2 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 59.8/72.2 MB 1.2 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 60.0/72.2 MB 1.2 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 60.3/72.2 MB 1.2 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 60.6/72.2 MB 1.2 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 61.3/72.2 MB 1.2 MB/s eta 0:00:09\n",
      "   ---------------------------------- ----- 62.4/72.2 MB 1.3 MB/s eta 0:00:08\n",
      "   ----------------------------------- ---- 63.7/72.2 MB 1.3 MB/s eta 0:00:07\n",
      "   ------------------------------------ --- 66.1/72.2 MB 1.3 MB/s eta 0:00:05\n",
      "   -------------------------------------- - 68.9/72.2 MB 1.4 MB/s eta 0:00:03\n",
      "   ---------------------------------------  72.1/72.2 MB 1.5 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 72.2/72.2 MB 1.5 MB/s eta 0:00:00\n",
      "Installing collected packages: bitsandbytes\n",
      "Successfully installed bitsandbytes-0.46.1\n"
     ]
    }
   ],
   "source": [
    "pip install peft bitsandbytes accelerate transformers datasets\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "85ac7360-9a10-4485-82b8-3e1b6c28592d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Loaded: (553992, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug_id</th>\n",
       "      <th>drug_name</th>\n",
       "      <th>SMILES</th>\n",
       "      <th>logD</th>\n",
       "      <th>logP</th>\n",
       "      <th>psa</th>\n",
       "      <th>drug_likeness</th>\n",
       "      <th>max_phase</th>\n",
       "      <th>IC50</th>\n",
       "      <th>target</th>\n",
       "      <th>organism</th>\n",
       "      <th>target_type</th>\n",
       "      <th>mechanism_of_action</th>\n",
       "      <th>efo_term</th>\n",
       "      <th>efo_id</th>\n",
       "      <th>mesh_heading</th>\n",
       "      <th>mesh_id</th>\n",
       "      <th>toxicity_alert</th>\n",
       "      <th>pIC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>AIDS</td>\n",
       "      <td>EFO:0000765</td>\n",
       "      <td>Acquired Immunodeficiency Syndrome</td>\n",
       "      <td>D000163</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>immunodeficiency disease</td>\n",
       "      <td>MONDO:0021094</td>\n",
       "      <td>Autoimmune Diseases</td>\n",
       "      <td>D001327</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>type 2 diabetes mellitus</td>\n",
       "      <td>MONDO:0005148</td>\n",
       "      <td>Diabetes Mellitus, Type 2</td>\n",
       "      <td>D003924</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>hepatitis C virus infection</td>\n",
       "      <td>EFO:0003047</td>\n",
       "      <td>Hepatitis C</td>\n",
       "      <td>D006526</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>infection</td>\n",
       "      <td>EFO:0000544</td>\n",
       "      <td>Infections</td>\n",
       "      <td>D007239</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>non-Hodgkins lymphoma</td>\n",
       "      <td>EFO:0005952</td>\n",
       "      <td>Lymphoma, Non-Hodgkin</td>\n",
       "      <td>D008228</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>malaria</td>\n",
       "      <td>EFO:0001068</td>\n",
       "      <td>Malaria</td>\n",
       "      <td>D008288</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>neoplasm</td>\n",
       "      <td>EFO:0000616</td>\n",
       "      <td>Neoplasms</td>\n",
       "      <td>D009369</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>pancreatic carcinoma</td>\n",
       "      <td>EFO:0002618</td>\n",
       "      <td>Pancreatic Neoplasms</td>\n",
       "      <td>D010190</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>tuberculosis</td>\n",
       "      <td>MONDO:0018076</td>\n",
       "      <td>Tuberculosis</td>\n",
       "      <td>D014376</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>viral disease</td>\n",
       "      <td>EFO:0000763</td>\n",
       "      <td>Virus Diseases</td>\n",
       "      <td>D014777</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV-1 infection</td>\n",
       "      <td>EFO:0000180</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV infection</td>\n",
       "      <td>EFO:0000764</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>MONDO:0008315</td>\n",
       "      <td>Prostatic Neoplasms, Castration-Resistant</td>\n",
       "      <td>D064129</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHEMBL223228</td>\n",
       "      <td>EFAVIRENZ</td>\n",
       "      <td>O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1</td>\n",
       "      <td>4.46</td>\n",
       "      <td>4.46</td>\n",
       "      <td>38.33</td>\n",
       "      <td>0.73</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>triple-negative breast cancer</td>\n",
       "      <td>EFO:0005537</td>\n",
       "      <td>Triple Negative Breast Neoplasms</td>\n",
       "      <td>D064726</td>\n",
       "      <td>triple bond</td>\n",
       "      <td>9.124939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>AIDS</td>\n",
       "      <td>EFO:0000765</td>\n",
       "      <td>Acquired Immunodeficiency Syndrome</td>\n",
       "      <td>D000163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>infection</td>\n",
       "      <td>EFO:0000544</td>\n",
       "      <td>Infections</td>\n",
       "      <td>D007239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>tuberculosis</td>\n",
       "      <td>MONDO:0018076</td>\n",
       "      <td>Tuberculosis</td>\n",
       "      <td>D014376</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>viral disease</td>\n",
       "      <td>EFO:0000763</td>\n",
       "      <td>Virus Diseases</td>\n",
       "      <td>D014777</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV-1 infection</td>\n",
       "      <td>EFO:0000180</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23000.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV infection</td>\n",
       "      <td>EFO:0000764</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>AIDS</td>\n",
       "      <td>EFO:0000765</td>\n",
       "      <td>Acquired Immunodeficiency Syndrome</td>\n",
       "      <td>D000163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>infection</td>\n",
       "      <td>EFO:0000544</td>\n",
       "      <td>Infections</td>\n",
       "      <td>D007239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>tuberculosis</td>\n",
       "      <td>MONDO:0018076</td>\n",
       "      <td>Tuberculosis</td>\n",
       "      <td>D014376</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>viral disease</td>\n",
       "      <td>EFO:0000763</td>\n",
       "      <td>Virus Diseases</td>\n",
       "      <td>D014777</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV-1 infection</td>\n",
       "      <td>EFO:0000180</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CHEMBL57</td>\n",
       "      <td>NEVIRAPINE</td>\n",
       "      <td>Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>2.49</td>\n",
       "      <td>58.12</td>\n",
       "      <td>0.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>230.00</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Human immunodeficiency virus type 1 reverse tr...</td>\n",
       "      <td>HIV infection</td>\n",
       "      <td>EFO:0000764</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>D015658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.638272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CHEMBL440498</td>\n",
       "      <td>CTS-1027</td>\n",
       "      <td>O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...</td>\n",
       "      <td>2.24</td>\n",
       "      <td>2.25</td>\n",
       "      <td>101.93</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.00</td>\n",
       "      <td>Matrix metalloproteinase 9</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Matrix metalloproteinase 9 inhibitor</td>\n",
       "      <td>hepatitis C virus infection</td>\n",
       "      <td>EFO:0003047</td>\n",
       "      <td>Hepatitis C</td>\n",
       "      <td>D006526</td>\n",
       "      <td>hydroxamic acid</td>\n",
       "      <td>8.522879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CHEMBL440498</td>\n",
       "      <td>CTS-1027</td>\n",
       "      <td>O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...</td>\n",
       "      <td>2.24</td>\n",
       "      <td>2.25</td>\n",
       "      <td>101.93</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.00</td>\n",
       "      <td>Matrix metalloproteinase 9</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Matrix metalloproteinase 9 inhibitor</td>\n",
       "      <td>hepatitis C virus infection</td>\n",
       "      <td>EFO:0003047</td>\n",
       "      <td>Hepatitis C</td>\n",
       "      <td>D006526</td>\n",
       "      <td>Oxygen-nitrogen single bond</td>\n",
       "      <td>8.522879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CHEMBL440498</td>\n",
       "      <td>CTS-1027</td>\n",
       "      <td>O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...</td>\n",
       "      <td>2.24</td>\n",
       "      <td>2.25</td>\n",
       "      <td>101.93</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.00</td>\n",
       "      <td>Matrix metalloproteinase 9</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Matrix metalloproteinase 9 inhibitor</td>\n",
       "      <td>hepatitis C virus infection</td>\n",
       "      <td>EFO:0003047</td>\n",
       "      <td>Hepatitis C</td>\n",
       "      <td>D006526</td>\n",
       "      <td>hydroxamate_warhead</td>\n",
       "      <td>8.522879</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         drug_id   drug_name  \\\n",
       "0   CHEMBL223228   EFAVIRENZ   \n",
       "1   CHEMBL223228   EFAVIRENZ   \n",
       "2   CHEMBL223228   EFAVIRENZ   \n",
       "3   CHEMBL223228   EFAVIRENZ   \n",
       "4   CHEMBL223228   EFAVIRENZ   \n",
       "5   CHEMBL223228   EFAVIRENZ   \n",
       "6   CHEMBL223228   EFAVIRENZ   \n",
       "7   CHEMBL223228   EFAVIRENZ   \n",
       "8   CHEMBL223228   EFAVIRENZ   \n",
       "9   CHEMBL223228   EFAVIRENZ   \n",
       "10  CHEMBL223228   EFAVIRENZ   \n",
       "11  CHEMBL223228   EFAVIRENZ   \n",
       "12  CHEMBL223228   EFAVIRENZ   \n",
       "13  CHEMBL223228   EFAVIRENZ   \n",
       "14  CHEMBL223228   EFAVIRENZ   \n",
       "15      CHEMBL57  NEVIRAPINE   \n",
       "16      CHEMBL57  NEVIRAPINE   \n",
       "17      CHEMBL57  NEVIRAPINE   \n",
       "18      CHEMBL57  NEVIRAPINE   \n",
       "19      CHEMBL57  NEVIRAPINE   \n",
       "20      CHEMBL57  NEVIRAPINE   \n",
       "21      CHEMBL57  NEVIRAPINE   \n",
       "22      CHEMBL57  NEVIRAPINE   \n",
       "23      CHEMBL57  NEVIRAPINE   \n",
       "24      CHEMBL57  NEVIRAPINE   \n",
       "25      CHEMBL57  NEVIRAPINE   \n",
       "26      CHEMBL57  NEVIRAPINE   \n",
       "27  CHEMBL440498    CTS-1027   \n",
       "28  CHEMBL440498    CTS-1027   \n",
       "29  CHEMBL440498    CTS-1027   \n",
       "\n",
       "                                               SMILES  logD  logP     psa  \\\n",
       "0        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "1        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "2        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "3        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "4        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "5        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "6        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "7        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "8        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "9        O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "10       O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "11       O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "12       O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "13       O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "14       O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1  4.46  4.46   38.33   \n",
       "15                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "16                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "17                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "18                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "19                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "20                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "21                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "22                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "23                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "24                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "25                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "26                    Cc1ccnc2c1NC(=O)c1cccnc1N2C1CC1  2.49  2.49   58.12   \n",
       "27  O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...  2.24  2.25  101.93   \n",
       "28  O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...  2.24  2.25  101.93   \n",
       "29  O=C(NO)C1(CS(=O)(=O)c2ccc(Oc3ccc(Cl)cc3)cc2)CC...  2.24  2.25  101.93   \n",
       "\n",
       "    drug_likeness  max_phase      IC50  \\\n",
       "0            0.73        4.0      0.75   \n",
       "1            0.73        4.0      0.75   \n",
       "2            0.73        4.0      0.75   \n",
       "3            0.73        4.0      0.75   \n",
       "4            0.73        4.0      0.75   \n",
       "5            0.73        4.0      0.75   \n",
       "6            0.73        4.0      0.75   \n",
       "7            0.73        4.0      0.75   \n",
       "8            0.73        4.0      0.75   \n",
       "9            0.73        4.0      0.75   \n",
       "10           0.73        4.0      0.75   \n",
       "11           0.73        4.0      0.75   \n",
       "12           0.73        4.0      0.75   \n",
       "13           0.73        4.0      0.75   \n",
       "14           0.73        4.0      0.75   \n",
       "15           0.86        4.0  23000.00   \n",
       "16           0.86        4.0  23000.00   \n",
       "17           0.86        4.0  23000.00   \n",
       "18           0.86        4.0  23000.00   \n",
       "19           0.86        4.0  23000.00   \n",
       "20           0.86        4.0  23000.00   \n",
       "21           0.86        4.0    230.00   \n",
       "22           0.86        4.0    230.00   \n",
       "23           0.86        4.0    230.00   \n",
       "24           0.86        4.0    230.00   \n",
       "25           0.86        4.0    230.00   \n",
       "26           0.86        4.0    230.00   \n",
       "27           0.54        2.0      3.00   \n",
       "28           0.54        2.0      3.00   \n",
       "29           0.54        2.0      3.00   \n",
       "\n",
       "                                               target  \\\n",
       "0   Human immunodeficiency virus type 1 reverse tr...   \n",
       "1   Human immunodeficiency virus type 1 reverse tr...   \n",
       "2   Human immunodeficiency virus type 1 reverse tr...   \n",
       "3   Human immunodeficiency virus type 1 reverse tr...   \n",
       "4   Human immunodeficiency virus type 1 reverse tr...   \n",
       "5   Human immunodeficiency virus type 1 reverse tr...   \n",
       "6   Human immunodeficiency virus type 1 reverse tr...   \n",
       "7   Human immunodeficiency virus type 1 reverse tr...   \n",
       "8   Human immunodeficiency virus type 1 reverse tr...   \n",
       "9   Human immunodeficiency virus type 1 reverse tr...   \n",
       "10  Human immunodeficiency virus type 1 reverse tr...   \n",
       "11  Human immunodeficiency virus type 1 reverse tr...   \n",
       "12  Human immunodeficiency virus type 1 reverse tr...   \n",
       "13  Human immunodeficiency virus type 1 reverse tr...   \n",
       "14  Human immunodeficiency virus type 1 reverse tr...   \n",
       "15  Human immunodeficiency virus type 1 reverse tr...   \n",
       "16  Human immunodeficiency virus type 1 reverse tr...   \n",
       "17  Human immunodeficiency virus type 1 reverse tr...   \n",
       "18  Human immunodeficiency virus type 1 reverse tr...   \n",
       "19  Human immunodeficiency virus type 1 reverse tr...   \n",
       "20  Human immunodeficiency virus type 1 reverse tr...   \n",
       "21  Human immunodeficiency virus type 1 reverse tr...   \n",
       "22  Human immunodeficiency virus type 1 reverse tr...   \n",
       "23  Human immunodeficiency virus type 1 reverse tr...   \n",
       "24  Human immunodeficiency virus type 1 reverse tr...   \n",
       "25  Human immunodeficiency virus type 1 reverse tr...   \n",
       "26  Human immunodeficiency virus type 1 reverse tr...   \n",
       "27                         Matrix metalloproteinase 9   \n",
       "28                         Matrix metalloproteinase 9   \n",
       "29                         Matrix metalloproteinase 9   \n",
       "\n",
       "                          organism     target_type  \\\n",
       "0   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "1   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "2   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "3   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "4   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "5   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "6   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "7   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "8   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "9   Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "10  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "11  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "12  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "13  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "14  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "15  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "16  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "17  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "18  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "19  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "20  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "21  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "22  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "23  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "24  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "25  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "26  Human immunodeficiency virus 1  SINGLE PROTEIN   \n",
       "27                    Homo sapiens  SINGLE PROTEIN   \n",
       "28                    Homo sapiens  SINGLE PROTEIN   \n",
       "29                    Homo sapiens  SINGLE PROTEIN   \n",
       "\n",
       "                                  mechanism_of_action  \\\n",
       "0   Human immunodeficiency virus type 1 reverse tr...   \n",
       "1   Human immunodeficiency virus type 1 reverse tr...   \n",
       "2   Human immunodeficiency virus type 1 reverse tr...   \n",
       "3   Human immunodeficiency virus type 1 reverse tr...   \n",
       "4   Human immunodeficiency virus type 1 reverse tr...   \n",
       "5   Human immunodeficiency virus type 1 reverse tr...   \n",
       "6   Human immunodeficiency virus type 1 reverse tr...   \n",
       "7   Human immunodeficiency virus type 1 reverse tr...   \n",
       "8   Human immunodeficiency virus type 1 reverse tr...   \n",
       "9   Human immunodeficiency virus type 1 reverse tr...   \n",
       "10  Human immunodeficiency virus type 1 reverse tr...   \n",
       "11  Human immunodeficiency virus type 1 reverse tr...   \n",
       "12  Human immunodeficiency virus type 1 reverse tr...   \n",
       "13  Human immunodeficiency virus type 1 reverse tr...   \n",
       "14  Human immunodeficiency virus type 1 reverse tr...   \n",
       "15  Human immunodeficiency virus type 1 reverse tr...   \n",
       "16  Human immunodeficiency virus type 1 reverse tr...   \n",
       "17  Human immunodeficiency virus type 1 reverse tr...   \n",
       "18  Human immunodeficiency virus type 1 reverse tr...   \n",
       "19  Human immunodeficiency virus type 1 reverse tr...   \n",
       "20  Human immunodeficiency virus type 1 reverse tr...   \n",
       "21  Human immunodeficiency virus type 1 reverse tr...   \n",
       "22  Human immunodeficiency virus type 1 reverse tr...   \n",
       "23  Human immunodeficiency virus type 1 reverse tr...   \n",
       "24  Human immunodeficiency virus type 1 reverse tr...   \n",
       "25  Human immunodeficiency virus type 1 reverse tr...   \n",
       "26  Human immunodeficiency virus type 1 reverse tr...   \n",
       "27               Matrix metalloproteinase 9 inhibitor   \n",
       "28               Matrix metalloproteinase 9 inhibitor   \n",
       "29               Matrix metalloproteinase 9 inhibitor   \n",
       "\n",
       "                         efo_term         efo_id  \\\n",
       "0                            AIDS    EFO:0000765   \n",
       "1        immunodeficiency disease  MONDO:0021094   \n",
       "2        type 2 diabetes mellitus  MONDO:0005148   \n",
       "3     hepatitis C virus infection    EFO:0003047   \n",
       "4                       infection    EFO:0000544   \n",
       "5           non-Hodgkins lymphoma    EFO:0005952   \n",
       "6                         malaria    EFO:0001068   \n",
       "7                        neoplasm    EFO:0000616   \n",
       "8            pancreatic carcinoma    EFO:0002618   \n",
       "9                    tuberculosis  MONDO:0018076   \n",
       "10                  viral disease    EFO:0000763   \n",
       "11                HIV-1 infection    EFO:0000180   \n",
       "12                  HIV infection    EFO:0000764   \n",
       "13                prostate cancer  MONDO:0008315   \n",
       "14  triple-negative breast cancer    EFO:0005537   \n",
       "15                           AIDS    EFO:0000765   \n",
       "16                      infection    EFO:0000544   \n",
       "17                   tuberculosis  MONDO:0018076   \n",
       "18                  viral disease    EFO:0000763   \n",
       "19                HIV-1 infection    EFO:0000180   \n",
       "20                  HIV infection    EFO:0000764   \n",
       "21                           AIDS    EFO:0000765   \n",
       "22                      infection    EFO:0000544   \n",
       "23                   tuberculosis  MONDO:0018076   \n",
       "24                  viral disease    EFO:0000763   \n",
       "25                HIV-1 infection    EFO:0000180   \n",
       "26                  HIV infection    EFO:0000764   \n",
       "27    hepatitis C virus infection    EFO:0003047   \n",
       "28    hepatitis C virus infection    EFO:0003047   \n",
       "29    hepatitis C virus infection    EFO:0003047   \n",
       "\n",
       "                                 mesh_heading  mesh_id  \\\n",
       "0          Acquired Immunodeficiency Syndrome  D000163   \n",
       "1                         Autoimmune Diseases  D001327   \n",
       "2                   Diabetes Mellitus, Type 2  D003924   \n",
       "3                                 Hepatitis C  D006526   \n",
       "4                                  Infections  D007239   \n",
       "5                       Lymphoma, Non-Hodgkin  D008228   \n",
       "6                                     Malaria  D008288   \n",
       "7                                   Neoplasms  D009369   \n",
       "8                        Pancreatic Neoplasms  D010190   \n",
       "9                                Tuberculosis  D014376   \n",
       "10                             Virus Diseases  D014777   \n",
       "11                             HIV Infections  D015658   \n",
       "12                             HIV Infections  D015658   \n",
       "13  Prostatic Neoplasms, Castration-Resistant  D064129   \n",
       "14           Triple Negative Breast Neoplasms  D064726   \n",
       "15         Acquired Immunodeficiency Syndrome  D000163   \n",
       "16                                 Infections  D007239   \n",
       "17                               Tuberculosis  D014376   \n",
       "18                             Virus Diseases  D014777   \n",
       "19                             HIV Infections  D015658   \n",
       "20                             HIV Infections  D015658   \n",
       "21         Acquired Immunodeficiency Syndrome  D000163   \n",
       "22                                 Infections  D007239   \n",
       "23                               Tuberculosis  D014376   \n",
       "24                             Virus Diseases  D014777   \n",
       "25                             HIV Infections  D015658   \n",
       "26                             HIV Infections  D015658   \n",
       "27                                Hepatitis C  D006526   \n",
       "28                                Hepatitis C  D006526   \n",
       "29                                Hepatitis C  D006526   \n",
       "\n",
       "                 toxicity_alert     pIC50  \n",
       "0                   triple bond  9.124939  \n",
       "1                   triple bond  9.124939  \n",
       "2                   triple bond  9.124939  \n",
       "3                   triple bond  9.124939  \n",
       "4                   triple bond  9.124939  \n",
       "5                   triple bond  9.124939  \n",
       "6                   triple bond  9.124939  \n",
       "7                   triple bond  9.124939  \n",
       "8                   triple bond  9.124939  \n",
       "9                   triple bond  9.124939  \n",
       "10                  triple bond  9.124939  \n",
       "11                  triple bond  9.124939  \n",
       "12                  triple bond  9.124939  \n",
       "13                  triple bond  9.124939  \n",
       "14                  triple bond  9.124939  \n",
       "15                          NaN  4.638272  \n",
       "16                          NaN  4.638272  \n",
       "17                          NaN  4.638272  \n",
       "18                          NaN  4.638272  \n",
       "19                          NaN  4.638272  \n",
       "20                          NaN  4.638272  \n",
       "21                          NaN  6.638272  \n",
       "22                          NaN  6.638272  \n",
       "23                          NaN  6.638272  \n",
       "24                          NaN  6.638272  \n",
       "25                          NaN  6.638272  \n",
       "26                          NaN  6.638272  \n",
       "27              hydroxamic acid  8.522879  \n",
       "28  Oxygen-nitrogen single bond  8.522879  \n",
       "29          hydroxamate_warhead  8.522879  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load your dataset\n",
    "df = pd.read_csv(\"data/cleaned_clinical_drugs_dataset.csv\")\n",
    "print(\"✅ Loaded:\", df.shape)\n",
    "df.head(30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8c085f11-447b-46ae-9170-b2a1613911f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🧼 After cleaning: (553987, 7)\n"
     ]
    }
   ],
   "source": [
    "columns_to_keep = [\n",
    "    \"drug_name\", \"SMILES\", \"target\", \"mechanism_of_action\",\n",
    "    \"pIC50\", \"efo_term\", \"mesh_heading\"\n",
    "]\n",
    "\n",
    "df = df[columns_to_keep].dropna()\n",
    "print(\"🧼 After cleaning:\", df.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f6f00908-1962-4738-afe2-32839115706c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "📄 Sample Prompt:\n",
      " ### Instruction:\n",
      "Given the following drug information, predict the target and mechanism.\n",
      "\n",
      "Drug Name: EFAVIRENZ\n",
      "SMILES: O=C1Nc2ccc(Cl)cc2[C@@](C#CC2CC2)(C(F)(F)F)O1\n",
      "\n",
      "### Response:\n",
      "Target: Human immunodeficiency virus type 1 reverse transcriptase\n",
      "Mechanism: Human immunodeficiency virus type 1 reverse transcriptase inhibitor\n"
     ]
    }
   ],
   "source": [
    "def create_prompt(row):\n",
    "    prompt = f\"\"\"### Instruction:\n",
    "Given the following drug information, predict the target and mechanism.\n",
    "\n",
    "Drug Name: {row['drug_name']}\n",
    "SMILES: {row['SMILES']}\n",
    "\n",
    "### Response:\n",
    "Target: {row['target']}\n",
    "Mechanism: {row['mechanism_of_action']}\"\"\"\n",
    "    return prompt\n",
    "\n",
    "df['text'] = df.apply(create_prompt, axis=1)\n",
    "formatted_df = df[['text']].rename(columns={'text': 'text'})\n",
    "\n",
    "print(\"📄 Sample Prompt:\\n\", formatted_df.iloc[0]['text'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "089c45d7-20a0-47fe-a4ae-fd7c926b06b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "💾 Saved formatted dataset to formatted_data.jsonl\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "output_path = \"formatted_data.jsonl\"\n",
    "\n",
    "with open(output_path, \"w\", encoding=\"utf-8\") as f:\n",
    "    for record in formatted_df['text']:\n",
    "        json.dump({\"text\": record}, f)\n",
    "        f.write(\"\\n\")\n",
    "\n",
    "print(f\"💾 Saved formatted dataset to {output_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b4343e6d-d843-41ea-8ddf-90c1c320da2c",
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "Torch not compiled with CUDA enabled",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[3], line 13\u001b[0m\n\u001b[0;32m      7\u001b[0m tokenizer \u001b[38;5;241m=\u001b[39m AutoTokenizer\u001b[38;5;241m.\u001b[39mfrom_pretrained(model_name)\n\u001b[0;32m      9\u001b[0m \u001b[38;5;66;03m# Load the model in FP16 and manually push to GPU\u001b[39;00m\n\u001b[0;32m     10\u001b[0m model \u001b[38;5;241m=\u001b[39m \u001b[43mAutoModelForCausalLM\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_pretrained\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m     11\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmodel_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m     12\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtorch_dtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfloat16\u001b[49m\n\u001b[1;32m---> 13\u001b[0m \u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcuda\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m     15\u001b[0m \u001b[38;5;66;03m# Prepare for LoRA training (no quantization)\u001b[39;00m\n\u001b[0;32m     16\u001b[0m model \u001b[38;5;241m=\u001b[39m prepare_model_for_kbit_training(model)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\transformers\\modeling_utils.py:2576\u001b[0m, in \u001b[0;36mPreTrainedModel.to\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m   2571\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m dtype_present_in_args:\n\u001b[0;32m   2572\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[0;32m   2573\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mYou cannot cast a GPTQ model in a new `dtype`. Make sure to load the model using `from_pretrained` using the desired\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   2574\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m `dtype` by passing the correct `torch_dtype` argument.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   2575\u001b[0m         )\n\u001b[1;32m-> 2576\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\torch\\nn\\modules\\module.py:1355\u001b[0m, in \u001b[0;36mModule.to\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1352\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m   1353\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m\n\u001b[1;32m-> 1355\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\torch\\nn\\modules\\module.py:915\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    913\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[0;32m    914\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[1;32m--> 915\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    917\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[0;32m    918\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[0;32m    919\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[0;32m    920\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    925\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[0;32m    926\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\torch\\nn\\modules\\module.py:915\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    913\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[0;32m    914\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[1;32m--> 915\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    917\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[0;32m    918\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[0;32m    919\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[0;32m    920\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    925\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[0;32m    926\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\torch\\nn\\modules\\module.py:942\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    938\u001b[0m \u001b[38;5;66;03m# Tensors stored in modules are graph leaves, and we don't want to\u001b[39;00m\n\u001b[0;32m    939\u001b[0m \u001b[38;5;66;03m# track autograd history of `param_applied`, so we have to use\u001b[39;00m\n\u001b[0;32m    940\u001b[0m \u001b[38;5;66;03m# `with torch.no_grad():`\u001b[39;00m\n\u001b[0;32m    941\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m torch\u001b[38;5;241m.\u001b[39mno_grad():\n\u001b[1;32m--> 942\u001b[0m     param_applied \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparam\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    943\u001b[0m p_should_use_set_data \u001b[38;5;241m=\u001b[39m compute_should_use_set_data(param, param_applied)\n\u001b[0;32m    945\u001b[0m \u001b[38;5;66;03m# subclasses may have multiple child tensors so we need to use swap_tensors\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\torch\\nn\\modules\\module.py:1341\u001b[0m, in \u001b[0;36mModule.to.<locals>.convert\u001b[1;34m(t)\u001b[0m\n\u001b[0;32m   1334\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m convert_to_format \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m t\u001b[38;5;241m.\u001b[39mdim() \u001b[38;5;129;01min\u001b[39;00m (\u001b[38;5;241m4\u001b[39m, \u001b[38;5;241m5\u001b[39m):\n\u001b[0;32m   1335\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m t\u001b[38;5;241m.\u001b[39mto(\n\u001b[0;32m   1336\u001b[0m             device,\n\u001b[0;32m   1337\u001b[0m             dtype \u001b[38;5;28;01mif\u001b[39;00m t\u001b[38;5;241m.\u001b[39mis_floating_point() \u001b[38;5;129;01mor\u001b[39;00m t\u001b[38;5;241m.\u001b[39mis_complex() \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m   1338\u001b[0m             non_blocking,\n\u001b[0;32m   1339\u001b[0m             memory_format\u001b[38;5;241m=\u001b[39mconvert_to_format,\n\u001b[0;32m   1340\u001b[0m         )\n\u001b[1;32m-> 1341\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   1342\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1343\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mis_floating_point\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mis_complex\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m   1344\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnon_blocking\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1345\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1346\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m   1347\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mstr\u001b[39m(e) \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot copy out of meta tensor; no data!\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\torch\\cuda\\__init__.py:363\u001b[0m, in \u001b[0;36m_lazy_init\u001b[1;34m()\u001b[0m\n\u001b[0;32m    358\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[0;32m    359\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot re-initialize CUDA in forked subprocess. To use CUDA with \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    360\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmultiprocessing, you must use the \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mspawn\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m start method\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    361\u001b[0m     )\n\u001b[0;32m    362\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(torch\u001b[38;5;241m.\u001b[39m_C, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_cuda_getDeviceCount\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m--> 363\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAssertionError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTorch not compiled with CUDA enabled\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    364\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m _cudart \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    365\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAssertionError\u001b[39;00m(\n\u001b[0;32m    366\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlibcudart functions unavailable. It looks like you have a broken build?\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    367\u001b[0m     )\n",
      "\u001b[1;31mAssertionError\u001b[0m: Torch not compiled with CUDA enabled"
     ]
    }
   ],
   "source": [
    "from transformers import AutoTokenizer, AutoModelForCausalLM\n",
    "from peft import prepare_model_for_kbit_training, LoraConfig, get_peft_model\n",
    "import torch\n",
    "\n",
    "# Load tokenizer and model\n",
    "model_name = \"microsoft/BioGPT-Large\"\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "\n",
    "# Load the model in FP16 and manually push to GPU\n",
    "model = AutoModelForCausalLM.from_pretrained(\n",
    "    model_name,\n",
    "    torch_dtype=torch.float16\n",
    ").to(\"cuda\")\n",
    "\n",
    "# Prepare for LoRA training (no quantization)\n",
    "model = prepare_model_for_kbit_training(model)\n",
    "\n",
    "# LoRA config\n",
    "lora_config = LoraConfig(\n",
    "    r=8,\n",
    "    lora_alpha=32,\n",
    "    target_modules=[\"q_proj\", \"v_proj\"],  # fallback: try [\"c_attn\"] or inspect if error\n",
    "    lora_dropout=0.05,\n",
    "    bias=\"none\",\n",
    "    task_type=\"CAUSAL_LM\"\n",
    ")\n",
    "\n",
    "model = get_peft_model(model, lora_config)\n",
    "\n",
    "# ✅ Show trainable parameters\n",
    "def print_trainable_parameters(model):\n",
    "    trainable = sum(p.numel() for p in model.parameters() if p.requires_grad)\n",
    "    total = sum(p.numel() for p in model.parameters())\n",
    "    print(f\"✅ Trainable params: {trainable:,} / {total:,} ({100 * trainable / total:.2f}%)\")\n",
    "\n",
    "print_trainable_parameters(model)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "15a1e4d5-bae1-4073-8286-6a71f0080107",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "📊 Lines saved in formatted_data.jsonl: 553987\n",
      "📁 Expected rows: 553987\n"
     ]
    }
   ],
   "source": [
    "# Count number of lines in the saved JSONL file\n",
    "with open(\"formatted_data.jsonl\", \"r\", encoding=\"utf-8\") as f:\n",
    "    num_lines = sum(1 for _ in f)\n",
    "\n",
    "print(f\"📊 Lines saved in formatted_data.jsonl: {num_lines}\")\n",
    "print(f\"📁 Expected rows: {len(formatted_df)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "72f0e970-32a9-4c21-ba70-fbdfe606fadb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trainable params: 614,400 || all params: 1,571,803,200 || trainable%: 0.039088863033234694\n"
     ]
    }
   ],
   "source": [
    "from transformers import AutoModelForCausalLM\n",
    "from peft import LoraConfig, get_peft_model, TaskType\n",
    "\n",
    "# Define LoRA configuration\n",
    "peft_config = LoraConfig(\n",
    "    r=4,  # Lower rank to reduce trainable parameters\n",
    "    lora_alpha=16,\n",
    "    target_modules=[\"q_proj\"],  # Only adapting query projection layer\n",
    "    lora_dropout=0.05,\n",
    "    bias=\"none\",\n",
    "    task_type=TaskType.CAUSAL_LM\n",
    ")\n",
    "\n",
    "# Load base BioGPT model\n",
    "base_model = AutoModelForCausalLM.from_pretrained(\"microsoft/BioGPT-Large\")\n",
    "\n",
    "# Apply LoRA configuration\n",
    "model = get_peft_model(base_model, peft_config)\n",
    "\n",
    "# Print trainable parameters\n",
    "model.print_trainable_parameters()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cecfe77a-6a18-4f5a-9373-0b7526a8303f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The installed version of bitsandbytes was compiled without GPU support. 8-bit optimizers and GPU quantization are unavailable.\n",
      "C:\\Users\\konde\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\transformers\\data\\metrics\\__init__.py:19: UserWarning: A NumPy version >=1.23.5 and <2.3.0 is required for this version of SciPy (detected version 2.3.1)\n",
      "  from scipy.stats import pearsonr, spearmanr\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Failed to import transformers.trainer because of the following error (look up to see its traceback):\nnumpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\transformers\\utils\\import_utils.py:1472\u001b[0m, in \u001b[0;36m_LazyModule._get_module\u001b[1;34m(self, module_name)\u001b[0m\n\u001b[0;32m   1471\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 1472\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mimportlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mimport_module\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m.\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mmodule_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;18;43m__name__\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1473\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\importlib\\__init__.py:90\u001b[0m, in \u001b[0;36mimport_module\u001b[1;34m(name, package)\u001b[0m\n\u001b[0;32m     89\u001b[0m         level \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m---> 90\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_bootstrap\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_gcd_import\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[43m[\u001b[49m\u001b[43mlevel\u001b[49m\u001b[43m:\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpackage\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlevel\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m<frozen importlib._bootstrap>:1387\u001b[0m, in \u001b[0;36m_gcd_import\u001b[1;34m(name, package, level)\u001b[0m\n",
      "File \u001b[1;32m<frozen importlib._bootstrap>:1360\u001b[0m, in \u001b[0;36m_find_and_load\u001b[1;34m(name, import_)\u001b[0m\n",
      "File \u001b[1;32m<frozen importlib._bootstrap>:1331\u001b[0m, in \u001b[0;36m_find_and_load_unlocked\u001b[1;34m(name, import_)\u001b[0m\n",
      "File \u001b[1;32m<frozen importlib._bootstrap>:935\u001b[0m, in \u001b[0;36m_load_unlocked\u001b[1;34m(spec)\u001b[0m\n",
      "File \u001b[1;32m<frozen importlib._bootstrap_external>:995\u001b[0m, in \u001b[0;36mexec_module\u001b[1;34m(self, module)\u001b[0m\n",
      "File \u001b[1;32m<frozen importlib._bootstrap>:488\u001b[0m, in \u001b[0;36m_call_with_frames_removed\u001b[1;34m(f, *args, **kwds)\u001b[0m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\transformers\\trainer.py:59\u001b[0m\n\u001b[0;32m     58\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mconfiguration_utils\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m PretrainedConfig\n\u001b[1;32m---> 59\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdata\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdata_collator\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m DataCollator, DataCollatorWithPadding, default_data_collator\n\u001b[0;32m     60\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdebug_utils\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m DebugOption, DebugUnderflowOverflow\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\transformers\\data\\__init__.py:26\u001b[0m\n\u001b[0;32m     15\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdata_collator\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[0;32m     16\u001b[0m     DataCollatorForLanguageModeling,\n\u001b[0;32m     17\u001b[0m     DataCollatorForPermutationLanguageModeling,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     24\u001b[0m     default_data_collator,\n\u001b[0;32m     25\u001b[0m )\n\u001b[1;32m---> 26\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmetrics\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m glue_compute_metrics, xnli_compute_metrics\n\u001b[0;32m     27\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mprocessors\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[0;32m     28\u001b[0m     DataProcessor,\n\u001b[0;32m     29\u001b[0m     InputExample,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     43\u001b[0m     xnli_tasks_num_labels,\n\u001b[0;32m     44\u001b[0m )\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\transformers\\data\\metrics\\__init__.py:20\u001b[0m\n\u001b[0;32m     19\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mscipy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mstats\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m pearsonr, spearmanr\n\u001b[1;32m---> 20\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01msklearn\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmetrics\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m f1_score, matthews_corrcoef\n\u001b[0;32m     23\u001b[0m DEPRECATION_WARNING \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m     24\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThis metric will be removed from the library soon, metrics should be handled with the 🤗 Evaluate \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m     25\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlibrary. You can have a look at this example script for pointers: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m     26\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://github.com/huggingface/transformers/blob/main/examples/pytorch/text-classification/run_glue.py\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m     27\u001b[0m )\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\sklearn\\__init__.py:83\u001b[0m\n\u001b[0;32m     79\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[0;32m     80\u001b[0m     __check_build,  \u001b[38;5;66;03m# noqa: F401\u001b[39;00m\n\u001b[0;32m     81\u001b[0m     _distributor_init,  \u001b[38;5;66;03m# noqa: F401\u001b[39;00m\n\u001b[0;32m     82\u001b[0m )\n\u001b[1;32m---> 83\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m clone\n\u001b[0;32m     84\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_show_versions\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m show_versions\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\sklearn\\base.py:19\u001b[0m\n\u001b[0;32m     18\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mexceptions\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m InconsistentVersionWarning\n\u001b[1;32m---> 19\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m _IS_32BIT\n\u001b[0;32m     20\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_estimator_html_repr\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m estimator_html_repr\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\sklearn\\utils\\__init__.py:27\u001b[0m\n\u001b[0;32m     26\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mfixes\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m parse_version, threadpool_info\n\u001b[1;32m---> 27\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmurmurhash\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m murmurhash3_32\n\u001b[0;32m     28\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mvalidation\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[0;32m     29\u001b[0m     _is_arraylike_not_scalar,\n\u001b[0;32m     30\u001b[0m     as_float_array,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     39\u001b[0m     indexable,\n\u001b[0;32m     40\u001b[0m )\n",
      "File \u001b[1;32msklearn\\utils\\murmurhash.pyx:1\u001b[0m, in \u001b[0;36minit sklearn.utils.murmurhash\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[1], line 4\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtransformers\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AutoTokenizer, AutoModelForCausalLM\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpeft\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m LoraConfig, get_peft_model, TaskType\n\u001b[1;32m----> 4\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtransformers\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m TrainingArguments, Trainer, DataCollatorForLanguageModeling\n\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtorch\u001b[39;00m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;66;03m# Step 1: Load tokenizer\u001b[39;00m\n",
      "File \u001b[1;32m<frozen importlib._bootstrap>:1412\u001b[0m, in \u001b[0;36m_handle_fromlist\u001b[1;34m(module, fromlist, import_, recursive)\u001b[0m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\transformers\\utils\\import_utils.py:1462\u001b[0m, in \u001b[0;36m_LazyModule.__getattr__\u001b[1;34m(self, name)\u001b[0m\n\u001b[0;32m   1460\u001b[0m     value \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_module(name)\n\u001b[0;32m   1461\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m name \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_class_to_module\u001b[38;5;241m.\u001b[39mkeys():\n\u001b[1;32m-> 1462\u001b[0m     module \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_module\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_class_to_module\u001b[49m\u001b[43m[\u001b[49m\u001b[43mname\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1463\u001b[0m     value \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(module, name)\n\u001b[0;32m   1464\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\transformers\\utils\\import_utils.py:1474\u001b[0m, in \u001b[0;36m_LazyModule._get_module\u001b[1;34m(self, module_name)\u001b[0m\n\u001b[0;32m   1472\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m importlib\u001b[38;5;241m.\u001b[39mimport_module(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m module_name, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m)\n\u001b[0;32m   1473\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m-> 1474\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[0;32m   1475\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFailed to import \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmodule_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m because of the following error (look up to see its\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   1476\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m traceback):\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   1477\u001b[0m     ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n",
      "\u001b[1;31mRuntimeError\u001b[0m: Failed to import transformers.trainer because of the following error (look up to see its traceback):\nnumpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject"
     ]
    }
   ],
   "source": [
    "from datasets import load_dataset\n",
    "from transformers import AutoTokenizer, AutoModelForCausalLM\n",
    "from peft import LoraConfig, get_peft_model, TaskType\n",
    "from transformers import TrainingArguments, Trainer, DataCollatorForLanguageModeling\n",
    "import torch\n",
    "\n",
    "# Step 1: Load tokenizer\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"microsoft/BioGPT-Large\")\n",
    "\n",
    "# Step 2: Load formatted dataset\n",
    "dataset = load_dataset(\"json\", data_files=\"formatted_data.jsonl\", split=\"train\")\n",
    "\n",
    "# Step 3: Tokenization\n",
    "def tokenize_function(example):\n",
    "    texts = [p + \" \" + c for p, c in zip(example[\"prompt\"], example[\"completion\"])]\n",
    "    return tokenizer(texts, padding=\"max_length\", truncation=True, max_length=512)\n",
    "\n",
    "tokenized_dataset = dataset.map(tokenize_function, batched=True)\n",
    "\n",
    "# Step 4: LoRA configuration (already optimized)\n",
    "peft_config = LoraConfig(\n",
    "    r=4,\n",
    "    lora_alpha=16,\n",
    "    target_modules=[\"q_proj\"],\n",
    "    lora_dropout=0.05,\n",
    "    bias=\"none\",\n",
    "    task_type=TaskType.CAUSAL_LM\n",
    ")\n",
    "\n",
    "# Step 5: Load base model and apply LoRA\n",
    "base_model = AutoModelForCausalLM.from_pretrained(\"microsoft/BioGPT-Large\")\n",
    "model = get_peft_model(base_model, peft_config)\n",
    "model.print_trainable_parameters()\n",
    "\n",
    "# Step 6: Training arguments\n",
    "training_args = TrainingArguments(\n",
    "    output_dir=\"./biogpt-lora-llm4mol\",\n",
    "    per_device_train_batch_size=4,       # Tune if GPU OOM\n",
    "    gradient_accumulation_steps=4,       # Effective batch size = 16\n",
    "    num_train_epochs=1,\n",
    "    learning_rate=2e-4,\n",
    "    weight_decay=0.01,\n",
    "    logging_dir=\"./logs\",\n",
    "    logging_steps=50,\n",
    "    save_strategy=\"epoch\",\n",
    "    fp16=torch.cuda.is_available(),      # Use FP16 if CUDA available\n",
    "    report_to=\"none\"\n",
    ")\n",
    "\n",
    "# Step 7: Data collator (for causal LM)\n",
    "data_collator = DataCollatorForLanguageModeling(\n",
    "    tokenizer=tokenizer,\n",
    "    mlm=False\n",
    ")\n",
    "\n",
    "# Step 8: Trainer setup\n",
    "trainer = Trainer(\n",
    "    model=model,\n",
    "    args=training_args,\n",
    "    train_dataset=tokenized_dataset,\n",
    "    data_collator=data_collator\n",
    ")\n",
    "\n",
    "# Step 9: Train the model\n",
    "trainer.train()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "919f8699-e608-4a62-9d44-dafb173b0bdf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
